{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import pylab\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "import itertools\n",
    "import operator\n",
    "import subprocess\n",
    "import twobitreader\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pysam\n",
    "import shutil\n",
    "\n",
    "from LAM_scripts.LAM_helpersDanner.py import *\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Pipeline1\n",
    "# This, for the most part, is the UDITAS pipeline modified to allow for REPLACE targeting. \n",
    "\n",
    "Overview:\n",
    "- Trim off the 5 nt on both sides and check for primering (how many to trim off of read 1 or two)\n",
    "- need to check for priming (make it able to check priming on Read1 or two. add missmatches. and if input is trimmed or not. pull primer from data sheet. pull primer using coordiantes and get seq downstream depending on lenght. \n",
    "- trims short amplicons for reads that go into the adapter or illumina primers\n",
    "- local align to the plasmid without the AAV seq at all\n",
    "- local align to the AAV seq without the HDR arms\n",
    "- after pulling out the reads that didn't align to AAV seq or plasmid backbone, analyze the breaks\n",
    "\n",
    "- generate table of expected amplicons (use the hdr sample and just replace seq between breaks if thats possible)\n",
    "- align agasint the reads that didn't map to AAV or plasmid backbone,\n",
    "  look at indels and quantification\n",
    "\n",
    "Test pipeline 2:\n",
    "- Trim up to the cut site. Run end-to-end across the genome. Look for expected off target translocations\n",
    "- also generally measure the frequency at which integrrations happen\n",
    "\n",
    "Questions:\n",
    "- which side is the HDR arm on? which of the two sides should i be looking for extra hdr transcripts from\n",
    "- Can I bring in some of the analysis tools from crispresso to understand the indel profiles better.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/eric/Data/Spaced_Nicking/LAM2_Miniseq_ELANE\n"
     ]
    }
   ],
   "source": [
    "#Directory\n",
    "\n",
    "directory = '/home/eric/Data/Spaced_Nicking/LAM2_Miniseq_ELANE'\n",
    "print(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/eric/Data/Ref_Genomes/hg38.2bit\n",
      "env: BOWTIE2_INDEXES=/home/eric/Data/Ref_Genomes\n"
     ]
    }
   ],
   "source": [
    "##########        Assign the file_genome_2bit location.     ############ \n",
    "#\n",
    "#   This is needed for pulling sequence from the referene genome by location\n",
    "#assembly = amplicon_info['genome']\n",
    "assembly = 'hg38'\n",
    "file_genome_2bit = os.path.join('/home/eric/Data/Ref_Genomes', assembly + '.2bit')\n",
    "print(file_genome_2bit)\n",
    "\n",
    "###############   BOWTIE2_INDEXES for genome alignments    ################\n",
    "#\n",
    "#check in bash: > ECHO $GENOMES_2BIT\n",
    "\n",
    "%env BOWTIE2_INDEXES=/home/eric/Data/Ref_Genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#THIS trimming doesn't compress the trimmed files because there was some error where compressing into a .gz \n",
    "# was clipping the last ~24 reads or something like that.\n",
    "\n",
    "def trimming_R1_R2(dir_sample, amplicon_info, R1trim, R2trim, pipeline):\n",
    "    \n",
    "    \n",
    "    #define the input files and output files\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "    \n",
    "    #input/output files\n",
    "    if pipeline == 1:\n",
    "        #inputs\n",
    "        r1_fastq = create_filename(dir_sample, N7, N5, 'R1fastqgz')\n",
    "        r2_fastq = create_filename(dir_sample, N7, N5, 'R2fastqgz')\n",
    "        \n",
    "         #output files not gz compressed\n",
    "        r1_clipped_fastq = create_filename(dir_sample, N7, N5, 'R1fastq_LAM')\n",
    "        r2_clipped_fastq = create_filename(dir_sample, N7, N5, 'R2fastq_LAM')\n",
    "    \n",
    "    elif pipeline == 2:\n",
    "        print('makingpipeline2 files')\n",
    "        r1_fastq = create_filename(dir_sample, N7, N5, 'unmapped_plasmid_R1fastqgz')\n",
    "        r2_fastq = create_filename(dir_sample, N7, N5, 'unmapped_plasmid_R2fastqgz')\n",
    "        \n",
    "        #output files not gz compressed\n",
    "        r1_clipped_fastq = create_filename(dir_sample, N7, N5, 'breaktrimmed_R1fastq')\n",
    "        r2_clipped_fastq = create_filename(dir_sample, N7, N5, 'breaktrimmed_R2fastq')\n",
    "   \n",
    "\n",
    "    \n",
    "    #open/create all these files\n",
    "    ref_file_r1_fasta_good = open(r1_clipped_fastq, \"w\")\n",
    "    ref_file_r2_fasta_good = open(r2_clipped_fastq, \"w\")\n",
    "\n",
    "      \n",
    "    \n",
    "    # We open r1,r2 files and distribute reads\n",
    "    with open_fastq_or_gz(r1_fastq) as r1_file, open_fastq_or_gz(r2_fastq) as r2_file:\n",
    "\n",
    "        r1_r2 = itertools.izip(r1_file, r2_file)\n",
    "\n",
    "        for header_r1, header_r2 in r1_r2:\n",
    "           \n",
    "            seq_r1, seq_r2 = r1_r2.next()\n",
    "\n",
    "            r1_r2.next()\n",
    "\n",
    "            qual_r1, qual_r2 = r1_r2.next()\n",
    "            seq_r1, seq_r2 = seq_r1.rstrip(), seq_r2.rstrip()\n",
    "            qual_r1, qual_r2 = qual_r1.rstrip(), qual_r2.rstrip()\n",
    "\n",
    "            seq_r1_use = seq_r1[R1trim:]\n",
    "            seq_q1_use = qual_r1[R1trim:]\n",
    "            seq_r2_use = seq_r2[R2trim:]\n",
    "            seq_q2_use = seq_r2[R2trim:]\n",
    "            \n",
    "            r1f = ref_file_r1_fasta_good\n",
    "            r2f = ref_file_r2_fasta_good\n",
    "\n",
    "            print(\"\\n\".join([header_r1.rstrip(), seq_r1_use.rstrip(), \"+\", seq_q1_use.rstrip()]), file=r1f)\n",
    "            print(\"\\n\".join([header_r2.rstrip(), seq_r2_use.rstrip(), \"+\", seq_q2_use.rstrip()]), file=r2f)\n",
    "                    \n",
    "           \n",
    "    #for fo in [r1_clipped_fastq, r2_clipped_fastq]:\n",
    "     #   with open(fo) as f_in, gzip.open(fo + '.gz', 'wb') as f_out:\n",
    "      #      f_out.writelines(f_in)\n",
    "       # os.remove(fo)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read2trim is : 24  nucleotides.\n",
      "done with sample 0 1.1.F N701 N501\n",
      "done with sample 1 1.2.F N701 N502\n",
      "done with sample 2 1.3.F N701 N504\n",
      "done with sample 3 2.1.F N701 N505\n",
      "done with sample 4 2.2.F N701 N506\n",
      "done with sample 5 2.3.F N701 N507\n",
      "done with sample 6 3.1.F N701 N508\n",
      "done with sample 7 3.2.F N701 N510\n",
      "done with sample 8 3.3.F N702 N501\n",
      "done with sample 9 12.1.F N702 N502\n",
      "done with sample 10 12.2.F N702 N504\n",
      "done with sample 11 12.3.F N702 N505\n",
      "done with sample 12 13.1.F N702 N506\n",
      "done with sample 13 13.2.F N702 N507\n",
      "done with sample 14 13.3.F N702 N508\n",
      "done with sample 15 1.1.R N703 N501\n",
      "done with sample 16 1.2.R N703 N502\n",
      "done with sample 17 1.3.R N703 N504\n",
      "done with sample 18 2.1.R N703 N505\n",
      "done with sample 19 2.2.R N703 N506\n",
      "done with sample 20 2.3.R N703 N507\n",
      "done with sample 21 3.1.R N703 N508\n",
      "done with sample 22 3.2.R N703 N510\n",
      "done with sample 23 3.3.R N704 N501\n",
      "done with sample 24 12.1.R N704 N502\n",
      "done with sample 25 12.2.R N704 N504\n",
      "done with sample 26 12.3.R N704 N505\n",
      "done with sample 27 13.1.R N704 N506\n",
      "done with sample 28 13.2.R N704 N507\n",
      "done with sample 29 13.3.R N704 N508\n"
     ]
    }
   ],
   "source": [
    "########## Remove first 5 nts to remove adapter seq and spot generation sequencs  #########\n",
    "# \n",
    "#     Misha adds 4 nt's to help spot generation on Illumina on both read 1 and 2\n",
    "#    Read 1 has gene binding primer\n",
    "#    Read 2 has adapter to ligate on universal reverse seq\n",
    "\n",
    "\n",
    "                    # decide on the length to trim\n",
    "adapter_seq = 'GACTATAGGGCACGCGTGG'\n",
    "adapt_len = len(adapter_seq)\n",
    "read2trim = 5 + adapt_len\n",
    "print('read2trim is :', read2trim, ' nucleotides.')\n",
    "\n",
    "\n",
    "                  # run the trimming\n",
    "for i in range(30):\n",
    "        \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    \n",
    "    trimming_R1_R2(directory, amplicon_info, R1trim = 3, R2trim = read2trim, pipeline = 1)\n",
    "    \n",
    "    #print('done with sample', i)\n",
    "    print('done with sample', i , amplicon_info['name'], amplicon_info['index_I1'],amplicon_info['index_I2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TGTCACAGTGCAGCTCACTCAG\n",
      "CTGAGTGAGCTGCACTGTGACA\n"
     ]
    }
   ],
   "source": [
    "#ignore this\n",
    "\n",
    "#could make an automatic way to generate the primer mismatch and downstream if I want later\n",
    "start = 5226605\n",
    "end = 5226627\n",
    "\n",
    "genome = twobitreader.TwoBitFile(file_genome_2bit)\n",
    "primer = genome['chr11'][int(start):int(end)]\n",
    "print(primer)\n",
    "print(reverse_complement(primer))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Discard Mispriming Reads\n",
    "\n",
    "LAM uses a anchored primer and then gets ride of the background gDNA. Then it uses a nested primer and so there should be a very clean product. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with sample 0\n",
      "done with sample 1\n",
      "done with sample 2\n",
      "done with sample 3\n",
      "done with sample 4\n",
      "done with sample 5\n",
      "done with sample 6\n",
      "done with sample 7\n",
      "done with sample 8\n",
      "done with sample 9\n",
      "done with sample 10\n",
      "done with sample 11\n",
      "done with sample 12\n",
      "done with sample 13\n",
      "done with sample 14\n",
      "done with sample 15\n",
      "done with sample 16\n",
      "done with sample 17\n",
      "done with sample 18\n",
      "done with sample 19\n",
      "done with sample 20\n",
      "done with sample 21\n",
      "done with sample 22\n",
      "done with sample 23\n",
      "done with sample 24\n",
      "done with sample 25\n",
      "done with sample 26\n",
      "done with sample 27\n",
      "done with sample 28\n",
      "done with sample 29\n",
      "            sample_name    i7    i5  total_reads  reads_with_good_priming  \\\n",
      "0         ELANE_AAV_Fwd  N701  N501        70127                    34280   \n",
      "1         ELANE_AAV_Fwd  N701  N502        49097                    18503   \n",
      "2         ELANE_AAV_Fwd  N701  N504        40795                    10553   \n",
      "3      ELANE_Cas9WT_Fwd  N701  N505        33080                     9220   \n",
      "4      ELANE_Cas9WT_Fwd  N701  N506        59458                    14322   \n",
      "5      ELANE_Cas9WT_Fwd  N701  N507       100161                    27822   \n",
      "6        ELANE_D10A_Fwd  N701  N508       100236                    42864   \n",
      "7        ELANE_D10A_Fwd  N701  N510       141824                    63270   \n",
      "8        ELANE_D10A_Fwd  N702  N501        82559                    50752   \n",
      "9   ELANE_Cas9WT_d3_Fwd  N702  N502         6509                     3095   \n",
      "10  ELANE_Cas9WT_d3_Fwd  N702  N504        27838                    13814   \n",
      "11  ELANE_Cas9WT_d3_Fwd  N702  N505         6678                     6293   \n",
      "12    ELANE_D10A_d3_Fwd  N702  N506        92793                    66375   \n",
      "13    ELANE_D10A_d3_Fwd  N702  N507       125967                    99595   \n",
      "14    ELANE_D10A_d3_Fwd  N702  N508       222697                   187454   \n",
      "15        ELANE_AAV_Rev  N703  N501        84528                    77678   \n",
      "16        ELANE_AAV_Rev  N703  N502       130622                   119340   \n",
      "17        ELANE_AAV_Rev  N703  N504        47074                    42699   \n",
      "18     ELANE_Cas9WT_Rev  N703  N505       104228                    96217   \n",
      "19     ELANE_Cas9WT_Rev  N703  N506       111695                   101614   \n",
      "20     ELANE_Cas9WT_Rev  N703  N507       132173                   121394   \n",
      "21       ELANE_D10A_Rev  N703  N508       171550                   160979   \n",
      "22       ELANE_D10A_Rev  N703  N510       158664                   147418   \n",
      "23       ELANE_D10A_Rev  N704  N501        40347                    37903   \n",
      "24  ELANE_Cas9WT_d3_Rev  N704  N502        67628                    62868   \n",
      "25  ELANE_Cas9WT_d3_Rev  N704  N504         2980                     2225   \n",
      "26  ELANE_Cas9WT_d3_Rev  N704  N505        34445                    33073   \n",
      "27    ELANE_D10A_d3_Rev  N704  N506        93910                    88989   \n",
      "28    ELANE_D10A_d3_Rev  N704  N507        25731                    25044   \n",
      "29    ELANE_D10A_d3_Rev  N704  N508        99242                    94279   \n",
      "\n",
      "    reads_with_gene_specific_primer_misprimed  \\\n",
      "0                                       34788   \n",
      "1                                       29627   \n",
      "2                                       29215   \n",
      "3                                       23192   \n",
      "4                                       44300   \n",
      "5                                       70565   \n",
      "6                                       55564   \n",
      "7                                       74337   \n",
      "8                                       30848   \n",
      "9                                        3304   \n",
      "10                                      13522   \n",
      "11                                        321   \n",
      "12                                      25658   \n",
      "13                                      25137   \n",
      "14                                      32892   \n",
      "15                                       5515   \n",
      "16                                       9059   \n",
      "17                                       3507   \n",
      "18                                       6352   \n",
      "19                                       8424   \n",
      "20                                       8752   \n",
      "21                                       7929   \n",
      "22                                       7855   \n",
      "23                                       1912   \n",
      "24                                       3650   \n",
      "25                                        696   \n",
      "26                                        910   \n",
      "27                                       3718   \n",
      "28                                        374   \n",
      "29                                       3476   \n",
      "\n",
      "    reads_with_indexing_primer_mispriming  \n",
      "0                                    1059  \n",
      "1                                     967  \n",
      "2                                    1027  \n",
      "3                                     668  \n",
      "4                                     836  \n",
      "5                                    1774  \n",
      "6                                    1808  \n",
      "7                                    4217  \n",
      "8                                     959  \n",
      "9                                     110  \n",
      "10                                    502  \n",
      "11                                     64  \n",
      "12                                    760  \n",
      "13                                   1235  \n",
      "14                                   2351  \n",
      "15                                   1335  \n",
      "16                                   2223  \n",
      "17                                    868  \n",
      "18                                   1659  \n",
      "19                                   1657  \n",
      "20                                   2027  \n",
      "21                                   2642  \n",
      "22                                   3391  \n",
      "23                                    532  \n",
      "24                                   1110  \n",
      "25                                     59  \n",
      "26                                    462  \n",
      "27                                   1203  \n",
      "28                                    313  \n",
      "29                                   1487  \n"
     ]
    }
   ],
   "source": [
    "############### GOOD PRIMING Filter ##########\n",
    "#\n",
    "#    Here we assume the NNNNN is no longer on Read1 and I used 'trimmed_R1R2() function'\n",
    "#    gene specific primer on READ 1 in this case. The program only understnad checing primer 1***\n",
    "# The program \n",
    "\n",
    "# make a dataframe to capture all of the priming information and put it in the 'results' folder\n",
    "results_df_all = pd.DataFrame()\n",
    "\n",
    "results_folder = os.path.join(directory, 'results')\n",
    "if not os.path.exists(results_folder):\n",
    "    os.mkdir(results_folder)\n",
    "results_file = os.path.join(directory, 'results','all_priming.xlsx')\n",
    "    \n",
    "## inputs for correct_priming2() function\n",
    "mismatches =3                       # I am using the Levenshtein distance so if it is out of frame it costs 2\n",
    "trimmed_R1R2 = True                 #if the file has been already trimed\n",
    "removePrimerPlusDownstream = False  # remove the primer/downstream seq if it is good (good for guideseq)\n",
    "exportMismatch = True               # export the file of mismatches sequences\n",
    "\n",
    "for i in range(30):\n",
    "        \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "\n",
    "    \n",
    "    #EVERYTHING IS CAPITAL\n",
    "    #                       \n",
    "    #                    \n",
    "    ThreePrimeEnd_seq =       'CCTCGGAGCGTTGGATGATAGAGTCGATCCAGTTTACAAA'\n",
    "    #      Ch11:5226605:5226627                   \n",
    "    ThrePrimeEnd_primeronly = 'CCTCGGAGCGTTGGATGATAGAGTCGATCC'\n",
    "    \n",
    "    #3primer \n",
    "    FivePrimeEnd_seq =            'CTTCTGGGCAGGAACCGTGGGATCGCCAGCGT'\n",
    "    #      Ch11:5227054:5227084\n",
    "    FivePrimeEnd_seq_primeronly = 'CTTCTGGGCAGGAACCGTG'\n",
    "\n",
    "    direction = amplicon_info['Direction']\n",
    "\n",
    "    if direction == 3:\n",
    "        primer_seq_plus_downstream = ThreePrimeEnd_seq\n",
    "        primer_seq = ThrePrimeEnd_primeronly\n",
    "    elif direction == 5:\n",
    "        primer_seq_plus_downstream = FivePrimeEnd_seq\n",
    "        primer_seq = FivePrimeEnd_seq_primeronly\n",
    "    \n",
    "    df_sample_results = correct_priming2(directory, amplicon_info, primer_seq, primer_seq_plus_downstream, \n",
    "                                          mismatches, trimmed_R1R2, removePrimerPlusDownstream,\n",
    "                                          exportMismatch)\n",
    "    #add the results to the ongoing dataframe        \n",
    "    results_df_all = results_df_all.append(df_sample_results, ignore_index=True)\n",
    "    print('done with sample', i)\n",
    "\n",
    "#export the final table\n",
    "results_df_all.to_excel(results_file)    \n",
    "print(results_df_all)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def trim_short_fastq(dir_sample, amplicon_info, direction5primer, direction3primer, adapter):\n",
    "\n",
    "    \n",
    "    direction = amplicon_info['Direction']\n",
    "\n",
    "    # Read1  This is the gene specific primering binding\n",
    "    if direction == 5:\n",
    "        gene_specific_primer = direction5primer         # this is the gene specific primer for the 5 side\n",
    "    elif direction == 3:\n",
    "        gene_specific_primer = direction3primer         # gene specific primer for the 3 side\n",
    "    rev_gene_specific_primer = reverse_complement(gene_specific_primer)\n",
    "    \n",
    "    # LAM adapter\n",
    "    adapter_rev = reverse_complement(adapter)\n",
    "    \n",
    "    # We first check if the experiment had any guides\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "    \n",
    "    file_R1 = create_filename(dir_sample, N7, N5, 'R1fastq_CorrPrime')\n",
    "    file_R2 = create_filename(dir_sample, N7, N5, 'R2fastq_CorrPrime')\n",
    "\n",
    "    file_cutadapt_R1 = create_filename(dir_sample, N7, N5, 'R1trimmed')\n",
    "    file_cutadapt_R2 = create_filename(dir_sample, N7, N5, 'R2trimmed')\n",
    "    file_cutadapt_report = create_filename(dir_sample, N7, N5, 'trimmed_report')\n",
    "    \n",
    "    \n",
    "    if not os.path.exists(os.path.dirname(file_cutadapt_R1)):\n",
    "        os.mkdir(os.path.dirname(file_cutadapt_R1))\n",
    "    \n",
    "    \n",
    "    # remove adapters with cutadapt\n",
    "    #original uditas peramiter had an error -e 0.33 (but was cutting of random stuff too much)\n",
    "    # -a is hte 3' adapter for Read1\n",
    "    # -A is 3' adapter for Read2\n",
    "    # -m minium length\n",
    "    cutadapt_command = ['cutadapt',\n",
    "                        '-m', '50',\n",
    "                        '-e', '0.1',\n",
    "                        '-a', adapter_rev,\n",
    "                        '-A', rev_gene_specific_primer,\n",
    "                        '-o', file_cutadapt_R1, '-p', file_cutadapt_R2,\n",
    "                        file_R1, file_R2]\n",
    "\n",
    "    handle_cutadapt_report = open(file_cutadapt_report, 'wb')\n",
    "    subprocess.call(cutadapt_command, stdout=handle_cutadapt_report)\n",
    "    handle_cutadapt_report.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with sample 0\n",
      "done with sample 1\n",
      "done with sample 2\n",
      "done with sample 3\n",
      "done with sample 4\n",
      "done with sample 5\n",
      "done with sample 6\n",
      "done with sample 7\n",
      "done with sample 8\n",
      "done with sample 9\n",
      "done with sample 10\n",
      "done with sample 11\n",
      "done with sample 12\n",
      "done with sample 13\n",
      "done with sample 14\n",
      "done with sample 15\n",
      "done with sample 16\n",
      "done with sample 17\n",
      "done with sample 18\n",
      "done with sample 19\n",
      "done with sample 20\n",
      "done with sample 21\n",
      "done with sample 22\n",
      "done with sample 23\n",
      "done with sample 24\n",
      "done with sample 25\n",
      "done with sample 26\n",
      "done with sample 27\n",
      "done with sample 28\n",
      "done with sample 29\n"
     ]
    }
   ],
   "source": [
    "### TRIMMING ####\n",
    "#need to trim off the end of the short reads. This is for amplicons that were too short and have the other side on them.\n",
    "\n",
    "direction5primer = 'CTTCTGGGCAGGAACCGTG'\n",
    "direction3primer = 'CCTCGGAGCGTTGGATGATAGAGTCGATCC' \n",
    "adapter ='GACTATAGGGCACGCGTGG'\n",
    "\n",
    "for i in range(30):\n",
    "        \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    \n",
    "    trim_short_fastq(directory, amplicon_info, direction5primer, direction3primer, adapter)\n",
    "    \n",
    "    print('done with sample', i)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Need to make an new bowtie2 index file that includes targeting for alignment agains the whole genome so it is all in one sheet together. \n",
    "\n",
    "### Other option would be to align it to the amplicon, extract unaligned files and then align to the genome but seems cleaner this way. \n",
    "#### Ideally every sequence is unique between the targeting vector and genome.\n",
    "\n",
    "\n",
    "1. Build your fastas of interest and label .fa files.\n",
    "    1. You need fasta of hg38 or reference genome. You can pull this from downloaded bowtie indexed sampels and then use the following command to turn the index into a fasta file: bowtie2-inspect hg38 > hg38.fa   \n",
    "    2. Put all the fasta files in the same folder. Should also use the transfected plasmid\n",
    "2. index the files with bowtie\n",
    "    1. use the command bowtie2-build -f pE049,pe038_mc.fa,hg38.fa -p hg38_plus_targetvectorandplasmid\n",
    "    2. this has the -p to make it take less ram in my case.\n",
    "    3. In this case it adds the hg38 and the minicircle targeting file together\n",
    "    4. I have a Intel® Core™ i7-5500U CPU @ 2.40GHz × 4 with 15.1 GiB ram and it required about 13.8 gigs of ram and 2 hours to do a hg38+small fasta index\n",
    "    5. be sure to pay attention to the name of the new indexed file. \"hg38_plus_targetvector\" in example above. Add it to the sample_info.csv sheet. Under the tab 'genome_plus_targeting'.\n",
    "    6. you can check it indexed correcctly: bowtie2-inspect -s hg38_plus_targetvector\n",
    "    \n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Making the reference alingment sequences\n",
    "\n",
    "- plasmid sequence in .csv:  entire thing without HDR sequence (but with the ITRS) to map backbone integration\n",
    "- AAV: reference has entire AAV including Homology arms\n",
    "- HDR seq: The hdr reference sequence will be directly copied out ofthe .csv file and not altered.\n",
    "\n",
    "### Logic for following AAV integration\n",
    "1. Map locally to the plasmid without homology. This check for plasmid integration or ITR integration\n",
    "2. Extract out mappings. Things that don't map will be aligned to amplicons. Things that do map will be aligned to the AAV file.\n",
    "3. Align files that mapped to the plasmid to the complete AAV vector (they should contain ITR seq)\n",
    "4. Align files that did not map to the plasmid to the amplicons\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N7 : N701    N5 :  N501\n",
      "sample name:  1.1.F    sample name:  ELANE_AAV_Fwd\n",
      "rection type double_cut_same_chromosome_and_HDR\n"
     ]
    }
   ],
   "source": [
    "# single smaple for testing\n",
    "\n",
    "amplicon_info = get_csv_data(directory, 0)\n",
    "print('N7 :', amplicon_info['index_I1'], \"   N5 : \", amplicon_info['index_I2'])\n",
    "print (\"sample name: \", amplicon_info['name'], \"   sample name: \", amplicon_info['description'])\n",
    "print('rection type', get_reaction_type(amplicon_info))\n",
    "\n",
    "create_plasmid_reference(directory, amplicon_info)\n",
    "#create_AAV_reference(directory, amplicon_info)\n",
    "create_amplicon(directory, amplicon_info, file_genome_2bit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N7 : N701    N5 :  N501\n",
      "sample name:  1.1.F    sample name:  ELANE_AAV_Fwd\n",
      "N7 : N701    N5 :  N502\n",
      "sample name:  1.2.F    sample name:  ELANE_AAV_Fwd\n",
      "N7 : N701    N5 :  N504\n",
      "sample name:  1.3.F    sample name:  ELANE_AAV_Fwd\n",
      "N7 : N701    N5 :  N505\n",
      "sample name:  2.1.F    sample name:  ELANE_Cas9WT_Fwd\n",
      "N7 : N701    N5 :  N506\n",
      "sample name:  2.2.F    sample name:  ELANE_Cas9WT_Fwd\n",
      "N7 : N701    N5 :  N507\n",
      "sample name:  2.3.F    sample name:  ELANE_Cas9WT_Fwd\n",
      "N7 : N701    N5 :  N508\n",
      "sample name:  3.1.F    sample name:  ELANE_D10A_Fwd\n",
      "N7 : N701    N5 :  N510\n",
      "sample name:  3.2.F    sample name:  ELANE_D10A_Fwd\n",
      "N7 : N702    N5 :  N501\n",
      "sample name:  3.3.F    sample name:  ELANE_D10A_Fwd\n",
      "N7 : N702    N5 :  N502\n",
      "sample name:  12.1.F    sample name:  ELANE_Cas9WT_d3_Fwd\n",
      "N7 : N702    N5 :  N504\n",
      "sample name:  12.2.F    sample name:  ELANE_Cas9WT_d3_Fwd\n",
      "N7 : N702    N5 :  N505\n",
      "sample name:  12.3.F    sample name:  ELANE_Cas9WT_d3_Fwd\n",
      "N7 : N702    N5 :  N506\n",
      "sample name:  13.1.F    sample name:  ELANE_D10A_d3_Fwd\n",
      "N7 : N702    N5 :  N507\n",
      "sample name:  13.2.F    sample name:  ELANE_D10A_d3_Fwd\n",
      "N7 : N702    N5 :  N508\n",
      "sample name:  13.3.F    sample name:  ELANE_D10A_d3_Fwd\n",
      "N7 : N703    N5 :  N501\n",
      "sample name:  1.1.R    sample name:  ELANE_AAV_Rev\n",
      "N7 : N703    N5 :  N502\n",
      "sample name:  1.2.R    sample name:  ELANE_AAV_Rev\n",
      "N7 : N703    N5 :  N504\n",
      "sample name:  1.3.R    sample name:  ELANE_AAV_Rev\n",
      "N7 : N703    N5 :  N505\n",
      "sample name:  2.1.R    sample name:  ELANE_Cas9WT_Rev\n",
      "N7 : N703    N5 :  N506\n",
      "sample name:  2.2.R    sample name:  ELANE_Cas9WT_Rev\n",
      "N7 : N703    N5 :  N507\n",
      "sample name:  2.3.R    sample name:  ELANE_Cas9WT_Rev\n",
      "N7 : N703    N5 :  N508\n",
      "sample name:  3.1.R    sample name:  ELANE_D10A_Rev\n",
      "N7 : N703    N5 :  N510\n",
      "sample name:  3.2.R    sample name:  ELANE_D10A_Rev\n",
      "N7 : N704    N5 :  N501\n",
      "sample name:  3.3.R    sample name:  ELANE_D10A_Rev\n",
      "N7 : N704    N5 :  N502\n",
      "sample name:  12.1.R    sample name:  ELANE_Cas9WT_d3_Rev\n",
      "N7 : N704    N5 :  N504\n",
      "sample name:  12.2.R    sample name:  ELANE_Cas9WT_d3_Rev\n",
      "N7 : N704    N5 :  N505\n",
      "sample name:  12.3.R    sample name:  ELANE_Cas9WT_d3_Rev\n",
      "N7 : N704    N5 :  N506\n",
      "sample name:  13.1.R    sample name:  ELANE_D10A_d3_Rev\n",
      "N7 : N704    N5 :  N507\n",
      "sample name:  13.2.R    sample name:  ELANE_D10A_d3_Rev\n",
      "N7 : N704    N5 :  N508\n",
      "sample name:  13.3.R    sample name:  ELANE_D10A_d3_Rev\n"
     ]
    }
   ],
   "source": [
    "######## GENERATING REFERENCE SEQUENCES ###########\n",
    "#\n",
    "# MAKE SURE PLASMID DOESNT HAVE THE HDR PORTION IN IT\n",
    "# MAKE SURE AAV is entire AAV sequence\n",
    "# HDR should be entire sequence for reference knock-in\n",
    "\n",
    "\n",
    "for i in range(30):\n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    print('N7 :', amplicon_info['index_I1'], \"   N5 : \", amplicon_info['index_I2'])\n",
    "    print (\"sample name: \", amplicon_info['name'], \"   sample name: \", amplicon_info['description'])\n",
    "    get_reaction_type(amplicon_info)\n",
    "\n",
    "    create_plasmid_reference(directory, amplicon_info)\n",
    "    create_AAV_reference(directory, amplicon_info)\n",
    "    create_amplicon(directory, amplicon_info, file_genome_2bit)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Aligning reads to plasmid/AAV (local alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N7 : N701    N5 :  N501\n",
      "sample name:  1.1.F    sample name:  ELANE_AAV_Fwd\n",
      "made filenames\n",
      "N7 : N701    N5 :  N502\n",
      "sample name:  1.2.F    sample name:  ELANE_AAV_Fwd\n",
      "made filenames\n",
      "N7 : N701    N5 :  N504\n",
      "sample name:  1.3.F    sample name:  ELANE_AAV_Fwd\n",
      "made filenames\n",
      "N7 : N701    N5 :  N505\n",
      "sample name:  2.1.F    sample name:  ELANE_Cas9WT_Fwd\n",
      "made filenames\n",
      "N7 : N701    N5 :  N506\n",
      "sample name:  2.2.F    sample name:  ELANE_Cas9WT_Fwd\n",
      "made filenames\n",
      "N7 : N701    N5 :  N507\n",
      "sample name:  2.3.F    sample name:  ELANE_Cas9WT_Fwd\n",
      "made filenames\n",
      "N7 : N701    N5 :  N508\n",
      "sample name:  3.1.F    sample name:  ELANE_D10A_Fwd\n",
      "made filenames\n",
      "N7 : N701    N5 :  N510\n",
      "sample name:  3.2.F    sample name:  ELANE_D10A_Fwd\n",
      "made filenames\n",
      "N7 : N702    N5 :  N501\n",
      "sample name:  3.3.F    sample name:  ELANE_D10A_Fwd\n",
      "made filenames\n",
      "N7 : N702    N5 :  N502\n",
      "sample name:  12.1.F    sample name:  ELANE_Cas9WT_d3_Fwd\n",
      "made filenames\n",
      "N7 : N702    N5 :  N504\n",
      "sample name:  12.2.F    sample name:  ELANE_Cas9WT_d3_Fwd\n",
      "made filenames\n",
      "N7 : N702    N5 :  N505\n",
      "sample name:  12.3.F    sample name:  ELANE_Cas9WT_d3_Fwd\n",
      "made filenames\n",
      "N7 : N702    N5 :  N506\n",
      "sample name:  13.1.F    sample name:  ELANE_D10A_d3_Fwd\n",
      "made filenames\n",
      "N7 : N702    N5 :  N507\n",
      "sample name:  13.2.F    sample name:  ELANE_D10A_d3_Fwd\n",
      "made filenames\n",
      "N7 : N702    N5 :  N508\n",
      "sample name:  13.3.F    sample name:  ELANE_D10A_d3_Fwd\n",
      "made filenames\n",
      "N7 : N703    N5 :  N501\n",
      "sample name:  1.1.R    sample name:  ELANE_AAV_Rev\n",
      "made filenames\n",
      "N7 : N703    N5 :  N502\n",
      "sample name:  1.2.R    sample name:  ELANE_AAV_Rev\n",
      "made filenames\n",
      "N7 : N703    N5 :  N504\n",
      "sample name:  1.3.R    sample name:  ELANE_AAV_Rev\n",
      "made filenames\n",
      "N7 : N703    N5 :  N505\n",
      "sample name:  2.1.R    sample name:  ELANE_Cas9WT_Rev\n",
      "made filenames\n",
      "N7 : N703    N5 :  N506\n",
      "sample name:  2.2.R    sample name:  ELANE_Cas9WT_Rev\n",
      "made filenames\n",
      "N7 : N703    N5 :  N507\n",
      "sample name:  2.3.R    sample name:  ELANE_Cas9WT_Rev\n",
      "made filenames\n",
      "N7 : N703    N5 :  N508\n",
      "sample name:  3.1.R    sample name:  ELANE_D10A_Rev\n",
      "made filenames\n",
      "N7 : N703    N5 :  N510\n",
      "sample name:  3.2.R    sample name:  ELANE_D10A_Rev\n",
      "made filenames\n",
      "N7 : N704    N5 :  N501\n",
      "sample name:  3.3.R    sample name:  ELANE_D10A_Rev\n",
      "made filenames\n",
      "N7 : N704    N5 :  N502\n",
      "sample name:  12.1.R    sample name:  ELANE_Cas9WT_d3_Rev\n",
      "made filenames\n",
      "N7 : N704    N5 :  N504\n",
      "sample name:  12.2.R    sample name:  ELANE_Cas9WT_d3_Rev\n",
      "made filenames\n",
      "N7 : N704    N5 :  N505\n",
      "sample name:  12.3.R    sample name:  ELANE_Cas9WT_d3_Rev\n",
      "made filenames\n",
      "N7 : N704    N5 :  N506\n",
      "sample name:  13.1.R    sample name:  ELANE_D10A_d3_Rev\n",
      "made filenames\n",
      "N7 : N704    N5 :  N507\n",
      "sample name:  13.2.R    sample name:  ELANE_D10A_d3_Rev\n",
      "made filenames\n",
      "N7 : N704    N5 :  N508\n",
      "sample name:  13.3.R    sample name:  ELANE_D10A_d3_Rev\n",
      "made filenames\n"
     ]
    }
   ],
   "source": [
    "######## ALIGNING LOCAL TO PLASMID FILES ###########\n",
    "#\n",
    "# MAKE SURE PLASMID DOESNT HAVE THE AAV PORTION IN IT\n",
    "# MAKE SURE AAV DOESN'T HAVE HDR ARMS IN IT\n",
    "\n",
    "\n",
    "for i in range(30):\n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    print('N7 :', amplicon_info['index_I1'], \"   N5 : \", amplicon_info['index_I2'])\n",
    "    print (\"sample name: \", amplicon_info['name'], \"   sample name: \", amplicon_info['description'])\n",
    "    \n",
    "    align_plasmid_local(directory, amplicon_info, ncpu=12)\n",
    "    extract_reads_plasmid(directory, amplicon_info)\n",
    "    align_AAV_local(directory, amplicon_info, ncpu=12)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with sample 0\n",
      "done with sample 1\n",
      "done with sample 2\n",
      "done with sample 3\n",
      "done with sample 4\n",
      "done with sample 5\n",
      "done with sample 6\n",
      "done with sample 7\n",
      "done with sample 8\n",
      "done with sample 9\n",
      "done with sample 10\n",
      "done with sample 11\n",
      "done with sample 12\n",
      "done with sample 13\n",
      "done with sample 14\n",
      "done with sample 15\n",
      "done with sample 16\n",
      "done with sample 17\n",
      "done with sample 18\n",
      "done with sample 19\n",
      "done with sample 20\n",
      "done with sample 21\n",
      "done with sample 22\n",
      "done with sample 23\n",
      "done with sample 24\n",
      "done with sample 25\n",
      "done with sample 26\n",
      "done with sample 27\n",
      "done with sample 28\n",
      "done with sample 29\n",
      "   sample_name    i7    i5  number_plasmid_alignments\n",
      "0        1.1.F  N701  N501                          0\n",
      "1        1.2.F  N701  N502                          0\n",
      "2        1.3.F  N701  N504                          0\n",
      "3        2.1.F  N701  N505                          8\n",
      "4        2.2.F  N701  N506                         21\n",
      "5        2.3.F  N701  N507                         28\n",
      "6        3.1.F  N701  N508                         17\n",
      "7        3.2.F  N701  N510                         11\n",
      "8        3.3.F  N702  N501                          0\n",
      "9       12.1.F  N702  N502                          0\n",
      "10      12.2.F  N702  N504                          0\n",
      "11      12.3.F  N702  N505                          0\n",
      "12      13.1.F  N702  N506                          0\n",
      "13      13.2.F  N702  N507                          0\n",
      "14      13.3.F  N702  N508                          0\n",
      "15       1.1.R  N703  N501                          0\n",
      "16       1.2.R  N703  N502                          0\n",
      "17       1.3.R  N703  N504                          0\n",
      "18       2.1.R  N703  N505                          0\n",
      "19       2.2.R  N703  N506                          7\n",
      "20       2.3.R  N703  N507                          3\n",
      "21       3.1.R  N703  N508                         28\n",
      "22       3.2.R  N703  N510                         15\n",
      "23       3.3.R  N704  N501                         10\n",
      "24      12.1.R  N704  N502                        399\n",
      "25      12.2.R  N704  N504                         10\n",
      "26      12.3.R  N704  N505                        158\n",
      "27      13.1.R  N704  N506                          0\n",
      "28      13.2.R  N704  N507                          0\n",
      "29      13.3.R  N704  N508                          0\n"
     ]
    }
   ],
   "source": [
    "######## QUANTIFYING PLASMID/AAV INTEGRATION ###########\n",
    "#\n",
    "\n",
    "\n",
    "\n",
    "#SUMMARY FILE\n",
    "results_df_all = pd.DataFrame()\n",
    "\n",
    "results_folder = os.path.join(directory, 'results')\n",
    "if not os.path.exists(results_folder):\n",
    "    os.mkdir(results_folder)\n",
    "results_file = os.path.join(directory, 'results','plasmid_AAV_integration.xlsx')\n",
    "\n",
    "\n",
    "# FUNCTION INPUTS\n",
    "analysis = 'plasmid'\n",
    "min_MAPQ = 1\n",
    "\n",
    "for i in range(30):\n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i) #cas9 cutting sample\n",
    "\n",
    "    df_sample_results = analyze_local_alignments(directory, amplicon_info, min_MAPQ, analysis)\n",
    "\n",
    "    results_df_all = results_df_all.append(df_sample_results, ignore_index=True)\n",
    "    print('done with sample', i)\n",
    "\n",
    "#export the final table\n",
    "results_df_all.to_excel(results_file)   \n",
    "\n",
    "print(results_df_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align filtered reads to amplicons (end-to-end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "############################\n",
    "#\n",
    "# Aligns reads globally to amplicon. \"end-to-end\" as the default function in bowtie2\n",
    "# Input: directory to be analyzed\n",
    "#        amplicon_info, slice of sample_info.csv for the sample being processed\n",
    "#        file_genome_2bit, 2bit file with the reference genome being used\n",
    "#\n",
    "#        paired = True or False\n",
    "# ##########################\n",
    "def align_ampliconLAM(dir_sample, amplicon_info, check_plasmid_insertions, paired, ncpu=4):\n",
    "\n",
    "    # We first check if the experiment had any guides\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "\n",
    "    has_plasmid = type(amplicon_info['plasmid_sequence']) is str or type(amplicon_info['plasmid_sequence']) is unicode\n",
    "\n",
    "    if check_plasmid_insertions == 1 and has_plasmid:\n",
    "        file_R1 = create_filename(dir_sample, N7, N5, 'unmapped_plasmid_R1fastqgz')\n",
    "        file_R2 = create_filename(dir_sample, N7, N5, 'unmapped_plasmid_R2fastqgz')\n",
    "    else:\n",
    "        file_R1 = create_filename(dir_sample, N7, N5, 'R1trimmed')\n",
    "        file_R2 = create_filename(dir_sample, N7, N5, 'R2trimmed')\n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_R1)):\n",
    "        os.mkdir(os.path.dirname(file_R1))\n",
    "\n",
    "    file_sam_amplicons = create_filename(dir_sample, N7, N5, 'sam_amplicons')\n",
    "    file_sam_report_amplicons = create_filename(dir_sample, N7, N5, 'sam_report_amplicons')\n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_sam_amplicons)):\n",
    "        os.mkdir(os.path.dirname(file_sam_amplicons))\n",
    "\n",
    "    file_bam_amplicons = create_filename(dir_sample, N7, N5, 'bam_amplicons')\n",
    "    file_sorted_bam_amplicons = create_filename(dir_sample, N7, N5, 'sorted_bam_amplicons')\n",
    "\n",
    "    if not os.path.exists(os.path.dirname(file_bam_amplicons)):\n",
    "        os.mkdir(os.path.dirname(file_bam_amplicons))\n",
    "\n",
    "    # global alignment to the amplicons with bowtie2\n",
    "    initial_dir = os.getcwd()\n",
    "    folder_amplicons = create_filename(dir_sample, N7, N5, 'amplicons')\n",
    "\n",
    "    os.chdir(folder_amplicons)\n",
    "    \n",
    "    if paired == True:\n",
    "        bowtie2_command = ['bowtie2', '-p', str(ncpu), '--very-sensitive', \n",
    "                           '-X', '5000', '-k', '3', '-x', 'amplicons',\n",
    "                           '-1', file_R1, '-2', file_R2,\n",
    "                           '-S', file_sam_amplicons]\n",
    "    \n",
    "    else:\n",
    "        bowtie2_command = ['bowtie2', '--very-sensitive', '-p', str(ncpu),\n",
    "                   '-X', '5000', '-k', '3', '-x', 'amplicons',\n",
    "                   '-U', file_R1, '-S', file_sam_amplicons]\n",
    "\n",
    "\n",
    "    handle_sam_report_amplicons = open(file_sam_report_amplicons, 'wb')\n",
    "\n",
    "    subprocess.call(bowtie2_command, stderr=handle_sam_report_amplicons)\n",
    "\n",
    "    handle_sam_report_amplicons.close()\n",
    "\n",
    "    # convert sam to bam\n",
    "    sam_to_bam_amplicons_command = ['samtools', 'view', '-Sb', file_sam_amplicons]\n",
    "\n",
    "    handle_file_bam_amplicons = open(file_bam_amplicons, 'wb')\n",
    "\n",
    "    subprocess.call(sam_to_bam_amplicons_command, stdout=handle_file_bam_amplicons)\n",
    "\n",
    "    # sort bam files\n",
    "    sort_bam_amplicons_command = ['samtools', 'sort', file_bam_amplicons, '-o', file_sorted_bam_amplicons]\n",
    "\n",
    "    subprocess.call(sort_bam_amplicons_command)\n",
    "\n",
    "    # Clean up\n",
    "    #os.remove(file_sam_amplicons)\n",
    "    os.remove(file_bam_amplicons)\n",
    "\n",
    "    # Create bam index files\n",
    "    create_bam_amplicons_index_command = ['samtools', 'index', file_sorted_bam_amplicons]\n",
    "    subprocess.call(create_bam_amplicons_index_command)\n",
    "\n",
    "    os.chdir(initial_dir)\n",
    " \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/eric/Data/Spaced_Nicking/LAM2_Miniseq_ELANE\n"
     ]
    }
   ],
   "source": [
    "######## Aligning reads to amplicons ###########\n",
    "#\n",
    "# \n",
    "#  This will go thorugh and align the amplicons. \n",
    "# check_plamid_insertions is if the aligned reads should come from plasmid integration removed fastq\n",
    "# paired, do you use paired ends for alignment or just Read1 (gene specific primer)\n",
    "print(directory)\n",
    "\n",
    "for i in range(30):#was 30\n",
    "    \n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    align_ampliconLAM(directory, amplicon_info, check_plasmid_insertions = True, paired = True, ncpu = 14)\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'extract_unmapped_reads_amplicons_LAM' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-c8b80f209195>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m     \u001b[0mamplicon_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_csv_data\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdirectory\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 10\u001b[0;31m     \u001b[0mextract_unmapped_reads_amplicons_LAM\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdirectory\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mamplicon_info\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'paired'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'extract_unmapped_reads_amplicons_LAM' is not defined"
     ]
    }
   ],
   "source": [
    "################################################################################\n",
    "# Function to extract all unmapped reads to the amplicons\n",
    "# reads, 'single', 'paired'\n",
    "################################################################################\n",
    "#  THIS IS USED FOR PIPELINE 2 WHERE WE ALIGN THEM TO THE REST OF HTE GENOME\n",
    "\n",
    "for i in range(30):\n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i)    \n",
    "    extract_unmapped_reads_amplicons_LAM(directory, amplicon_info, 'paired')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyze the aligned amplicon reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Allreads is :  62208\n",
      "HDRcount is :  1978\n",
      "done with sample 0 ELANE_AAV_Fwd N701 N501\n",
      "Allreads is :  32798\n",
      "HDRcount is :  814\n",
      "done with sample 1 ELANE_AAV_Fwd N701 N502\n",
      "Allreads is :  17949\n",
      "HDRcount is :  329\n",
      "done with sample 2 ELANE_AAV_Fwd N701 N504\n",
      "Allreads is :  14974\n",
      "HDRcount is :  8405\n",
      "done with sample 3 ELANE_Cas9WT_Fwd N701 N505\n",
      "Allreads is :  23336\n",
      "HDRcount is :  13179\n",
      "done with sample 4 ELANE_Cas9WT_Fwd N701 N506\n",
      "Allreads is :  45874\n",
      "HDRcount is :  26885\n",
      "done with sample 5 ELANE_Cas9WT_Fwd N701 N507\n",
      "Allreads is :  74285\n",
      "HDRcount is :  27734\n",
      "done with sample 6 ELANE_D10A_Fwd N701 N508\n",
      "Allreads is :  108044\n",
      "HDRcount is :  36474\n",
      "done with sample 7 ELANE_D10A_Fwd N701 N510\n",
      "Allreads is :  89362\n",
      "HDRcount is :  34451\n",
      "done with sample 8 ELANE_D10A_Fwd N702 N501\n",
      "Allreads is :  5269\n",
      "HDRcount is :  317\n",
      "done with sample 9 ELANE_Cas9WT_d3_Fwd N702 N502\n",
      "Allreads is :  23201\n",
      "HDRcount is :  21935\n",
      "done with sample 10 ELANE_Cas9WT_d3_Fwd N702 N504\n",
      "Allreads is :  11389\n",
      "HDRcount is :  10877\n",
      "done with sample 11 ELANE_Cas9WT_d3_Fwd N702 N505\n",
      "Allreads is :  113938\n",
      "HDRcount is :  106492\n",
      "done with sample 12 ELANE_D10A_d3_Fwd N702 N506\n",
      "Allreads is :  172671\n",
      "HDRcount is :  161035\n",
      "done with sample 13 ELANE_D10A_d3_Fwd N702 N507\n",
      "Allreads is :  324959\n",
      "HDRcount is :  303888\n",
      "done with sample 14 ELANE_D10A_d3_Fwd N702 N508\n",
      "Allreads is :  32867\n",
      "HDRcount is :  27\n",
      "done with sample 15 ELANE_AAV_Rev N703 N501\n",
      "Allreads is :  51694\n",
      "HDRcount is :  40\n",
      "done with sample 16 ELANE_AAV_Rev N703 N502\n",
      "Allreads is :  11350\n",
      "HDRcount is :  7\n",
      "done with sample 17 ELANE_AAV_Rev N703 N504\n",
      "Allreads is :  35971\n",
      "HDRcount is :  20007\n",
      "done with sample 18 ELANE_Cas9WT_Rev N703 N505\n",
      "Allreads is :  38779\n",
      "HDRcount is :  21765\n",
      "done with sample 19 ELANE_Cas9WT_Rev N703 N506\n",
      "Allreads is :  42921\n",
      "HDRcount is :  20822\n",
      "done with sample 20 ELANE_Cas9WT_Rev N703 N507\n",
      "Allreads is :  66670\n",
      "HDRcount is :  18172\n",
      "done with sample 21 ELANE_D10A_Rev N703 N508\n",
      "Allreads is :  58892\n",
      "HDRcount is :  29107\n",
      "done with sample 22 ELANE_D10A_Rev N703 N510\n",
      "Allreads is :  10871\n",
      "HDRcount is :  3094\n",
      "done with sample 23 ELANE_D10A_Rev N704 N501\n",
      "Allreads is :  33593\n",
      "HDRcount is :  5795\n",
      "done with sample 24 ELANE_Cas9WT_d3_Rev N704 N502\n",
      "Allreads is :  571\n",
      "HDRcount is :  56\n",
      "done with sample 25 ELANE_Cas9WT_d3_Rev N704 N504\n",
      "Allreads is :  18295\n",
      "HDRcount is :  2775\n",
      "done with sample 26 ELANE_Cas9WT_d3_Rev N704 N505\n",
      "Allreads is :  47298\n",
      "HDRcount is :  4359\n",
      "done with sample 27 ELANE_D10A_d3_Rev N704 N506\n",
      "Allreads is :  8809\n",
      "HDRcount is :  1531\n",
      "done with sample 28 ELANE_D10A_d3_Rev N704 N507\n",
      "Allreads is :  51456\n",
      "HDRcount is :  4854\n",
      "done with sample 29 ELANE_D10A_d3_Rev N704 N508\n",
      "   sample_name    i7    i5      sample_descript  Total_mapped_reads  \\\n",
      "0        1.1.F  N701  N501        ELANE_AAV_Fwd               62208   \n",
      "1        1.2.F  N701  N502        ELANE_AAV_Fwd               32798   \n",
      "2        1.3.F  N701  N504        ELANE_AAV_Fwd               17949   \n",
      "3        2.1.F  N701  N505     ELANE_Cas9WT_Fwd               14974   \n",
      "4        2.2.F  N701  N506     ELANE_Cas9WT_Fwd               23336   \n",
      "5        2.3.F  N701  N507     ELANE_Cas9WT_Fwd               45874   \n",
      "6        3.1.F  N701  N508       ELANE_D10A_Fwd               74285   \n",
      "7        3.2.F  N701  N510       ELANE_D10A_Fwd              108044   \n",
      "8        3.3.F  N702  N501       ELANE_D10A_Fwd               89362   \n",
      "9       12.1.F  N702  N502  ELANE_Cas9WT_d3_Fwd                5269   \n",
      "10      12.2.F  N702  N504  ELANE_Cas9WT_d3_Fwd               23201   \n",
      "11      12.3.F  N702  N505  ELANE_Cas9WT_d3_Fwd               11389   \n",
      "12      13.1.F  N702  N506    ELANE_D10A_d3_Fwd              113938   \n",
      "13      13.2.F  N702  N507    ELANE_D10A_d3_Fwd              172671   \n",
      "14      13.3.F  N702  N508    ELANE_D10A_d3_Fwd              324959   \n",
      "15       1.1.R  N703  N501        ELANE_AAV_Rev               32867   \n",
      "16       1.2.R  N703  N502        ELANE_AAV_Rev               51694   \n",
      "17       1.3.R  N703  N504        ELANE_AAV_Rev               11350   \n",
      "18       2.1.R  N703  N505     ELANE_Cas9WT_Rev               35971   \n",
      "19       2.2.R  N703  N506     ELANE_Cas9WT_Rev               38779   \n",
      "20       2.3.R  N703  N507     ELANE_Cas9WT_Rev               42921   \n",
      "21       3.1.R  N703  N508       ELANE_D10A_Rev               66670   \n",
      "22       3.2.R  N703  N510       ELANE_D10A_Rev               58892   \n",
      "23       3.3.R  N704  N501       ELANE_D10A_Rev               10871   \n",
      "24      12.1.R  N704  N502  ELANE_Cas9WT_d3_Rev               33593   \n",
      "25      12.2.R  N704  N504  ELANE_Cas9WT_d3_Rev                 571   \n",
      "26      12.3.R  N704  N505  ELANE_Cas9WT_d3_Rev               18295   \n",
      "27      13.1.R  N704  N506    ELANE_D10A_d3_Rev               47298   \n",
      "28      13.2.R  N704  N507    ELANE_D10A_d3_Rev                8809   \n",
      "29      13.3.R  N704  N508    ELANE_D10A_d3_Rev               51456   \n",
      "\n",
      "    wt_cut1_total_reads  wt_cut1_total_indels  wt_cut1_total_deletions  \\\n",
      "0                 30028                     6                        5   \n",
      "1                 16080                     7                        7   \n",
      "2                  8799                     3                        3   \n",
      "3                  3251                  1551                     1253   \n",
      "4                  5110                  2539                     2228   \n",
      "5                  9565                  4917                     3870   \n",
      "6                 23572                   502                      404   \n",
      "7                 36629                   547                      441   \n",
      "8                 27807                   457                      417   \n",
      "9                  2703                     1                        1   \n",
      "10                  610                    11                       10   \n",
      "11                  254                     5                        4   \n",
      "12                 3704                     3                        2   \n",
      "13                 5952                     4                        3   \n",
      "14                10744                     6                        4   \n",
      "15                12769                     3                        2   \n",
      "16                19515                     4                        0   \n",
      "17                 4511                     3                        3   \n",
      "18                 6066                  2734                     2239   \n",
      "19                 6028                  3118                     2740   \n",
      "20                 8406                  3567                     2993   \n",
      "21                15962                   334                      281   \n",
      "22                 9761                   242                      155   \n",
      "23                 2549                    53                       44   \n",
      "24                 7013                   125                       85   \n",
      "25                  100                     0                        0   \n",
      "26                 4504                   117                       98   \n",
      "27                15609                     4                        2   \n",
      "28                 2769                     0                        0   \n",
      "29                16835                     5                        3   \n",
      "\n",
      "    wt_cut1_total_insertions  wt_cut2_total_reads  ...  \\\n",
      "0                          1                 3418  ...   \n",
      "1                          0                 1457  ...   \n",
      "2                          0                  293  ...   \n",
      "3                        301                  180  ...   \n",
      "4                        323                  268  ...   \n",
      "5                       1053                  741  ...   \n",
      "6                        117                 2269  ...   \n",
      "7                        106                 5177  ...   \n",
      "8                         40                 3877  ...   \n",
      "9                          0                    0  ...   \n",
      "10                         1                   24  ...   \n",
      "11                         1                   25  ...   \n",
      "12                         1                  138  ...   \n",
      "13                         1                  510  ...   \n",
      "14                         2                  910  ...   \n",
      "15                         1                16346  ...   \n",
      "16                         4                25659  ...   \n",
      "17                         0                 5602  ...   \n",
      "18                       505                 7763  ...   \n",
      "19                       389                 7968  ...   \n",
      "20                       585                10717  ...   \n",
      "21                        60                19098  ...   \n",
      "22                        93                11744  ...   \n",
      "23                         9                 2975  ...   \n",
      "24                        40                 8434  ...   \n",
      "25                         0                  117  ...   \n",
      "26                        19                 4943  ...   \n",
      "27                         2                20836  ...   \n",
      "28                         0                 3553  ...   \n",
      "29                         2                22682  ...   \n",
      "\n",
      "    1a_1a_cut1_total_reads  1a_1a_cut1_total_indels  \\\n",
      "0                        0                        0   \n",
      "1                        0                        0   \n",
      "2                        0                        0   \n",
      "3                        0                        0   \n",
      "4                        0                        0   \n",
      "5                        0                        0   \n",
      "6                        0                        0   \n",
      "7                        0                        0   \n",
      "8                        0                        0   \n",
      "9                        0                        0   \n",
      "10                       0                        0   \n",
      "11                       0                        0   \n",
      "12                       0                        0   \n",
      "13                       0                        0   \n",
      "14                       0                        0   \n",
      "15                       0                        0   \n",
      "16                       0                        0   \n",
      "17                       0                        0   \n",
      "18                       0                        0   \n",
      "19                       0                        0   \n",
      "20                       0                        0   \n",
      "21                       0                        0   \n",
      "22                       0                        0   \n",
      "23                       0                        0   \n",
      "24                       0                        0   \n",
      "25                       0                        0   \n",
      "26                       0                        0   \n",
      "27                       0                        0   \n",
      "28                       0                        0   \n",
      "29                       0                        0   \n",
      "\n",
      "    1a_1a_cut1_total_deletions  1a_1a_cut1_total_insertions  \\\n",
      "0                            0                            0   \n",
      "1                            0                            0   \n",
      "2                            0                            0   \n",
      "3                            0                            0   \n",
      "4                            0                            0   \n",
      "5                            0                            0   \n",
      "6                            0                            0   \n",
      "7                            0                            0   \n",
      "8                            0                            0   \n",
      "9                            0                            0   \n",
      "10                           0                            0   \n",
      "11                           0                            0   \n",
      "12                           0                            0   \n",
      "13                           0                            0   \n",
      "14                           0                            0   \n",
      "15                           0                            0   \n",
      "16                           0                            0   \n",
      "17                           0                            0   \n",
      "18                           0                            0   \n",
      "19                           0                            0   \n",
      "20                           0                            0   \n",
      "21                           0                            0   \n",
      "22                           0                            0   \n",
      "23                           0                            0   \n",
      "24                           0                            0   \n",
      "25                           0                            0   \n",
      "26                           0                            0   \n",
      "27                           0                            0   \n",
      "28                           0                            0   \n",
      "29                           0                            0   \n",
      "\n",
      "    2b_2b_cut1_total_reads  2b_2b_cut1_total_indels  \\\n",
      "0                        0                        0   \n",
      "1                        0                        0   \n",
      "2                        0                        0   \n",
      "3                        0                        0   \n",
      "4                        0                        0   \n",
      "5                        0                        0   \n",
      "6                        0                        0   \n",
      "7                        0                        0   \n",
      "8                        0                        0   \n",
      "9                        0                        0   \n",
      "10                       0                        0   \n",
      "11                       0                        0   \n",
      "12                       0                        0   \n",
      "13                       0                        0   \n",
      "14                       0                        0   \n",
      "15                       0                        0   \n",
      "16                       0                        0   \n",
      "17                       0                        0   \n",
      "18                       0                        0   \n",
      "19                       0                        0   \n",
      "20                       0                        0   \n",
      "21                       0                        0   \n",
      "22                       0                        0   \n",
      "23                       0                        0   \n",
      "24                       0                        0   \n",
      "25                       0                        0   \n",
      "26                       0                        0   \n",
      "27                       0                        0   \n",
      "28                       0                        0   \n",
      "29                       0                        0   \n",
      "\n",
      "    2b_2b_cut1_total_deletions  2b_2b_cut1_total_insertions  HDR_total_reads  \\\n",
      "0                            0                            0             1978   \n",
      "1                            0                            0              814   \n",
      "2                            0                            0              329   \n",
      "3                            0                            0             8405   \n",
      "4                            0                            0            13179   \n",
      "5                            0                            0            26885   \n",
      "6                            0                            0            27734   \n",
      "7                            0                            0            36474   \n",
      "8                            0                            0            34451   \n",
      "9                            0                            0              317   \n",
      "10                           0                            0            21935   \n",
      "11                           0                            0            10877   \n",
      "12                           0                            0           106492   \n",
      "13                           0                            0           161035   \n",
      "14                           0                            0           303888   \n",
      "15                           0                            0               27   \n",
      "16                           0                            0               40   \n",
      "17                           0                            0                7   \n",
      "18                           0                            0            20007   \n",
      "19                           0                            0            21765   \n",
      "20                           0                            0            20822   \n",
      "21                           0                            0            18172   \n",
      "22                           0                            0            29107   \n",
      "23                           0                            0             3094   \n",
      "24                           0                            0             5795   \n",
      "25                           0                            0               56   \n",
      "26                           0                            0             2775   \n",
      "27                           0                            0             4359   \n",
      "28                           0                            0             1531   \n",
      "29                           0                            0             4854   \n",
      "\n",
      "    median_fragment_size  \n",
      "0                  261.0  \n",
      "1                  260.0  \n",
      "2                  214.0  \n",
      "3                  224.0  \n",
      "4                  224.0  \n",
      "5                  250.0  \n",
      "6                  265.0  \n",
      "7                  280.0  \n",
      "8                  278.0  \n",
      "9                  589.0  \n",
      "10                 240.0  \n",
      "11                 282.0  \n",
      "12                 233.0  \n",
      "13                 259.0  \n",
      "14                 252.0  \n",
      "15                 418.0  \n",
      "16                 416.0  \n",
      "17                 404.0  \n",
      "18                 419.0  \n",
      "19                 422.0  \n",
      "20                 417.0  \n",
      "21                 415.0  \n",
      "22                 421.0  \n",
      "23                 405.0  \n",
      "24                 406.0  \n",
      "25                 383.0  \n",
      "26                 403.0  \n",
      "27                 407.0  \n",
      "28                 404.0  \n",
      "29                 406.0  \n",
      "\n",
      "[30 rows x 35 columns]\n"
     ]
    }
   ],
   "source": [
    "################### ANALYZING THE ALIGNMENTS ###############\n",
    "#\n",
    "#          INDELS AND QUANTIFICATION\n",
    "\n",
    "\n",
    "#SUMMARY FILE\n",
    "results_df_all = pd.DataFrame()\n",
    "\n",
    "results_folder = os.path.join(directory, 'results')\n",
    "if not os.path.exists(results_folder):\n",
    "    os.mkdir(results_folder)\n",
    "results_file = os.path.join(directory, 'results','all_amplicon_counts.xlsx')\n",
    "\n",
    "# function inputs\n",
    "window_size = 15\n",
    "amplicon_window_around_cut = 1000\n",
    "min_MAPQ = 5\n",
    "min_AS = -180\n",
    "\n",
    "for i in range(30):\n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i)    \n",
    "\n",
    "    result_amplicon_df = analyze_alignments_LAM(directory, amplicon_info, window_size,\n",
    "                                                    amplicon_window_around_cut, min_MAPQ, min_AS)\n",
    "    \n",
    "    \n",
    "    results_df_all = results_df_all.append(result_amplicon_df, ignore_index=True)\n",
    "    print('done with sample', i, amplicon_info['description'], amplicon_info['index_I1'],\n",
    "                                                                  amplicon_info['index_I2'])\n",
    "\n",
    "results_df_all.to_excel(results_file)   \n",
    "\n",
    "print(results_df_all)    \n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: []\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "results_df_all = pd.DataFrame()\n",
    "print(results_df_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NGS_req-ID                                                             NaN\n",
      "name                                                                 1.2.F\n",
      "Sample                                                                 NaN\n",
      "description                                                  ELANE_AAV_Fwd\n",
      "Direction                                                                5\n",
      "Control sample (Y/N)                                                     Y\n",
      "Notes                                                                  NaN\n",
      "Dilution                                                               NaN\n",
      "Cell name_type                                                 human CD34+\n",
      "UMI_Len                                                                NaN\n",
      "IndexI7Primer                                                          NaN\n",
      "I7_Index_ID                                           P7_N701_SBS12nextera\n",
      "index_I1                                                              N701\n",
      "barcode_I1                                                        TAAGGCGA\n",
      "I5_Index_ID                                           P5_N502_SBS12nextera\n",
      "index_I2                                                              N502\n",
      "barcode_I2                                                        ATAGAGAG\n",
      "primer_seq                                             cttctgggcaggaaccgtg\n",
      "primer                                                        ELANE_F_nest\n",
      "genome                                                                hg38\n",
      "chr                                                                  chr19\n",
      "start                                                               850997\n",
      "end                                                                 856250\n",
      "strand                                                                   +\n",
      "guide_1                                                       sgRNA1_ELANE\n",
      "sequence_guide_1                                      gagtgcagacgttgctgcga\n",
      "genome_guide_1                                                        hg38\n",
      "chr_guide_1                                                          chr19\n",
      "start_guide_1                                                       855742\n",
      "end_guide_1                                                         855762\n",
      "strand_guide_1                                                           -\n",
      "guide_2                                                       sgRNA2_ELANE\n",
      "sequence_guide_2                                      attgcctccttcgtccgggg\n",
      "genome_guide_2                                                        hg38\n",
      "chr_guide_2                                                          chr19\n",
      "start_guide_2                                                       856003\n",
      "end_guide_2                                                         856022\n",
      "strand_guide_2                                                           +\n",
      "guide_3                                                                NaN\n",
      "sequence_guide_3                                                       NaN\n",
      "genome_guide_3                                                         NaN\n",
      "chr_guide_3                                                            NaN\n",
      "start_guide_3                                                          NaN\n",
      "end_guide_3                                                            NaN\n",
      "strand_guide_3                                                         NaN\n",
      "genome_plus_targeting                                                  NaN\n",
      "Replace_Doner_Name                                                     NaN\n",
      "Replace_Donor                                                          NaN\n",
      "HDR_correct_seq          ataatatttataattattttataattataataatatttatataatt...\n",
      "AAVseq                   CCTGCAGGCAGCTGCGCGCTCGCTCGCTCACTGAGGCCGCCCGGGC...\n",
      "Donor (single strand                                                   NaN\n",
      "double strand)                                                         NaN\n",
      "plamid_name                                                  pAAV_noHDRseq\n",
      "plasmid_sequence         GCGGCCGCAGGAACCCCTAGTGATGGAGTTGGCCACTCCCTCTCTG...\n",
      "genomic_background                                                     NaN\n",
      "Name: 1, dtype: object\n"
     ]
    }
   ],
   "source": [
    "amplicon_info = get_csv_data(directory, 1) \n",
    "print(amplicon_info)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  sample_name    i7    i5 sample_descript  Total_mapped_reads  \\\n",
      "0       1.2.F  N701  N502   ELANE_AAV_Fwd               32799   \n",
      "\n",
      "   wt_cut1_total_reads  wt_cut1_total_indels  wt_cut1_total_deletions  \\\n",
      "0                16080                     7                        7   \n",
      "\n",
      "   wt_cut1_total_insertions  wt_cut2_total_reads  ...  1a_1a_cut1_total_reads  \\\n",
      "0                         0                 1457  ...                       0   \n",
      "\n",
      "   1a_1a_cut1_total_indels  1a_1a_cut1_total_deletions  \\\n",
      "0                        0                           0   \n",
      "\n",
      "   1a_1a_cut1_total_insertions  2b_2b_cut1_total_reads  \\\n",
      "0                            0                       0   \n",
      "\n",
      "   2b_2b_cut1_total_indels  2b_2b_cut1_total_deletions  \\\n",
      "0                        0                           0   \n",
      "\n",
      "   2b_2b_cut1_total_insertions  HDR_total_reads  median_fragment_size  \n",
      "0                            0              814                 260.0  \n",
      "\n",
      "[1 rows x 35 columns]\n"
     ]
    }
   ],
   "source": [
    "window_size = 15\n",
    "amplicon_window_around_cut = 1000\n",
    "min_MAPQ = 5\n",
    "min_AS = -180\n",
    "#result_amplicon_df = analyze_alignments_LAM(directory, amplicon_info, window_size, amplicon_window_around_cut, min_MAPQ, min_AS)\n",
    "print(result_amplicon_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
