{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'matplotlib'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-ea8f67e2bc5d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0m__future__\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mprint_function\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mmatplotlib\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mmatplotlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Agg'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mmatplotlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpyplot\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpylab\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'matplotlib'"
     ]
    }
   ],
   "source": [
    "from __future__ import print_function\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import pylab\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "import itertools\n",
    "import operator\n",
    "import subprocess\n",
    "import twobitreader\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pysam\n",
    "import shutil\n",
    "\n",
    "from LAM_scripts.LAM_helpersDanner.py import *\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Pipeline1\n",
    "# This, for the most part, is the UDITAS pipeline modified to allow for REPLACE targeting. \n",
    "\n",
    "## It is not possible to have \n",
    "\n",
    "### 1\n",
    "\n",
    "Ideas:\n",
    "- take sure that the HDR knockin doesnt change between the 5' and 3' because that might due to excess from fake HDR transcripts from the integrated AAV loci\n",
    "- will use the AAV like plasmid\n",
    "- need to check for priming\n",
    "\n",
    "## Inputs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#things taken from original uditas args\n",
    "\n",
    "check_plasmid_insertions = 1\n",
    "ncpu = 4\n",
    "window_size = 15\n",
    "amplicon_window_around_cut = 1000\n",
    "min_MAPQ = 5\n",
    "min_AS = -180\n",
    "process_AMP_seq_run = 0 #off"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/eric/Data/Spaced_Nicking/LAM_MiSeq_HBB_1/Demulitiplexed_corefacility\n"
     ]
    }
   ],
   "source": [
    "#Directory\n",
    "\n",
    "directory = '/home/eric/Data/Spaced_Nicking/LAM_MiSeq_HBB_1/Demulitiplexed_corefacility'\n",
    "\n",
    "#this is a minimal directory of only 300 files for quick debugging\n",
    "#directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/Mishas_Demultiplexed_LAM/10000Reads'\n",
    "\n",
    "print(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'os' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-4edb91d16cb9>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m#assembly = amplicon_info['genome']\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0massembly\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'hg38'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0mfile_genome_2bit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/home/eric/Data/Ref_Genomes'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0massembly\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'.2bit'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_genome_2bit\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'os' is not defined"
     ]
    }
   ],
   "source": [
    "# Assign the file_genome_2bit location. This is needed for pulling sequence from the referene genome by location\n",
    "#assembly = amplicon_info['genome']\n",
    "assembly = 'hg38'\n",
    "file_genome_2bit = os.path.join('/home/eric/Data/Ref_Genomes', assembly + '.2bit')\n",
    "print(file_genome_2bit)\n",
    "\n",
    "# BOWTIE2_INDEXES are needed for global alignments\n",
    "#not sure if this will work\n",
    "#normally in bash: export BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "#check in bash: > ECHO $GENOMES_2BIT\n",
    "\n",
    "%env BOWTIE2_INDEXES=/home/eric/Data/Ref_Genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#the amplicon info is related to the line on the csv file. It is indexed from 0.\n",
    "#it is on my extera space so need to make sure that is mounted. (first thing to check if it throws and error)\n",
    "\n",
    "\n",
    "# THIS IS WHAT RUNS THE PRIMER CHECK THROUGH ALL SAMPLES\n",
    "#input the sample range or list below\n",
    "\n",
    "\n",
    "results_df_all = pd.DataFrame()\n",
    "\n",
    "results_folder = os.path.join(directory, 'results')\n",
    "if not os.path.exists(results_folder):\n",
    "    os.mkdir(results_folder)\n",
    "results_file = os.path.join(directory, 'results','all_priming.xlsx')\n",
    "    \n",
    "## inputs for mispriming\n",
    "seq_skip = 4 # this the number of nucleotides at the begining of the read to skip before checking mismatch\n",
    "mismatches = # the number of mismatches you can have in the primer and downstream seq total\n",
    "downstream = 10 # lenght of sequence downstream of the primer\n",
    "    \n",
    "#make results tree\n",
    "\n",
    "for i in range(24):\n",
    "        \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "\n",
    "    #5primer is everything but AT, AT is for checking mispriming. The full sequence is the olgio from guideseq.\n",
    "    #EVERYTHING IS CAPITAL\n",
    "    ThreePrimeEnd_seq = ''\n",
    "    ThrePrimeEnd_primeronly = ''\n",
    "    #3primer \n",
    "    FivePrimeEnd_seq = ''\n",
    "    FivePrimeEnd_seq_primeronly = ''\n",
    "\n",
    "    direction = amplicon_info['Direction']\n",
    "\n",
    "    if direction == 3:\n",
    "        primer_seq_plus_downstream = ThreePrimeEnd_seq\n",
    "        primer_seq = ThrePrimeEnd_primeronly\n",
    "    elif direction == 5:\n",
    "        primer_seq_plus_downstream = FivePrimeEnd_seq\n",
    "        primer_seq = FivePrimeEnd_seq_primeronly\n",
    "    \n",
    "    df_sample_results = correct_priming2(directory, amplicon_info, primer_seq, primer_seq_plus_downstream, seq_skip, mismatches)\n",
    "    results_df_all = results_df_all.append(df_sample_results, ignore_index=True)\n",
    "    print('done with sample', i)\n",
    "\n",
    "print(results_df_all)\n",
    "results_df_all.to_excel(results_file)    \n",
    "#print reults to excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "this is for pipeline1 LAM- polb rev WT N706 N504 (row 0 in sheet) \n",
      "\n",
      "primer_seq_plus_downstream GGCCAAGCTGGAGCAGGAAATAGATGCACACG\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "NGS_req-ID                                                                     A000H2GWGF\n",
       "name                                                                           LAM_WT_pbR\n",
       "Sample                                                                                 WT\n",
       "description                                                  MiniSeq_K562_Polb_Rev_WT_LAM\n",
       "Control sample (Y/N)                                                                    N\n",
       "Notes                                                                                 NaN\n",
       "Dilution                                                                              NaN\n",
       "Cell name_type                                                                       K562\n",
       "UMI_Len                                                                        NNNNNNNNNN\n",
       "IndexI7Primer                                                                         NaN\n",
       "I7_Index_ID                                                          P7_N706_SBS12nextera\n",
       "index_I1                                                                             N706\n",
       "barcode_I1                                                                       TAGGCATG\n",
       "I5_Index_ID                                                             P5_UMI_N504_Tn5-A\n",
       "index_I2                                                                             N504\n",
       "barcode_I2                                                                       TCTACTCT\n",
       "primer_outfwd                                                                         NaN\n",
       "primer                                                                U1-C-PolB_Rev_nest1\n",
       "genome                                                                               hg38\n",
       "chr                                                                                  chr8\n",
       "start                                                                            42350359\n",
       "end                                                                              42350378\n",
       "strand                                                                                  -\n",
       "guide_1                                                                            Polb#3\n",
       "sequence_guide_1                                                     TAATTTTGTGTGGGTCACCC\n",
       "genome_guide_1                                                                       hg38\n",
       "chr_guide_1                                                                          chr8\n",
       "start_guide_1                                                                    42349873\n",
       "end_guide_1                                                                      42349892\n",
       "strand_guide_1                                                                          +\n",
       "guide_2                                                                            Polb#4\n",
       "sequence_guide_2                                                     TGAAACCAGTTTGGTTACCC\n",
       "genome_guide_2                                                                       hg38\n",
       "chr_guide_2                                                                          chr8\n",
       "start_guide_2                                                                    42350280\n",
       "end_guide_2                                                                      42350299\n",
       "strand_guide_2                                                                          +\n",
       "guide_3                                                                               NaN\n",
       "sequence_guide_3                                                                      NaN\n",
       "genome_guide_3                                                                        NaN\n",
       "chr_guide_3                                                                           NaN\n",
       "start_guide_3                                                                         NaN\n",
       "end_guide_3                                                                           NaN\n",
       "strand_guide_3                                                                        NaN\n",
       "genome_plus_targeting                                              hg38_plus_targetvector\n",
       "Replace_Doner_Name                                                               pE038_MC\n",
       "Replace_Donor                           TGACCCACACAAAATTAGTCTTTTAGCAGACTGGTATGTTTCCAAT...\n",
       "Donor (single strand; double strand)                                                  NaN\n",
       "plamid_name                                                            pE049_WO_redundant\n",
       "plasmid_sequence                        ttacataacttacggtaaatggcccgcctggctgaccgcccaacga...\n",
       "genomic_background                                                                    NaN\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "master_number = 0\n",
    "\n",
    "\n",
    "#################################################################\n",
    "#primer sequences for mispriming. Primer sequence + 12nt downstream \n",
    "#this will only be avialable for read1  (different than for Tn5 setup)\n",
    "\n",
    "##### the sequences need to be capital for this to work########\n",
    "\n",
    "\n",
    "primerlist = [['GGCCAAGCTGGAGCAGGAAATAGATGCACACG', 'this is for pipeline1 LAM- polb rev WT N706 N504 (row 0 in sheet)'],\n",
    "              ['GGCCAAGCTGGAGCAGGAAATAGATGCACACG', 'this is for pipeline1 LAM polb rev Targeting 50% N706 N505(row 1 in sheet)'],\n",
    "              ['TGCATCGCATTGTCTGAGTAGGTGTCATTCTATTC', 'this is for pipeline1 LAM mCherryfwd N707 N505, row2 in sheet'],\n",
    "              ['CCTCCACGTCACCGCATGTTAGCAGACTTCCT', 'this is for mCherryRev LAM  N708 N505 , row3 in sheet']]\n",
    "\n",
    "                \n",
    "\n",
    "primer_seq_plus_downstream = primerlist[master_number][0]          \n",
    "print(primerlist[master_number][1], '\\n')    \n",
    "print('primer_seq_plus_downstream', primer_seq_plus_downstream)\n",
    "\n",
    "\n",
    "#check what is printed below!!!\n",
    "\n",
    "########################################################################\n",
    "# Function to make the 'amplicon_info' list. Taking the line of our experiment csv file\n",
    "\n",
    "def get_csv_data(dir_sample, line_of_data_from_csv):\n",
    "    sample_info_filename = os.path.join(dir_sample, 'sample_info.csv')\n",
    "    experiments = pd.read_csv(sample_info_filename)\n",
    "    return experiments.loc[line_of_data_from_csv]\n",
    "\n",
    "\n",
    "#the amplicon info is related to the line on the csv file. It is indexed from 0.\n",
    "#it is on my extera space so need to make sure that is mounted. (first thing to check if it throws and error)\n",
    "\n",
    "\n",
    "amplicon_info = get_csv_data(directory, master_number)\n",
    "\n",
    "###########check waht is printed below #######\n",
    "amplicon_info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pull out the UMI info and clip the adapter sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of reads: 721904\n",
      "number of unique umis 492408\n"
     ]
    }
   ],
   "source": [
    "umi_joining(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Discard Mispriming Reads\n",
    "When you put a universal primer on the ends of everything, every mispriming event will amplify. An effect we normally don't deal with. I did nested PCR to reduce this. However 85% of the alignments in the UDITAS data I looked at seemed to be mispriming. They did all their blasting and analysis before removing mispriming. But to save computational power and remove error early on I will discard mispriming events. \n",
    "They discard these only for plasmid alignments analyze_alignments_plasmid for some reason which comes from the bam file.\n",
    "\n",
    "### read 1 and 2 are switched. By that I mean the gene specific primer is attached to the P7 sequence instead of P5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total_read_number 721904\n",
      "reads_with_good_priming 694822\n",
      "reads_with_bad_priming 27082\n",
      "perfect matched 677259\n",
      "good but mismatched 17563\n"
     ]
    }
   ],
   "source": [
    "correct_priming_lam(directory, amplicon_info, primer_seq_plus_downstream)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Running the trimming\n",
    "\n",
    "trim_fastq(directory, amplicon_info, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Running the reference plasmid creation\n",
    "create_plasmid_reference(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'replace'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Test the get_reaction_type is working\n",
    "get_reaction_type(amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Make the amplicons by calling the create_amplicon function\n",
    "create_amplicon(directory, amplicon_info, file_genome_2bit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "#try out the alignment to the plasmid\n",
    "align_plasmid_local(directory, amplicon_info, ncpu=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "#extract the unmapped reads\n",
    "extract_unmapped_reads_plasmid(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>target_plus_plasmid_total_reads</th>\n",
       "      <th>target_plus_plasmid_total_reads_collapsed</th>\n",
       "      <th>plasmid_only_total_reads</th>\n",
       "      <th>plasmid_only_total_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11553</td>\n",
       "      <td>11484</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   target_plus_plasmid_total_reads  target_plus_plasmid_total_reads_collapsed  \\\n",
       "0                                0                                          0   \n",
       "\n",
       "   plasmid_only_total_reads  plasmid_only_total_reads_collapsed  \n",
       "0                     11553                               11484  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#run the plasmid analysis to coudn plasmid integration events\n",
    "result_plasmid_df = analyze_alignments_plasmid(directory, amplicon_info, min_MAPQ, file_genome_2bit, True)\n",
    "result_plasmid_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#align against our suite of amplicons\n",
    "align_amplicon(directory, amplicon_info, check_plasmid_insertions, ncpu)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#this will extract unmap reads new folder (files that did not align to the predicted structural variants)\n",
    "extract_unmapped_reads_amplicons(directory, amplicon_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This section is to do the analysis of the amplicon alignments\n",
    "\n",
    "##### need to check my UMI stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of reads before mispriming cleanup 721904\n",
      "number of unique umis before mispriming cleanup 492408\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>all_amplicons_total_reads</th>\n",
       "      <th>all_amplicons_total_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>412997</td>\n",
       "      <td>327383</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   all_amplicons_total_reads  all_amplicons_total_reads_collapsed\n",
       "0                     412997                               327383"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_reads_in_all_amplicons_df = analyze_alignments_all_amplicons(directory, amplicon_info, min_MAPQ, min_AS)\n",
    "result_reads_in_all_amplicons_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wt_cut1_total_reads</th>\n",
       "      <th>wt_cut1_total_indels</th>\n",
       "      <th>wt_cut1_total_deletions</th>\n",
       "      <th>wt_cut1_total_insertions</th>\n",
       "      <th>wt_cut1_total_reads_collapsed</th>\n",
       "      <th>wt_cut1_total_indels_collapsed</th>\n",
       "      <th>wt_cut1_total_deletions_collapsed</th>\n",
       "      <th>wt_cut1_total_insertions_collapsed</th>\n",
       "      <th>wt_cut2_total_reads</th>\n",
       "      <th>wt_cut2_total_indels</th>\n",
       "      <th>...</th>\n",
       "      <th>1a_1a_cut1_total_insertions_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_reads</th>\n",
       "      <th>2b_2b_cut1_total_indels</th>\n",
       "      <th>2b_2b_cut1_total_deletions</th>\n",
       "      <th>2b_2b_cut1_total_insertions</th>\n",
       "      <th>2b_2b_cut1_total_reads_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_indels_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_deletions_collapsed</th>\n",
       "      <th>2b_2b_cut1_total_insertions_collapsed</th>\n",
       "      <th>median_fragment_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>384</td>\n",
       "      <td>384</td>\n",
       "      <td>156</td>\n",
       "      <td>243</td>\n",
       "      <td>384</td>\n",
       "      <td>384</td>\n",
       "      <td>156</td>\n",
       "      <td>243</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>330.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 145 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   wt_cut1_total_reads  wt_cut1_total_indels  wt_cut1_total_deletions  \\\n",
       "0                  384                   384                      156   \n",
       "\n",
       "   wt_cut1_total_insertions  wt_cut1_total_reads_collapsed  \\\n",
       "0                       243                            384   \n",
       "\n",
       "   wt_cut1_total_indels_collapsed  wt_cut1_total_deletions_collapsed  \\\n",
       "0                             384                                156   \n",
       "\n",
       "   wt_cut1_total_insertions_collapsed  wt_cut2_total_reads  \\\n",
       "0                                 243                    1   \n",
       "\n",
       "   wt_cut2_total_indels  ...  1a_1a_cut1_total_insertions_collapsed  \\\n",
       "0                     0  ...                                      0   \n",
       "\n",
       "   2b_2b_cut1_total_reads  2b_2b_cut1_total_indels  \\\n",
       "0                       0                        0   \n",
       "\n",
       "   2b_2b_cut1_total_deletions  2b_2b_cut1_total_insertions  \\\n",
       "0                           0                            0   \n",
       "\n",
       "   2b_2b_cut1_total_reads_collapsed  2b_2b_cut1_total_indels_collapsed  \\\n",
       "0                                 0                                  0   \n",
       "\n",
       "   2b_2b_cut1_total_deletions_collapsed  \\\n",
       "0                                     0   \n",
       "\n",
       "   2b_2b_cut1_total_insertions_collapsed  median_fragment_size  \n",
       "0                                      0                 330.0  \n",
       "\n",
       "[1 rows x 145 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_amplicon_df = analyze_alignments(directory, amplicon_info, window_size, amplicon_window_around_cut, min_MAPQ, min_AS)\n",
    "result_amplicon_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This is the Global alignment of the remaining reads\n",
    "\n",
    "#### What I need here is to modify it so that "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run the allignment\n",
    "assembly = amplicon_info['genome']\n",
    "align_genome_global(directory, amplicon_info, assembly)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of alignments that were attempted to align 5847\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genomewide_total_reads</th>\n",
       "      <th>genomewide_total_reads_collapsed</th>\n",
       "      <th>genomewide_target_only_reads</th>\n",
       "      <th>genomewide_target_only_reads_collapsed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2581</td>\n",
       "      <td>2577</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   genomewide_total_reads  genomewide_total_reads_collapsed  \\\n",
       "0                    2581                              2577   \n",
       "\n",
       "   genomewide_target_only_reads  genomewide_target_only_reads_collapsed  \n",
       "0                             4                                       4  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Do the global alignment analysis\n",
    "results_genome_global_df = analyze_alignments_genome_global(directory, amplicon_info, min_MAPQ, min_AS,  file_genome_2bit)\n",
    "results_genome_global_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summerize data Pipeline1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "694822.0"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_count = pd.DataFrame()\n",
    "read_count.loc[0,'read_count'] = count_reads(directory, amplicon_info)\n",
    "read_count.loc[0,'read_count']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_count</th>\n",
       "      <th>target_plus_plasmid_total_reads</th>\n",
       "      <th>target_plus_plasmid_total_reads_collapsed</th>\n",
       "      <th>plasmid_only_total_reads</th>\n",
       "      <th>plasmid_only_total_reads_collapsed</th>\n",
       "      <th>all_amplicons_total_reads</th>\n",
       "      <th>all_amplicons_total_reads_collapsed</th>\n",
       "      <th>genomewide_total_reads</th>\n",
       "      <th>genomewide_total_reads_collapsed</th>\n",
       "      <th>genomewide_target_only_reads</th>\n",
       "      <th>genomewide_target_only_reads_collapsed</th>\n",
       "      <th>total_aligned</th>\n",
       "      <th>total_aligned_collapsed</th>\n",
       "      <th>percent_aligned</th>\n",
       "      <th>percent_aligned_all_amplicons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>694822.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11553</td>\n",
       "      <td>11484</td>\n",
       "      <td>412997</td>\n",
       "      <td>327383</td>\n",
       "      <td>2581</td>\n",
       "      <td>2577</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>427131</td>\n",
       "      <td>341444</td>\n",
       "      <td>61.473442</td>\n",
       "      <td>96.690945</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   read_count  target_plus_plasmid_total_reads  \\\n",
       "0    694822.0                                0   \n",
       "\n",
       "   target_plus_plasmid_total_reads_collapsed  plasmid_only_total_reads  \\\n",
       "0                                          0                     11553   \n",
       "\n",
       "   plasmid_only_total_reads_collapsed  all_amplicons_total_reads  \\\n",
       "0                               11484                     412997   \n",
       "\n",
       "   all_amplicons_total_reads_collapsed  genomewide_total_reads  \\\n",
       "0                               327383                    2581   \n",
       "\n",
       "   genomewide_total_reads_collapsed  genomewide_target_only_reads  \\\n",
       "0                              2577                             4   \n",
       "\n",
       "   genomewide_target_only_reads_collapsed  total_aligned  \\\n",
       "0                                       4         427131   \n",
       "\n",
       "   total_aligned_collapsed  percent_aligned  percent_aligned_all_amplicons  \n",
       "0                   341444        61.473442                      96.690945  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#summerize the data\n",
    "\n",
    "summary_all_alignments = get_summary_all_alignments(directory, amplicon_info, read_count.loc[[0]], result_plasmid_df, result_reads_in_all_amplicons_df, results_genome_global_df)\n",
    "summary_all_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NGS_req-ID</th>\n",
       "      <th>name</th>\n",
       "      <th>Sample</th>\n",
       "      <th>description</th>\n",
       "      <th>Control sample (Y/N)</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Dilution</th>\n",
       "      <th>Cell name_type</th>\n",
       "      <th>UMI_Len</th>\n",
       "      <th>IndexI7Primer</th>\n",
       "      <th>...</th>\n",
       "      <th>hiti_3prime_replace_exon_rev_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>doner_tail_tail_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>doner_head_tail_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>doner_head_head_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>1a_1a_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>2b_2b_cut1_total_reads_collapsed_percent</th>\n",
       "      <th>target_plus_plasmid_total_reads_collapsed_percent</th>\n",
       "      <th>plasmid_only_total_reads_collapsed_percent</th>\n",
       "      <th>Type</th>\n",
       "      <th>Percent Editing</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.019656</td>\n",
       "      <td>0.274092</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.540541</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>0.411434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.019656</td>\n",
       "      <td>0.274092</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.540541</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>0.001071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>wt_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.019656</td>\n",
       "      <td>0.274092</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.540541</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_deletion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.019656</td>\n",
       "      <td>0.274092</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.540541</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>0.002143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.019656</td>\n",
       "      <td>0.274092</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.540541</td>\n",
       "      <td>large_inversion_cut2_total_reads_percent</td>\n",
       "      <td>0.099644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>large_inversion_cut2_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doner_head_head_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.019656</td>\n",
       "      <td>0.274092</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.540541</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1a_1a_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.019656</td>\n",
       "      <td>0.274092</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.540541</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2b_2b_cut1_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.019656</td>\n",
       "      <td>0.274092</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.540541</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>target_plus_plasmid_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_WT_pbR</td>\n",
       "      <td>WT</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_WT_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002184</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.019656</td>\n",
       "      <td>0.274092</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.540541</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>12.378391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol50_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_50per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_pbR</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_Rev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_bpA</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrFwd_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>A000H2GWGF</td>\n",
       "      <td>LAM_Pol80_mCherRev</td>\n",
       "      <td>Polb</td>\n",
       "      <td>MiniSeq_K562_Polb_mCherrRev_80per_LAM</td>\n",
       "      <td>N</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K562</td>\n",
       "      <td>NNNNNNNNNN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>plasmid_only_total_reads_percent</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>140 rows × 224 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     NGS_req-ID                name Sample  \\\n",
       "0    A000H2GWGF          LAM_WT_pbR     WT   \n",
       "1    A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "2    A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "3    A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "4    A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "5    A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "6    A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "7    A000H2GWGF          LAM_WT_pbR     WT   \n",
       "8    A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "9    A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "10   A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "11   A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "12   A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "13   A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "14   A000H2GWGF          LAM_WT_pbR     WT   \n",
       "15   A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "16   A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "17   A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "18   A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "19   A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "20   A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "21   A000H2GWGF          LAM_WT_pbR     WT   \n",
       "22   A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "23   A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "24   A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "25   A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "26   A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "27   A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "28   A000H2GWGF          LAM_WT_pbR     WT   \n",
       "29   A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "..          ...                 ...    ...   \n",
       "110  A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "111  A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "112  A000H2GWGF          LAM_WT_pbR     WT   \n",
       "113  A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "114  A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "115  A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "116  A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "117  A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "118  A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "119  A000H2GWGF          LAM_WT_pbR     WT   \n",
       "120  A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "121  A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "122  A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "123  A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "124  A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "125  A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "126  A000H2GWGF          LAM_WT_pbR     WT   \n",
       "127  A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "128  A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "129  A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "130  A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "131  A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "132  A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "133  A000H2GWGF          LAM_WT_pbR     WT   \n",
       "134  A000H2GWGF       LAM_Pol50_pbR   Polb   \n",
       "135  A000H2GWGF       LAM_Pol50_bpA   Polb   \n",
       "136  A000H2GWGF  LAM_Pol50_mCherRev   Polb   \n",
       "137  A000H2GWGF       LAM_Pol80_pbR   Polb   \n",
       "138  A000H2GWGF       LAM_Pol80_bpA   Polb   \n",
       "139  A000H2GWGF  LAM_Pol80_mCherRev   Polb   \n",
       "\n",
       "                               description Control sample (Y/N)  Notes  \\\n",
       "0             MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "1          MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "2    MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "3    MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "4          MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "5    MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "6    MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "7             MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "8          MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "9    MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "10   MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "11         MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "12   MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "13   MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "14            MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "15         MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "16   MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "17   MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "18         MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "19   MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "20   MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "21            MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "22         MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "23   MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "24   MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "25         MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "26   MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "27   MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "28            MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "29         MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "..                                     ...                  ...    ...   \n",
       "110  MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "111  MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "112           MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "113        MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "114  MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "115  MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "116        MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "117  MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "118  MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "119           MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "120        MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "121  MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "122  MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "123        MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "124  MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "125  MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "126           MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "127        MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "128  MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "129  MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "130        MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "131  MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "132  MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "133           MiniSeq_K562_Polb_Rev_WT_LAM                    N    NaN   \n",
       "134        MiniSeq_K562_Polb_Rev_50per_LAM                    N    NaN   \n",
       "135  MiniSeq_K562_Polb_mCherrFwd_50per_LAM                    N    NaN   \n",
       "136  MiniSeq_K562_Polb_mCherrRev_50per_LAM                    N    NaN   \n",
       "137        MiniSeq_K562_Polb_Rev_80per_LAM                    N    NaN   \n",
       "138  MiniSeq_K562_Polb_mCherrFwd_80per_LAM                    N    NaN   \n",
       "139  MiniSeq_K562_Polb_mCherrRev_80per_LAM                    N    NaN   \n",
       "\n",
       "     Dilution Cell name_type     UMI_Len  IndexI7Primer  ...  \\\n",
       "0         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "1         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "2         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "3         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "4         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "5         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "6         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "7         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "8         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "9         NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "10        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "11        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "12        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "13        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "14        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "15        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "16        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "17        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "18        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "19        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "20        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "21        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "22        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "23        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "24        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "25        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "26        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "27        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "28        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "29        NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "..        ...            ...         ...            ...  ...   \n",
       "110       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "111       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "112       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "113       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "114       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "115       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "116       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "117       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "118       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "119       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "120       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "121       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "122       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "123       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "124       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "125       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "126       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "127       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "128       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "129       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "130       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "131       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "132       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "133       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "134       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "135       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "136       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "137       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "138       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "139       NaN           K562  NNNNNNNNNN            NaN  ...   \n",
       "\n",
       "    hiti_3prime_replace_exon_rev_cut1_total_reads_collapsed_percent  \\\n",
       "0                                             0.002184                \n",
       "1                                                  NaN                \n",
       "2                                                  NaN                \n",
       "3                                                  NaN                \n",
       "4                                                  NaN                \n",
       "5                                                  NaN                \n",
       "6                                                  NaN                \n",
       "7                                             0.002184                \n",
       "8                                                  NaN                \n",
       "9                                                  NaN                \n",
       "10                                                 NaN                \n",
       "11                                                 NaN                \n",
       "12                                                 NaN                \n",
       "13                                                 NaN                \n",
       "14                                            0.002184                \n",
       "15                                                 NaN                \n",
       "16                                                 NaN                \n",
       "17                                                 NaN                \n",
       "18                                                 NaN                \n",
       "19                                                 NaN                \n",
       "20                                                 NaN                \n",
       "21                                            0.002184                \n",
       "22                                                 NaN                \n",
       "23                                                 NaN                \n",
       "24                                                 NaN                \n",
       "25                                                 NaN                \n",
       "26                                                 NaN                \n",
       "27                                                 NaN                \n",
       "28                                            0.002184                \n",
       "29                                                 NaN                \n",
       "..                                                 ...                \n",
       "110                                                NaN                \n",
       "111                                                NaN                \n",
       "112                                           0.002184                \n",
       "113                                                NaN                \n",
       "114                                                NaN                \n",
       "115                                                NaN                \n",
       "116                                                NaN                \n",
       "117                                                NaN                \n",
       "118                                                NaN                \n",
       "119                                           0.002184                \n",
       "120                                                NaN                \n",
       "121                                                NaN                \n",
       "122                                                NaN                \n",
       "123                                                NaN                \n",
       "124                                                NaN                \n",
       "125                                                NaN                \n",
       "126                                           0.002184                \n",
       "127                                                NaN                \n",
       "128                                                NaN                \n",
       "129                                                NaN                \n",
       "130                                                NaN                \n",
       "131                                                NaN                \n",
       "132                                                NaN                \n",
       "133                                           0.002184                \n",
       "134                                                NaN                \n",
       "135                                                NaN                \n",
       "136                                                NaN                \n",
       "137                                                NaN                \n",
       "138                                                NaN                \n",
       "139                                                NaN                \n",
       "\n",
       "    doner_tail_tail_cut1_total_reads_collapsed_percent  \\\n",
       "0                                                  0.0   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "5                                                  NaN   \n",
       "6                                                  NaN   \n",
       "7                                                  0.0   \n",
       "8                                                  NaN   \n",
       "9                                                  NaN   \n",
       "10                                                 NaN   \n",
       "11                                                 NaN   \n",
       "12                                                 NaN   \n",
       "13                                                 NaN   \n",
       "14                                                 0.0   \n",
       "15                                                 NaN   \n",
       "16                                                 NaN   \n",
       "17                                                 NaN   \n",
       "18                                                 NaN   \n",
       "19                                                 NaN   \n",
       "20                                                 NaN   \n",
       "21                                                 0.0   \n",
       "22                                                 NaN   \n",
       "23                                                 NaN   \n",
       "24                                                 NaN   \n",
       "25                                                 NaN   \n",
       "26                                                 NaN   \n",
       "27                                                 NaN   \n",
       "28                                                 0.0   \n",
       "29                                                 NaN   \n",
       "..                                                 ...   \n",
       "110                                                NaN   \n",
       "111                                                NaN   \n",
       "112                                                0.0   \n",
       "113                                                NaN   \n",
       "114                                                NaN   \n",
       "115                                                NaN   \n",
       "116                                                NaN   \n",
       "117                                                NaN   \n",
       "118                                                NaN   \n",
       "119                                                0.0   \n",
       "120                                                NaN   \n",
       "121                                                NaN   \n",
       "122                                                NaN   \n",
       "123                                                NaN   \n",
       "124                                                NaN   \n",
       "125                                                NaN   \n",
       "126                                                0.0   \n",
       "127                                                NaN   \n",
       "128                                                NaN   \n",
       "129                                                NaN   \n",
       "130                                                NaN   \n",
       "131                                                NaN   \n",
       "132                                                NaN   \n",
       "133                                                0.0   \n",
       "134                                                NaN   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "138                                                NaN   \n",
       "139                                                NaN   \n",
       "\n",
       "    doner_head_tail_cut1_total_reads_collapsed_percent  \\\n",
       "0                                            52.019656   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "5                                                  NaN   \n",
       "6                                                  NaN   \n",
       "7                                            52.019656   \n",
       "8                                                  NaN   \n",
       "9                                                  NaN   \n",
       "10                                                 NaN   \n",
       "11                                                 NaN   \n",
       "12                                                 NaN   \n",
       "13                                                 NaN   \n",
       "14                                           52.019656   \n",
       "15                                                 NaN   \n",
       "16                                                 NaN   \n",
       "17                                                 NaN   \n",
       "18                                                 NaN   \n",
       "19                                                 NaN   \n",
       "20                                                 NaN   \n",
       "21                                           52.019656   \n",
       "22                                                 NaN   \n",
       "23                                                 NaN   \n",
       "24                                                 NaN   \n",
       "25                                                 NaN   \n",
       "26                                                 NaN   \n",
       "27                                                 NaN   \n",
       "28                                           52.019656   \n",
       "29                                                 NaN   \n",
       "..                                                 ...   \n",
       "110                                                NaN   \n",
       "111                                                NaN   \n",
       "112                                          52.019656   \n",
       "113                                                NaN   \n",
       "114                                                NaN   \n",
       "115                                                NaN   \n",
       "116                                                NaN   \n",
       "117                                                NaN   \n",
       "118                                                NaN   \n",
       "119                                          52.019656   \n",
       "120                                                NaN   \n",
       "121                                                NaN   \n",
       "122                                                NaN   \n",
       "123                                                NaN   \n",
       "124                                                NaN   \n",
       "125                                                NaN   \n",
       "126                                          52.019656   \n",
       "127                                                NaN   \n",
       "128                                                NaN   \n",
       "129                                                NaN   \n",
       "130                                                NaN   \n",
       "131                                                NaN   \n",
       "132                                                NaN   \n",
       "133                                          52.019656   \n",
       "134                                                NaN   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "138                                                NaN   \n",
       "139                                                NaN   \n",
       "\n",
       "    doner_head_head_cut1_total_reads_collapsed_percent  \\\n",
       "0                                             0.274092   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "5                                                  NaN   \n",
       "6                                                  NaN   \n",
       "7                                             0.274092   \n",
       "8                                                  NaN   \n",
       "9                                                  NaN   \n",
       "10                                                 NaN   \n",
       "11                                                 NaN   \n",
       "12                                                 NaN   \n",
       "13                                                 NaN   \n",
       "14                                            0.274092   \n",
       "15                                                 NaN   \n",
       "16                                                 NaN   \n",
       "17                                                 NaN   \n",
       "18                                                 NaN   \n",
       "19                                                 NaN   \n",
       "20                                                 NaN   \n",
       "21                                            0.274092   \n",
       "22                                                 NaN   \n",
       "23                                                 NaN   \n",
       "24                                                 NaN   \n",
       "25                                                 NaN   \n",
       "26                                                 NaN   \n",
       "27                                                 NaN   \n",
       "28                                            0.274092   \n",
       "29                                                 NaN   \n",
       "..                                                 ...   \n",
       "110                                                NaN   \n",
       "111                                                NaN   \n",
       "112                                           0.274092   \n",
       "113                                                NaN   \n",
       "114                                                NaN   \n",
       "115                                                NaN   \n",
       "116                                                NaN   \n",
       "117                                                NaN   \n",
       "118                                                NaN   \n",
       "119                                           0.274092   \n",
       "120                                                NaN   \n",
       "121                                                NaN   \n",
       "122                                                NaN   \n",
       "123                                                NaN   \n",
       "124                                                NaN   \n",
       "125                                                NaN   \n",
       "126                                           0.274092   \n",
       "127                                                NaN   \n",
       "128                                                NaN   \n",
       "129                                                NaN   \n",
       "130                                                NaN   \n",
       "131                                                NaN   \n",
       "132                                                NaN   \n",
       "133                                           0.274092   \n",
       "134                                                NaN   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "138                                                NaN   \n",
       "139                                                NaN   \n",
       "\n",
       "    1a_1a_cut1_total_reads_collapsed_percent  \\\n",
       "0                                        0.0   \n",
       "1                                        NaN   \n",
       "2                                        NaN   \n",
       "3                                        NaN   \n",
       "4                                        NaN   \n",
       "5                                        NaN   \n",
       "6                                        NaN   \n",
       "7                                        0.0   \n",
       "8                                        NaN   \n",
       "9                                        NaN   \n",
       "10                                       NaN   \n",
       "11                                       NaN   \n",
       "12                                       NaN   \n",
       "13                                       NaN   \n",
       "14                                       0.0   \n",
       "15                                       NaN   \n",
       "16                                       NaN   \n",
       "17                                       NaN   \n",
       "18                                       NaN   \n",
       "19                                       NaN   \n",
       "20                                       NaN   \n",
       "21                                       0.0   \n",
       "22                                       NaN   \n",
       "23                                       NaN   \n",
       "24                                       NaN   \n",
       "25                                       NaN   \n",
       "26                                       NaN   \n",
       "27                                       NaN   \n",
       "28                                       0.0   \n",
       "29                                       NaN   \n",
       "..                                       ...   \n",
       "110                                      NaN   \n",
       "111                                      NaN   \n",
       "112                                      0.0   \n",
       "113                                      NaN   \n",
       "114                                      NaN   \n",
       "115                                      NaN   \n",
       "116                                      NaN   \n",
       "117                                      NaN   \n",
       "118                                      NaN   \n",
       "119                                      0.0   \n",
       "120                                      NaN   \n",
       "121                                      NaN   \n",
       "122                                      NaN   \n",
       "123                                      NaN   \n",
       "124                                      NaN   \n",
       "125                                      NaN   \n",
       "126                                      0.0   \n",
       "127                                      NaN   \n",
       "128                                      NaN   \n",
       "129                                      NaN   \n",
       "130                                      NaN   \n",
       "131                                      NaN   \n",
       "132                                      NaN   \n",
       "133                                      0.0   \n",
       "134                                      NaN   \n",
       "135                                      NaN   \n",
       "136                                      NaN   \n",
       "137                                      NaN   \n",
       "138                                      NaN   \n",
       "139                                      NaN   \n",
       "\n",
       "    2b_2b_cut1_total_reads_collapsed_percent  \\\n",
       "0                                        0.0   \n",
       "1                                        NaN   \n",
       "2                                        NaN   \n",
       "3                                        NaN   \n",
       "4                                        NaN   \n",
       "5                                        NaN   \n",
       "6                                        NaN   \n",
       "7                                        0.0   \n",
       "8                                        NaN   \n",
       "9                                        NaN   \n",
       "10                                       NaN   \n",
       "11                                       NaN   \n",
       "12                                       NaN   \n",
       "13                                       NaN   \n",
       "14                                       0.0   \n",
       "15                                       NaN   \n",
       "16                                       NaN   \n",
       "17                                       NaN   \n",
       "18                                       NaN   \n",
       "19                                       NaN   \n",
       "20                                       NaN   \n",
       "21                                       0.0   \n",
       "22                                       NaN   \n",
       "23                                       NaN   \n",
       "24                                       NaN   \n",
       "25                                       NaN   \n",
       "26                                       NaN   \n",
       "27                                       NaN   \n",
       "28                                       0.0   \n",
       "29                                       NaN   \n",
       "..                                       ...   \n",
       "110                                      NaN   \n",
       "111                                      NaN   \n",
       "112                                      0.0   \n",
       "113                                      NaN   \n",
       "114                                      NaN   \n",
       "115                                      NaN   \n",
       "116                                      NaN   \n",
       "117                                      NaN   \n",
       "118                                      NaN   \n",
       "119                                      0.0   \n",
       "120                                      NaN   \n",
       "121                                      NaN   \n",
       "122                                      NaN   \n",
       "123                                      NaN   \n",
       "124                                      NaN   \n",
       "125                                      NaN   \n",
       "126                                      0.0   \n",
       "127                                      NaN   \n",
       "128                                      NaN   \n",
       "129                                      NaN   \n",
       "130                                      NaN   \n",
       "131                                      NaN   \n",
       "132                                      NaN   \n",
       "133                                      0.0   \n",
       "134                                      NaN   \n",
       "135                                      NaN   \n",
       "136                                      NaN   \n",
       "137                                      NaN   \n",
       "138                                      NaN   \n",
       "139                                      NaN   \n",
       "\n",
       "     target_plus_plasmid_total_reads_collapsed_percent  \\\n",
       "0                                                  0.0   \n",
       "1                                                  NaN   \n",
       "2                                                  NaN   \n",
       "3                                                  NaN   \n",
       "4                                                  NaN   \n",
       "5                                                  NaN   \n",
       "6                                                  NaN   \n",
       "7                                                  0.0   \n",
       "8                                                  NaN   \n",
       "9                                                  NaN   \n",
       "10                                                 NaN   \n",
       "11                                                 NaN   \n",
       "12                                                 NaN   \n",
       "13                                                 NaN   \n",
       "14                                                 0.0   \n",
       "15                                                 NaN   \n",
       "16                                                 NaN   \n",
       "17                                                 NaN   \n",
       "18                                                 NaN   \n",
       "19                                                 NaN   \n",
       "20                                                 NaN   \n",
       "21                                                 0.0   \n",
       "22                                                 NaN   \n",
       "23                                                 NaN   \n",
       "24                                                 NaN   \n",
       "25                                                 NaN   \n",
       "26                                                 NaN   \n",
       "27                                                 NaN   \n",
       "28                                                 0.0   \n",
       "29                                                 NaN   \n",
       "..                                                 ...   \n",
       "110                                                NaN   \n",
       "111                                                NaN   \n",
       "112                                                0.0   \n",
       "113                                                NaN   \n",
       "114                                                NaN   \n",
       "115                                                NaN   \n",
       "116                                                NaN   \n",
       "117                                                NaN   \n",
       "118                                                NaN   \n",
       "119                                                0.0   \n",
       "120                                                NaN   \n",
       "121                                                NaN   \n",
       "122                                                NaN   \n",
       "123                                                NaN   \n",
       "124                                                NaN   \n",
       "125                                                NaN   \n",
       "126                                                0.0   \n",
       "127                                                NaN   \n",
       "128                                                NaN   \n",
       "129                                                NaN   \n",
       "130                                                NaN   \n",
       "131                                                NaN   \n",
       "132                                                NaN   \n",
       "133                                                0.0   \n",
       "134                                                NaN   \n",
       "135                                                NaN   \n",
       "136                                                NaN   \n",
       "137                                                NaN   \n",
       "138                                                NaN   \n",
       "139                                                NaN   \n",
       "\n",
       "    plasmid_only_total_reads_collapsed_percent  \\\n",
       "0                                    12.540541   \n",
       "1                                          NaN   \n",
       "2                                          NaN   \n",
       "3                                          NaN   \n",
       "4                                          NaN   \n",
       "5                                          NaN   \n",
       "6                                          NaN   \n",
       "7                                    12.540541   \n",
       "8                                          NaN   \n",
       "9                                          NaN   \n",
       "10                                         NaN   \n",
       "11                                         NaN   \n",
       "12                                         NaN   \n",
       "13                                         NaN   \n",
       "14                                   12.540541   \n",
       "15                                         NaN   \n",
       "16                                         NaN   \n",
       "17                                         NaN   \n",
       "18                                         NaN   \n",
       "19                                         NaN   \n",
       "20                                         NaN   \n",
       "21                                   12.540541   \n",
       "22                                         NaN   \n",
       "23                                         NaN   \n",
       "24                                         NaN   \n",
       "25                                         NaN   \n",
       "26                                         NaN   \n",
       "27                                         NaN   \n",
       "28                                   12.540541   \n",
       "29                                         NaN   \n",
       "..                                         ...   \n",
       "110                                        NaN   \n",
       "111                                        NaN   \n",
       "112                                  12.540541   \n",
       "113                                        NaN   \n",
       "114                                        NaN   \n",
       "115                                        NaN   \n",
       "116                                        NaN   \n",
       "117                                        NaN   \n",
       "118                                        NaN   \n",
       "119                                  12.540541   \n",
       "120                                        NaN   \n",
       "121                                        NaN   \n",
       "122                                        NaN   \n",
       "123                                        NaN   \n",
       "124                                        NaN   \n",
       "125                                        NaN   \n",
       "126                                  12.540541   \n",
       "127                                        NaN   \n",
       "128                                        NaN   \n",
       "129                                        NaN   \n",
       "130                                        NaN   \n",
       "131                                        NaN   \n",
       "132                                        NaN   \n",
       "133                                  12.540541   \n",
       "134                                        NaN   \n",
       "135                                        NaN   \n",
       "136                                        NaN   \n",
       "137                                        NaN   \n",
       "138                                        NaN   \n",
       "139                                        NaN   \n",
       "\n",
       "                                         Type Percent Editing  \n",
       "0                 wt_cut1_total_reads_percent        0.411434  \n",
       "1                 wt_cut1_total_reads_percent             NaN  \n",
       "2                 wt_cut1_total_reads_percent             NaN  \n",
       "3                 wt_cut1_total_reads_percent             NaN  \n",
       "4                 wt_cut1_total_reads_percent             NaN  \n",
       "5                 wt_cut1_total_reads_percent             NaN  \n",
       "6                 wt_cut1_total_reads_percent             NaN  \n",
       "7                 wt_cut2_total_reads_percent        0.001071  \n",
       "8                 wt_cut2_total_reads_percent             NaN  \n",
       "9                 wt_cut2_total_reads_percent             NaN  \n",
       "10                wt_cut2_total_reads_percent             NaN  \n",
       "11                wt_cut2_total_reads_percent             NaN  \n",
       "12                wt_cut2_total_reads_percent             NaN  \n",
       "13                wt_cut2_total_reads_percent             NaN  \n",
       "14    large_deletion_cut1_total_reads_percent        0.000000  \n",
       "15    large_deletion_cut1_total_reads_percent             NaN  \n",
       "16    large_deletion_cut1_total_reads_percent             NaN  \n",
       "17    large_deletion_cut1_total_reads_percent             NaN  \n",
       "18    large_deletion_cut1_total_reads_percent             NaN  \n",
       "19    large_deletion_cut1_total_reads_percent             NaN  \n",
       "20    large_deletion_cut1_total_reads_percent             NaN  \n",
       "21   large_inversion_cut1_total_reads_percent        0.002143  \n",
       "22   large_inversion_cut1_total_reads_percent             NaN  \n",
       "23   large_inversion_cut1_total_reads_percent             NaN  \n",
       "24   large_inversion_cut1_total_reads_percent             NaN  \n",
       "25   large_inversion_cut1_total_reads_percent             NaN  \n",
       "26   large_inversion_cut1_total_reads_percent             NaN  \n",
       "27   large_inversion_cut1_total_reads_percent             NaN  \n",
       "28   large_inversion_cut2_total_reads_percent        0.099644  \n",
       "29   large_inversion_cut2_total_reads_percent             NaN  \n",
       "..                                        ...             ...  \n",
       "110  doner_head_head_cut1_total_reads_percent             NaN  \n",
       "111  doner_head_head_cut1_total_reads_percent             NaN  \n",
       "112            1a_1a_cut1_total_reads_percent        0.000000  \n",
       "113            1a_1a_cut1_total_reads_percent             NaN  \n",
       "114            1a_1a_cut1_total_reads_percent             NaN  \n",
       "115            1a_1a_cut1_total_reads_percent             NaN  \n",
       "116            1a_1a_cut1_total_reads_percent             NaN  \n",
       "117            1a_1a_cut1_total_reads_percent             NaN  \n",
       "118            1a_1a_cut1_total_reads_percent             NaN  \n",
       "119            2b_2b_cut1_total_reads_percent        0.000000  \n",
       "120            2b_2b_cut1_total_reads_percent             NaN  \n",
       "121            2b_2b_cut1_total_reads_percent             NaN  \n",
       "122            2b_2b_cut1_total_reads_percent             NaN  \n",
       "123            2b_2b_cut1_total_reads_percent             NaN  \n",
       "124            2b_2b_cut1_total_reads_percent             NaN  \n",
       "125            2b_2b_cut1_total_reads_percent             NaN  \n",
       "126   target_plus_plasmid_total_reads_percent        0.000000  \n",
       "127   target_plus_plasmid_total_reads_percent             NaN  \n",
       "128   target_plus_plasmid_total_reads_percent             NaN  \n",
       "129   target_plus_plasmid_total_reads_percent             NaN  \n",
       "130   target_plus_plasmid_total_reads_percent             NaN  \n",
       "131   target_plus_plasmid_total_reads_percent             NaN  \n",
       "132   target_plus_plasmid_total_reads_percent             NaN  \n",
       "133          plasmid_only_total_reads_percent       12.378391  \n",
       "134          plasmid_only_total_reads_percent             NaN  \n",
       "135          plasmid_only_total_reads_percent             NaN  \n",
       "136          plasmid_only_total_reads_percent             NaN  \n",
       "137          plasmid_only_total_reads_percent             NaN  \n",
       "138          plasmid_only_total_reads_percent             NaN  \n",
       "139          plasmid_only_total_reads_percent             NaN  \n",
       "\n",
       "[140 rows x 224 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_alignments_junction = pd.concat([result_amplicon_df, result_plasmid_df], axis=1)\n",
    "sample_info_filename = os.path.join(directory, 'sample_info.csv')\n",
    "\n",
    "experiments = pd.read_csv(sample_info_filename)\n",
    "\n",
    "N7 = amplicon_info['index_I1']\n",
    "N5 = amplicon_info['index_I2']\n",
    "\n",
    "results_summary = summarize_results(results_alignments_junction)\n",
    "#this outputs it in a directory above the N501_701 folder as a summary file. Need to label it or it will be overwritten by next tone\n",
    "results_summary_with_experiments = pd.concat([experiments, results_summary], axis=1)\n",
    "\n",
    "results_summary_with_experiments.to_excel(os.path.join(directory, N7 + '_' + N5, N7 + '_' + N5+'_'+'results_summary.xlsx'))\n",
    "\n",
    "results_pivot = melt_results_lam(results_summary_with_experiments)\n",
    "results_pivot.to_excel(os.path.join(directory, N7 + '_' + N5, N7 + '_' + N5+'_'+'results_summary_pivot_no_UMI_collapse.xlsx'))\n",
    "results_pivot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline2 \n",
    "\n",
    "## Need to make an new bowtie2 index file that includes targeting for alignment agains the whole genome so it is all in one sheet together. \n",
    "\n",
    "### Other option would be to align it to the amplicon, extract unaligned files and then align to the genome but seems cleaner this way. \n",
    "#### Ideally every sequence is unique between the targeting vector and genome.\n",
    "\n",
    "\n",
    "1. Build your fastas of interest and label .fa files.\n",
    "    1. You need fasta of hg38 or reference genome. You can pull this from downloaded bowtie indexed sampels and then use the following command to turn the index into a fasta file: bowtie2-inspect hg38 > hg38.fa   \n",
    "    2. Put all the fasta files in the same folder. Should also use the transfected plasmid\n",
    "2. index the files with bowtie\n",
    "    1. use the command bowtie2-build -f pE049,pe038_mc.fa,hg38.fa -p hg38_plus_targetvectorandplasmid\n",
    "    2. this has the -p to make it take less ram in my case.\n",
    "    3. In this case it adds the hg38 and the minicircle targeting file together\n",
    "    4. I have a Intel® Core™ i7-5500U CPU @ 2.40GHz × 4 with 15.1 GiB ram and it required about 13.8 gigs of ram and 2 hours to do a hg38+small fasta index\n",
    "    5. be sure to pay attention to the name of the new indexed file. \"hg38_plus_targetvector\" in example above. Add it to the sample_info.csv sheet. Under the tab 'genome_plus_targeting'.\n",
    "    6. you can check it indexed correcctly: bowtie2-inspect -s hg38_plus_targetvector\n",
    "    \n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'amplicon_info' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-b48417b7d299>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Assign the file_genome_2bit location. This is needed for pulling sequence from the referene genome by location\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m#need a new build\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0massembly_plus_targetvector\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mamplicon_info\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'genome_plus_targeting'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mfile_genome_2bit_plus_target\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0massembly_plus_targetvector\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'.2bit'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_genome_2bit_plus_target\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'amplicon_info' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pipeline2 Logic\n",
    "#  This is a second pipeline. Different than Uditas but built in a simliar form. \n",
    "## It uses the read distal to the gene specic primer. This read covers the region downstream of the break. After clipping off anything remaing upstream of the break we make a global alignment to look at genomic integrations and translocations. This also quantifies the alignments.\n",
    "\n",
    "\n",
    "### 0. Use correct_prime function in pipeline1 to arrive at reads that were correctly primed.\n",
    "### 1. Trim correctly primed reads to the break site  \n",
    "### 2. Only use a single end of the pair end reads (distal gene specific primer)\n",
    "### 3. Use bowtie2 genome that has the targeting seq in it added\n",
    "### 4. Do global end-to-end alignment\n",
    "### 5. Do a local alignment\n",
    "### 6. process bam file and pull out the high mapQ reads\n",
    "\n",
    "The idea here is to trim the sequence up to the cut on both reads and then do a global alignment end-to-end agasint the human genome.\n",
    "\n",
    "The idea here is to take the files that are 'trimmed' and 'correct priming' checking. Then to do another round of trimming to get down to the cut site. I need to make another folder for 'trimmed_to_cut'\n",
    "\n",
    "For Tn5 setup the 'Read 1' contains the sequence away from the gene specific primer. So it is the read we should concentration on if we don't do paired end.\n",
    "\n",
    "For LAM the 'Read 2' is the the sequence distal to the gene specific primer"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# if you want to do pipeline2 in one piece"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the directory is: /media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/Mishas_Demultiplexed_LAM\n",
      "/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes/hg38_plus_targetvector.2bit\n",
      "env: BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
      "running row 2\n",
      "this is for pipeline1 LAM mCherryFwd N707 N505, row2 in sheet \n",
      "\n",
      "the analysis is on LAM sample: LAM_Pol50_bpA\n",
      "-----  now making bed file   ------\n",
      "made global bed file: N707 _ N505\n",
      "doing the quantification of global alignments\n",
      "   %_target_site_reads  %_target_vector_reads  %_translocations  \\\n",
      "0            26.190761              51.203509          22.60573   \n",
      "\n",
      "   alignments_target_site_all  alignments_target_vector_all  \\\n",
      "0                       65743                        128529   \n",
      "\n",
      "   alignments_translocations_all  total aligned reads filtered  \\\n",
      "0                          56744                      251016.0   \n",
      "\n",
      "   all_reads_unfiltered  \n",
      "0                311453  \n",
      "running row 3\n",
      "this is for mCherryrev LAM  N708 N505 , row3 in sheet \n",
      "\n",
      "the analysis is on LAM sample: LAM_Pol50_mCherRev\n",
      "-----  now making bed file   ------\n",
      "made global bed file: N708 _ N505\n",
      "doing the quantification of global alignments\n",
      "   %_target_site_reads  %_target_vector_reads  %_translocations  \\\n",
      "0            25.852889              52.861703         21.285408   \n",
      "\n",
      "   alignments_target_site_all  alignments_target_vector_all  \\\n",
      "0                      124151                        253853   \n",
      "\n",
      "   alignments_translocations_all  total aligned reads filtered  \\\n",
      "0                         102217                      480221.0   \n",
      "\n",
      "   all_reads_unfiltered  \n",
      "0                516825  \n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "check_plasmid_insertions = 1\n",
    "ncpu = 4\n",
    "window_size = 15\n",
    "amplicon_window_around_cut = 1000\n",
    "min_MAPQ = 5\n",
    "min_AS = -180\n",
    "process_AMP_seq_run = 0 #off\n",
    "\n",
    "def get_csv_data(dir_sample, line_of_data_from_csv):\n",
    "    sample_info_filename = os.path.join(dir_sample, 'sample_info.csv')\n",
    "    experiments = pd.read_csv(sample_info_filename)\n",
    "    return experiments.loc[line_of_data_from_csv]\n",
    "\n",
    "\n",
    "\n",
    "############# Directory ############\n",
    "directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/Mishas_Demultiplexed_LAM'\n",
    "\n",
    "#this is a minimal directory of only 300 files for quick debugging\n",
    "#directory = '/media/edanner/NewUbuntuSpace/Workspace/LinearAmp/Sequence2_191129_MN00157_0047_A000H2GWGF/Mishas_Demultiplexed_LAM/10000Reads'\n",
    "amplicon_info = get_csv_data(directory, 1) \n",
    "\n",
    "print('the directory is:', directory)\n",
    "\n",
    "#########################################\n",
    "# Assign the file_genome_2bit location. This is needed for pulling sequence from the referene genome by location\n",
    "assembly_plus_targetvector = amplicon_info['genome_plus_targeting']\n",
    "file_genome_2bit_plus_target = os.path.join('/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes', assembly_plus_targetvector + '.2bit')\n",
    "print(file_genome_2bit_plus_target)\n",
    "#Needs to have a bowtie build in which the targeting vector is included!!\n",
    "%env BOWTIE2_INDEXES=/media/edanner/NewUbuntuSpace/Workspace/Ref_Genomes\n",
    "#check the env variables\n",
    "#%env\n",
    "################################################\n",
    "\n",
    "\n",
    "seq_to_breaksite = [['GGCCAAGCTGGAGCAGGAAATAGATGCACACGGAGGAAATGGTAGTGGAGTTCAGAGGAGGGAGAGGTTCTTTCTGCCTGGG', 'this is for pipeline1 LAM- polb rev WT N706 N504 (row 0 in sheet)'],\n",
    "                    ['GGCCAAGCTGGAGCAGGAAATAGATGCACACGGAGGAAATGGTAGTGGAGTTCAGAGGAGGGAGAGGTTCTTTCTGCCTGGG','this is for pipeline1 LAM polb rev Targeting 50% N706 N505(row 1 in sheet)'],\n",
    "                    ['TGCATCGCATTGTCTGAGTAGGTGTCATTCTATTCTGGGGGGTGGGGTGGGGCAGGACAGCAAGGGGGAGGATTGGGAAGAGAATAGCAGGCATGCTGGGGGGATCCCTGCAGGAGCTCGTCGACCCATGGGGGCCCGCCCCAACTGGGGTAACCTTTGGGCTCCCCGGGCGCGACTAGTGAATTCAGATCTGATATCTCTAGAAGTCCTGGG','this is for pipeline1 LAM mCherryFwd N707 N505, row2 in sheet'],\n",
    "                    ['CCTCCACGTCACCGCATGTTAGCAGACTTCCTCTGCCCTCAATCTTTAAGAAAAAAAAAAGTCTAACAATGATTTAGGAATGCTTTGAGGACTTAAATGATCTTATTGGAAACATACCAGTCTGCTAAAAGACTAATTTTGTGTGGGTCA', 'this is for mCherryrev LAM  N708 N505 , row3 in sheet']]\n",
    "\n",
    "\n",
    "#0 is LAM polb rev WT N706 N504 (row 0 in sheet)\n",
    "#1 is LAM polb rev Targeting 50% N706 N505(row 1 in sheet)\n",
    "#2 is LAM mCherryFwd N707 N505, row2 in sheet'\n",
    "#3 is LAM mCherryRev   N708 N505 , row3 in sheet'\n",
    "\n",
    "#### important! ####\n",
    "lines_to_run = [2,3]\n",
    "\n",
    "#type in 'on' or 'off'\n",
    "local = 'off'\n",
    "global_align = 'on'\n",
    "trimming = 'off'\n",
    "align_global = 'off'\n",
    "\n",
    "for i in lines_to_run:\n",
    "    \n",
    "    print('running row', i)\n",
    "    seq_primer_to_breaksite = seq_to_breaksite[i][0]\n",
    "    print(seq_to_breaksite[i][1],'\\n')\n",
    "    \n",
    "    # access the information in the csv \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    print('the analysis is on LAM sample:', amplicon_info['name'])\n",
    "    \n",
    "    # trimming the reads to the cas9 break site\n",
    "    \n",
    "    if trimming == 'on':\n",
    "        trim_fastq_to_break(directory, amplicon_info, seq_primer_to_breaksite, lam_or_tn5='lam', length = 35)\n",
    "    \n",
    "    if global_align == 'on':\n",
    "        if align_global == 'on':\n",
    "            #####this aligns everything after the break to end-to-end keep_sam=1 means keep the sam file\n",
    "            align_afterbreak_end_to_end_genome_global(directory, amplicon_info, assembly_plus_targetvector, lam_or_tn5='lam', keep_sam=0)\n",
    "\n",
    "        #### Process the alignment files to make a bam file of only high quality reads\n",
    "        final_trimmed_bam_filtered_mapq_AS_primary(directory, 'global', amplicon_info, 'tn5')\n",
    "        \n",
    "        ### global alignment bed generation \n",
    "        print('-----  now making bed file   ------')\n",
    "\n",
    "        N7 = amplicon_info['index_I1']\n",
    "        N5 = amplicon_info['index_I2']\n",
    "        file_trimmed_genome_global_bed = create_filename(directory, N7, N5, 'break_trimmed_genome_global_bed')\n",
    "        file_sorted_bam_genome_global = create_filename(directory, N7, N5, 'break_trimmed_filtered_and_sorted_genome_global')\n",
    "\n",
    "        !bam2bed < {file_sorted_bam_genome_global} > {file_trimmed_genome_global_bed}\n",
    "        bed_folder = os.path.abspath(create_filename(directory, N7, N5, 'all_bed'))\n",
    "\n",
    "        if not os.path.exists(bed_folder):\n",
    "            os.mkdir(bed_folder)\n",
    "        shutil.copy(file_trimmed_genome_global_bed, bed_folder)\n",
    "\n",
    "        print('made global bed file:', N7, '_', N5 )\n",
    "        print('doing the quantification of global alignments')\n",
    "\n",
    "        # quantifying alignments\n",
    "        quantify_global = quantify_pipeline2_alignments(directory, 'global', amplicon_info, 'lam')\n",
    "        print(quantify_global)\n",
    "    \n",
    "    if local == 'on':\n",
    "        #### this aligns everything in the local format wehre it can soft clip the ends keep_sam=1 means keep the sam file\n",
    "        align_afterbreaks_genome_local(directory, 1, amplicon_info, assembly_plus_targetvector, lam_or_tn5='lam', keep_sam=0)\n",
    "        \n",
    "        #make bed file\n",
    "        print('-----  now making bed file   ------')\n",
    "        N7 = amplicon_info['index_I1']\n",
    "        N5 = amplicon_info['index_I2']\n",
    "        file_trimmed_genome_local_bed = create_filename(directory, N7, N5, 'break_trimmed_genome_local_bed')\n",
    "        file_sorted_bam_genome_local = create_filename(directory, N7, N5, 'break_trimmed_filtered_and_sorted_bam_genome_local')\n",
    "\n",
    "        !bam2bed < {file_sorted_bam_genome_local} > {file_trimmed_genome_local_bed}\n",
    "        bed_folder = os.path.abspath(create_filename(directory, N7, N5, 'all_bed'))\n",
    "\n",
    "        if not os.path.exists(bed_folder):\n",
    "            os.mkdir(bed_folder)\n",
    "        shutil.copy(file_trimmed_genome_local_bed, bed_folder)\n",
    "        print('made local bed file:', N7, '_', N5 )\n",
    "\n",
    "        # running overview of alignment quantification\n",
    "        print('doing the quantification of local alignments')\n",
    "        quantify_local = quantify_pipeline2_alignments(directory, 'local', amplicon_info, 'lam')\n",
    "        print(quantify_local)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## some of the pipeline2 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   alignments_target_site_all  alignments_target_vector_all  \\\n",
      "0                      128841                             1   \n",
      "\n",
      "   alignments_translocations_all  total aligned reads  %_target_site_reads  \\\n",
      "0                            258             129100.0             99.79938   \n",
      "\n",
      "   %_target_vector_reads  %_translocations  \n",
      "0               0.000775          0.199845  \n"
     ]
    }
   ],
   "source": [
    "#polb rev WT\n",
    "print(quantify_pipeline2_alignments(directory, 'global', amplicon_info, 'lam'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   alignments_target_site_all  alignments_target_vector_all  \\\n",
      "0                      200782                        164376   \n",
      "\n",
      "   alignments_translocations_all  total aligned reads  %_target_site_reads  \\\n",
      "0                          18514             383672.0            52.331679   \n",
      "\n",
      "   %_target_vector_reads  %_translocations  \n",
      "0              42.842845          4.825476  \n"
     ]
    }
   ],
   "source": [
    "#polb rev targeted\n",
    "\n",
    "print(quantify_pipeline2_alignments(directory, 'global', amplicon_info, 'lam'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   alignments_target_site_all  alignments_target_vector_all  \\\n",
      "0                      124151                        253853   \n",
      "\n",
      "   alignments_translocations_all  total aligned reads  %_target_site_reads  \\\n",
      "0                         102217             480221.0            25.852889   \n",
      "\n",
      "   %_target_vector_reads  %_translocations  \n",
      "0              52.861703         21.285408  \n",
      "   alignments_target_site_all  alignments_target_vector_all  \\\n",
      "0                      125237                        253041   \n",
      "\n",
      "   alignments_translocations_all  total aligned reads  %_target_site_reads  \\\n",
      "0                         104551             482829.0            25.938169   \n",
      "\n",
      "   %_target_vector_reads  %_translocations  \n",
      "0              52.407995         21.653836  \n"
     ]
    }
   ],
   "source": [
    "#mcherry rev\n",
    "print(quantify_pipeline2_alignments(directory, 'global', amplicon_info, 'lam'))\n",
    "\n",
    "print(quantify_pipeline2_alignments(directory, 'local', amplicon_info, 'lam'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   alignments_target_site_all  alignments_target_vector_all  \\\n",
      "0                       65743                        128529   \n",
      "\n",
      "   alignments_translocations_all  total aligned reads  %_target_site_reads  \\\n",
      "0                          56744             251016.0            26.190761   \n",
      "\n",
      "   %_target_vector_reads  %_translocations  \n",
      "0              51.203509          22.60573  \n",
      "   alignments_target_site_all  alignments_target_vector_all  \\\n",
      "0                       64946                        286877   \n",
      "\n",
      "   alignments_translocations_all  total aligned reads  %_target_site_reads  \\\n",
      "0                          57079             408902.0            15.883023   \n",
      "\n",
      "   %_target_vector_reads  %_translocations  \n",
      "0              70.157886          13.95909  \n"
     ]
    }
   ],
   "source": [
    "#cherry fwd\n",
    "print(quantify_pipeline2_alignments(directory, 'global', amplicon_info, 'lam'))\n",
    "\n",
    "print(quantify_pipeline2_alignments(directory, 'local', amplicon_info, 'lam'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
