{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import pylab\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "import itertools\n",
    "import operator\n",
    "import subprocess\n",
    "import twobitreader\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pysam\n",
    "import shutil\n",
    "\n",
    "from AAVseq.AAVseq_helpersDanner import *\n",
    "from AAVseq.AAVseqDanner import *\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Notes about AAV-Seq analysis\n",
    "\n",
    "- This was done on HSC by Tung. Analysis by me 9/2/21 Targeting of the ELANE, PRF1, or IL7R. ELANE and IL7R used only one gRNA for the Cas9 and 2 guides for the nickase for ELANE and IL7R. Samples were cultured for 18 days. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Danner_AAVseqLogic\n",
    "\n",
    "#  This is a simliar to the guideseq pipeline. Different than the oirignal guideseq pipeline but built in a simliar form. \n",
    "## It checks for the on target priming, trimms off the adapter seq,  uses bowtie2 . Then exports a bed file for comparing to CRISPRgold\n",
    "\n",
    "\n",
    "#### 1. Check reads were correctly primed.\n",
    "#### 2 Trim reads \n",
    "#### 3. Align the reads against the whole geneome with local alignment. \n",
    "#### 4. Alignprocess bam file and pull out the high mapQ reads\n",
    "#### 5. convert to bed\n",
    "#### A3. Do global end-to-end alignment\n",
    "#### A4 global end-to-end\n",
    "#### A5 convert to bed\n",
    "\n",
    "It is important to trim R1 and R2 because if it is short then the read oges onto the chip. Also we will use global and local and compare the results because it is hard to trim to the end of the AAV due ot the random nature of hte integration.\n",
    "\n",
    "\n",
    "\n",
    "We use read1 for checking priming. Also we could use that for mapping (or paired end). On both the 3 and 5' direction the first 34 nucleotides of the read contain the integrated ODN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/eric/Data/Spaced_Nicking/AAVSeq2_ELANE_PRF1_IL7R_2_2021\n"
     ]
    }
   ],
   "source": [
    "#Directory\n",
    "\n",
    "directory = '/home/eric/Data/Spaced_Nicking/AAVSeq2_ELANE_PRF1_IL7R_2_2021'\n",
    "\n",
    "\n",
    "print(directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Discard Mispriming Reads\n",
    "When you put a universal primer on the ends of everything, every mispriming event will amplify. An effect we normally don't deal with. We do a nested PCR to reduce this. However many reads still have mispriming. They did all their blasting and analysis before removing mispriming. But to save computational power and remove error early on I will discard mispriming events. \n",
    "They discard these only for plasmid alignments analyze_alignments_plasmid for some reason which comes from the bam file.\n",
    "\n",
    "### for AAVseq with the MiniSeq machine read1 is where the primer binds the AAV arm, and so that is how you tell targeting. For the reverse it was binidng to the transposon which is randomly integrated, so the sequence strait out of read 2 are random genoimc sequences. If the Rev primer misprimes it shouldnt matter as it would then just bind and make smaller amplicons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with sample 0\n",
      "done with sample 1\n",
      "done with sample 2\n",
      "done with sample 3\n",
      "done with sample 4\n",
      "done with sample 5\n",
      "done with sample 6\n",
      "done with sample 7\n",
      "done with sample 8\n",
      "done with sample 9\n",
      "done with sample 10\n",
      "done with sample 11\n",
      "done with sample 12\n",
      "done with sample 13\n",
      "done with sample 14\n",
      "done with sample 15\n",
      "done with sample 16\n",
      "done with sample 17\n",
      "done with sample 18\n",
      "done with sample 19\n",
      "done with sample 20\n",
      "done with sample 21\n",
      "done with sample 22\n",
      "done with sample 23\n",
      "done with sample 24\n",
      "done with sample 25\n",
      "done with sample 26\n",
      "                sample_name    i7    i5  total_reads  reads_with_good_priming  \\\n",
      "0               1_AAV_ELA_1  N701  S502       150142                    44983   \n",
      "1               1_AAV_ELA_2  N701  S503       131739                    57533   \n",
      "2               1_AAV_ELA_3  N701  S505       114944                    25066   \n",
      "3        2_Cas9sg5AAV_ELA_1  N702  S502       185742                   132868   \n",
      "4        2_Cas9sg5AAV_ELA_2  N702  S503       125404                    84620   \n",
      "5        2_Cas9sg5AAV_ELA_3  N702  S505       110031                    85139   \n",
      "6   3_D10Asg5_303_AAV_ELA_1  N703  S502       202608                   122049   \n",
      "7   3_D10Asg5_303_AAV_ELA_2  N703  S503       519147                   395837   \n",
      "8   3_D10Asg5_303_AAV_ELA_3  N703  S505       458879                   331827   \n",
      "9              4_AAV_PRF1_1  N704  S502       291105                   202597   \n",
      "10             4_AAV_PRF1_2  N704  S503       272063                   170080   \n",
      "11             4_AAV_PRF1_3  N704  S505       394952                   209346   \n",
      "12     5_Cas9RNP_AAV_PRF1_1  N705  S502       127388                    87441   \n",
      "13     5_Cas9RNP_AAV_PRF1_2  N705  S503       452344                   361235   \n",
      "14     5_Cas9RNP_AAV_PRF1_3  N705  S505       489009                   395759   \n",
      "15     6_D10ARNP_AAV_PRF1_1  N706  S502       136922                   102512   \n",
      "16     6_D10ARNP_AAV_PRF1_2  N706  S503       157403                   127004   \n",
      "17     6_D10ARNP_AAV_PRF1_3  N706  S505       211278                   159007   \n",
      "18             7_AAV_IL7R_1  N707  S502       123510                    93798   \n",
      "19             7_AAV_IL7R_2  N707  S503       296220                   211693   \n",
      "20             7_AAV_IL7R_3  N707  S505       380509                   293155   \n",
      "21     8_Cas9RNP_AAV_IL7R_1  N710  S502       178793                   145675   \n",
      "22     8_Cas9RNP_AAV_IL7R_2  N710  S503       182746                   138581   \n",
      "23     8_Cas9RNP_AAV_IL7R_3  N710  S505       205055                   162085   \n",
      "24     9_D10ARNP_AAV_IL7R_1  N711  S502       136337                    99509   \n",
      "25     9_D10ARNP_AAV_IL7R_2  N711  S503        69169                    50101   \n",
      "26     9_D10ARNP_AAV_IL7R_3  N711  S505       215735                   163012   \n",
      "\n",
      "    reads_with_guideseq_primer_misprimed  \\\n",
      "0                                   2511   \n",
      "1                                   3591   \n",
      "2                                   3472   \n",
      "3                                   8426   \n",
      "4                                  14635   \n",
      "5                                   6055   \n",
      "6                                   8956   \n",
      "7                                  37457   \n",
      "8                                  22007   \n",
      "9                                  19824   \n",
      "10                                 12061   \n",
      "11                                 11553   \n",
      "12                                 20999   \n",
      "13                                 45777   \n",
      "14                                 37301   \n",
      "15                                 13892   \n",
      "16                                  8605   \n",
      "17                                 24002   \n",
      "18                                  7257   \n",
      "19                                 13698   \n",
      "20                                 27131   \n",
      "21                                  7159   \n",
      "22                                 17757   \n",
      "23                                 12494   \n",
      "24                                  7694   \n",
      "25                                  3049   \n",
      "26                                 16291   \n",
      "\n",
      "    reads_with_indexing_primer_mispriming  \n",
      "0                                  102648  \n",
      "1                                   70615  \n",
      "2                                   86406  \n",
      "3                                   44448  \n",
      "4                                   26149  \n",
      "5                                   18837  \n",
      "6                                   71603  \n",
      "7                                   85853  \n",
      "8                                  105045  \n",
      "9                                   68684  \n",
      "10                                  89922  \n",
      "11                                 174053  \n",
      "12                                  18948  \n",
      "13                                  45332  \n",
      "14                                  55949  \n",
      "15                                  20518  \n",
      "16                                  21794  \n",
      "17                                  28269  \n",
      "18                                  22455  \n",
      "19                                  70829  \n",
      "20                                  60223  \n",
      "21                                  25959  \n",
      "22                                  26408  \n",
      "23                                  30476  \n",
      "24                                  29134  \n",
      "25                                  16019  \n",
      "26                                  36432  \n"
     ]
    }
   ],
   "source": [
    "############### Check for Mispriming ##################\n",
    "#\n",
    "\n",
    "\n",
    "#the amplicon info is related to the line on the csv file. It is indexed from 0.\n",
    "#it is on my extera space so need to make sure that is mounted. (first thing to check if it throws and error)\n",
    "\n",
    "\n",
    "# THIS IS WHAT RUNS THE PRIMER CHECK THROUGH ALL SAMPLES\n",
    "#input the sample range or list below\n",
    "\n",
    "\n",
    "results_df_all = pd.DataFrame()\n",
    "\n",
    "results_folder = os.path.join(directory, 'results')\n",
    "if not os.path.exists(results_folder):\n",
    "    os.mkdir(results_folder)\n",
    "results_file = os.path.join(directory, 'results','all_priming.xlsx')\n",
    "    \n",
    "\n",
    "#make results tree\n",
    "# 15 samples\n",
    "\n",
    "for i in range(27):\n",
    "        \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "\n",
    "  \n",
    "    \n",
    "    # the primer binds to one of the sides of the AAV. \n",
    "    primer_seq_plus_downstream = 'TCTCTGCGCGCTCGCTCGCTCACTGA' #6 nt downstream the primer\n",
    "    primer_seq = 'TCTCTGCGCGCTCGCTCG' #sequence binding portion\n",
    "    \n",
    "    df_sample_results = correct_priming_guideseq(directory, amplicon_info, primer_seq, primer_seq_plus_downstream)\n",
    "    results_df_all = results_df_all.append(df_sample_results, ignore_index=True)\n",
    "    print('done with sample', i)\n",
    "\n",
    "print(results_df_all)\n",
    "results_df_all.to_excel(results_file)    \n",
    "#print reults to excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with sample 0\n",
      "done with sample 1\n",
      "done with sample 2\n",
      "done with sample 3\n",
      "done with sample 4\n",
      "done with sample 5\n",
      "done with sample 6\n",
      "done with sample 7\n",
      "done with sample 8\n",
      "done with sample 9\n",
      "done with sample 10\n",
      "done with sample 11\n",
      "done with sample 12\n",
      "done with sample 13\n",
      "done with sample 14\n",
      "done with sample 15\n",
      "done with sample 16\n",
      "done with sample 17\n",
      "done with sample 18\n",
      "done with sample 19\n",
      "done with sample 20\n",
      "done with sample 21\n",
      "done with sample 22\n",
      "done with sample 23\n",
      "done with sample 24\n",
      "done with sample 25\n",
      "done with sample 26\n"
     ]
    }
   ],
   "source": [
    "############ TRIMMING adapter sequences off short reads #########\n",
    "\n",
    "\n",
    "#need to trim off the end of the short reads. This is for amplicons that were too short and have the other side on them.\n",
    "# expect more trimmed read 2 and than read 1. Read 2 is longer htan read one as we trim off the ~30 bases of good priming \n",
    "# form read1 so sort amplicons get trimmed for read2 before being trimmed for read1.\n",
    "\n",
    "for i in range(27):\n",
    "        \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    \n",
    "    trim_AAVseq(directory, amplicon_info)\n",
    "    \n",
    "    print('done with sample', i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: BOWTIE2_INDEXES=/home/eric/Data/Ref_Genomes\n"
     ]
    }
   ],
   "source": [
    "# BOWTIE2_INDEXES are needed for global alignments\n",
    "#check in bash: > ECHO $GENOMES_2BIT\n",
    "%env BOWTIE2_INDEXES=/home/eric/Data/Ref_Genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Need to make an new bowtie2 index file that includes targeting for alignment agains the whole genome so it is all in one sheet together. \n",
    "\n",
    "### Other option would be to align it to the amplicon, extract unaligned files and then align to the genome but seems cleaner this way. \n",
    "#### Ideally every sequence is unique between the targeting vector and genome.\n",
    "\n",
    "\n",
    "1. Build your fastas of interest and label .fa files.\n",
    "    1. You need fasta of hg38 or reference genome. You can pull this from downloaded bowtie indexed sampels and then use the following command to turn the index into a fasta file: bowtie2-inspect hg38 > hg38.fa   \n",
    "    2. Put all the fasta files in the same folder. Should also use the transfected plasmid\n",
    "2. index the files with bowtie\n",
    "    1. use the command bowtie2-build -f pE049,pe038_mc.fa,hg38.fa -p hg38_plus_targetvectorandplasmid\n",
    "    2. this has the -p to make it take less ram in my case.\n",
    "    3. In this case it adds the hg38 and the minicircle targeting file together\n",
    "    4. I have a Intel® Core™ i7-5500U CPU @ 2.40GHz × 4 with 15.1 GiB ram and it required about 13.8 gigs of ram and 2 hours to do a hg38+small fasta index\n",
    "    5. be sure to pay attention to the name of the new indexed file. \"hg38_plus_targetvector\" in example above. Add it to the sample_info.csv sheet. Under the tab 'genome_plus_targeting'.\n",
    "    6. you can check it indexed correcctly: bowtie2-inspect -s hg38_plus_targetvector\n",
    "    \n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now we blast the correctly primed sequences.\n",
    "We don't need to trim off anything. But we should remove the length of the ODN primer sequence before blasting for Read1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "running row 0\n",
      "the analysis is on AAV-seq sample: 1_AAV_ELA_1\n",
      "wooo\n",
      "single por paired paired\n",
      "   sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  1_AAV_ELA_1  N701  S502                 paired                 37576   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      16368  \n",
      "-----  now making local bed file for 1_AAV_ELA_1   N701   S502\n",
      "made local bed file: N701 _ S502\n",
      "It is called 1_AAV_ELA_1\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 1\n",
      "the analysis is on AAV-seq sample: 1_AAV_ELA_2\n",
      "wooo\n",
      "single por paired paired\n",
      "   sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  1_AAV_ELA_2  N701  S503                 paired                138901   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      38424  \n",
      "-----  now making local bed file for 1_AAV_ELA_2   N701   S503\n",
      "made local bed file: N701 _ S503\n",
      "It is called 1_AAV_ELA_2\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 2\n",
      "the analysis is on AAV-seq sample: 1_AAV_ELA_3\n",
      "wooo\n",
      "single por paired paired\n",
      "   sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  1_AAV_ELA_3  N701  S505                 paired                  1899   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                        117  \n",
      "-----  now making local bed file for 1_AAV_ELA_3   N701   S505\n",
      "made local bed file: N701 _ S505\n",
      "It is called 1_AAV_ELA_3\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 3\n",
      "the analysis is on AAV-seq sample: 2_Cas9sg5AAV_ELA_1\n",
      "wooo\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  2_Cas9sg5AAV_ELA_1  N702  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                237168                     136556  \n",
      "-----  now making local bed file for 2_Cas9sg5AAV_ELA_1   N702   S502\n",
      "made local bed file: N702 _ S502\n",
      "It is called 2_Cas9sg5AAV_ELA_1\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 4\n",
      "the analysis is on AAV-seq sample: 2_Cas9sg5AAV_ELA_2\n",
      "wooo\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  2_Cas9sg5AAV_ELA_2  N702  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                227389                      77149  \n",
      "-----  now making local bed file for 2_Cas9sg5AAV_ELA_2   N702   S503\n",
      "made local bed file: N702 _ S503\n",
      "It is called 2_Cas9sg5AAV_ELA_2\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 5\n",
      "the analysis is on AAV-seq sample: 2_Cas9sg5AAV_ELA_3\n",
      "wooo\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  2_Cas9sg5AAV_ELA_3  N702  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                148715                      97405  \n",
      "-----  now making local bed file for 2_Cas9sg5AAV_ELA_3   N702   S505\n",
      "made local bed file: N702 _ S505\n",
      "It is called 2_Cas9sg5AAV_ELA_3\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 6\n",
      "the analysis is on AAV-seq sample: 3_D10Asg5_303_AAV_ELA_1\n",
      "wooo\n",
      "single por paired paired\n",
      "               sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3_D10Asg5_303_AAV_ELA_1  N703  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                207157                      33906  \n",
      "-----  now making local bed file for 3_D10Asg5_303_AAV_ELA_1   N703   S502\n",
      "made local bed file: N703 _ S502\n",
      "It is called 3_D10Asg5_303_AAV_ELA_1\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 7\n",
      "the analysis is on AAV-seq sample: 3_D10Asg5_303_AAV_ELA_2\n",
      "wooo\n",
      "single por paired paired\n",
      "               sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3_D10Asg5_303_AAV_ELA_2  N703  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                393806                     225124  \n",
      "-----  now making local bed file for 3_D10Asg5_303_AAV_ELA_2   N703   S503\n",
      "made local bed file: N703 _ S503\n",
      "It is called 3_D10Asg5_303_AAV_ELA_2\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 8\n",
      "the analysis is on AAV-seq sample: 3_D10Asg5_303_AAV_ELA_3\n",
      "wooo\n",
      "single por paired paired\n",
      "               sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3_D10Asg5_303_AAV_ELA_3  N703  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                569604                     202659  \n",
      "-----  now making local bed file for 3_D10Asg5_303_AAV_ELA_3   N703   S505\n",
      "made local bed file: N703 _ S505\n",
      "It is called 3_D10Asg5_303_AAV_ELA_3\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 9\n",
      "the analysis is on AAV-seq sample: 4_AAV_PRF1_1\n",
      "wooo\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  4_AAV_PRF1_1  N704  S502                 paired                112229   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      39766  \n",
      "-----  now making local bed file for 4_AAV_PRF1_1   N704   S502\n",
      "made local bed file: N704 _ S502\n",
      "It is called 4_AAV_PRF1_1\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 10\n",
      "the analysis is on AAV-seq sample: 4_AAV_PRF1_2\n",
      "wooo\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  4_AAV_PRF1_2  N704  S503                 paired                 97069   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      56981  \n",
      "-----  now making local bed file for 4_AAV_PRF1_2   N704   S503\n",
      "made local bed file: N704 _ S503\n",
      "It is called 4_AAV_PRF1_2\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 11\n",
      "the analysis is on AAV-seq sample: 4_AAV_PRF1_3\n",
      "wooo\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  4_AAV_PRF1_3  N704  S505                 paired                156285   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                     116861  \n",
      "-----  now making local bed file for 4_AAV_PRF1_3   N704   S505\n",
      "made local bed file: N704 _ S505\n",
      "It is called 4_AAV_PRF1_3\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 12\n",
      "the analysis is on AAV-seq sample: 5_Cas9RNP_AAV_PRF1_1\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5_Cas9RNP_AAV_PRF1_1  N705  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                195494                      69025  \n",
      "-----  now making local bed file for 5_Cas9RNP_AAV_PRF1_1   N705   S502\n",
      "made local bed file: N705 _ S502\n",
      "It is called 5_Cas9RNP_AAV_PRF1_1\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 13\n",
      "the analysis is on AAV-seq sample: 5_Cas9RNP_AAV_PRF1_2\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5_Cas9RNP_AAV_PRF1_2  N705  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                307875                     208803  \n",
      "-----  now making local bed file for 5_Cas9RNP_AAV_PRF1_2   N705   S503\n",
      "made local bed file: N705 _ S503\n",
      "It is called 5_Cas9RNP_AAV_PRF1_2\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 14\n",
      "the analysis is on AAV-seq sample: 5_Cas9RNP_AAV_PRF1_3\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5_Cas9RNP_AAV_PRF1_3  N705  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                437632                     188396  \n",
      "-----  now making local bed file for 5_Cas9RNP_AAV_PRF1_3   N705   S505\n",
      "made local bed file: N705 _ S505\n",
      "It is called 5_Cas9RNP_AAV_PRF1_3\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 15\n",
      "the analysis is on AAV-seq sample: 6_D10ARNP_AAV_PRF1_1\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  6_D10ARNP_AAV_PRF1_1  N706  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                101628                      66237  \n",
      "-----  now making local bed file for 6_D10ARNP_AAV_PRF1_1   N706   S502\n",
      "made local bed file: N706 _ S502\n",
      "It is called 6_D10ARNP_AAV_PRF1_1\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 16\n",
      "the analysis is on AAV-seq sample: 6_D10ARNP_AAV_PRF1_2\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  6_D10ARNP_AAV_PRF1_2  N706  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                168302                     123982  \n",
      "-----  now making local bed file for 6_D10ARNP_AAV_PRF1_2   N706   S503\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "made local bed file: N706 _ S503\n",
      "It is called 6_D10ARNP_AAV_PRF1_2\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 17\n",
      "the analysis is on AAV-seq sample: 6_D10ARNP_AAV_PRF1_3\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  6_D10ARNP_AAV_PRF1_3  N706  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                268887                     120786  \n",
      "-----  now making local bed file for 6_D10ARNP_AAV_PRF1_3   N706   S505\n",
      "made local bed file: N706 _ S505\n",
      "It is called 6_D10ARNP_AAV_PRF1_3\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 18\n",
      "the analysis is on AAV-seq sample: 7_AAV_IL7R_1\n",
      "wooo\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  7_AAV_IL7R_1  N707  S502                 paired                 60392   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      27199  \n",
      "-----  now making local bed file for 7_AAV_IL7R_1   N707   S502\n",
      "made local bed file: N707 _ S502\n",
      "It is called 7_AAV_IL7R_1\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 19\n",
      "the analysis is on AAV-seq sample: 7_AAV_IL7R_2\n",
      "wooo\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  7_AAV_IL7R_2  N707  S503                 paired                166856   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      36846  \n",
      "-----  now making local bed file for 7_AAV_IL7R_2   N707   S503\n",
      "made local bed file: N707 _ S503\n",
      "It is called 7_AAV_IL7R_2\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 20\n",
      "the analysis is on AAV-seq sample: 7_AAV_IL7R_3\n",
      "wooo\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  7_AAV_IL7R_3  N707  S505                 paired                125411   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      59131  \n",
      "-----  now making local bed file for 7_AAV_IL7R_3   N707   S505\n",
      "made local bed file: N707 _ S505\n",
      "It is called 7_AAV_IL7R_3\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 21\n",
      "the analysis is on AAV-seq sample: 8_Cas9RNP_AAV_IL7R_1\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  8_Cas9RNP_AAV_IL7R_1  N710  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                320505                     185497  \n",
      "-----  now making local bed file for 8_Cas9RNP_AAV_IL7R_1   N710   S502\n",
      "made local bed file: N710 _ S502\n",
      "It is called 8_Cas9RNP_AAV_IL7R_1\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 22\n",
      "the analysis is on AAV-seq sample: 8_Cas9RNP_AAV_IL7R_2\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  8_Cas9RNP_AAV_IL7R_2  N710  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                150088                     120382  \n",
      "-----  now making local bed file for 8_Cas9RNP_AAV_IL7R_2   N710   S503\n",
      "made local bed file: N710 _ S503\n",
      "It is called 8_Cas9RNP_AAV_IL7R_2\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 23\n",
      "the analysis is on AAV-seq sample: 8_Cas9RNP_AAV_IL7R_3\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  8_Cas9RNP_AAV_IL7R_3  N710  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                216090                     171265  \n",
      "-----  now making local bed file for 8_Cas9RNP_AAV_IL7R_3   N710   S505\n",
      "made local bed file: N710 _ S505\n",
      "It is called 8_Cas9RNP_AAV_IL7R_3\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 24\n",
      "the analysis is on AAV-seq sample: 9_D10ARNP_AAV_IL7R_1\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  9_D10ARNP_AAV_IL7R_1  N711  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                163956                      51428  \n",
      "-----  now making local bed file for 9_D10ARNP_AAV_IL7R_1   N711   S502\n",
      "made local bed file: N711 _ S502\n",
      "It is called 9_D10ARNP_AAV_IL7R_1\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 25\n",
      "the analysis is on AAV-seq sample: 9_D10ARNP_AAV_IL7R_2\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  9_D10ARNP_AAV_IL7R_2  N711  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 44926                      41820  \n",
      "-----  now making local bed file for 9_D10ARNP_AAV_IL7R_2   N711   S503\n",
      "made local bed file: N711 _ S503\n",
      "It is called 9_D10ARNP_AAV_IL7R_2\n",
      "doing the quantification of local alignments\n",
      " \n",
      "running row 26\n",
      "the analysis is on AAV-seq sample: 9_D10ARNP_AAV_IL7R_3\n",
      "wooo\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  9_D10ARNP_AAV_IL7R_3  N711  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                268388                     132323  \n",
      "-----  now making local bed file for 9_D10ARNP_AAV_IL7R_3   N711   S505\n",
      "made local bed file: N711 _ S505\n",
      "It is called 9_D10ARNP_AAV_IL7R_3\n",
      "doing the quantification of local alignments\n"
     ]
    }
   ],
   "source": [
    "###  THIS IS FOR BLASTING THE LOCAL ALIGNMENT. WE WILL HAVE LIGATED AAV ON END OF READ AND SO WE SHOULD IGNORE  THIS\n",
    "#\n",
    "#     This section does two things. It makes a bam file that is filtered for AS and MAPQ \n",
    "#     It quantifies these bam files and sorts them. Then it converts to a bed file.\n",
    "#\n",
    "\n",
    "                #### Required Arguments\n",
    "cpu = 14\n",
    "min_MAPQ = 50\n",
    "min_AS = -180\n",
    "\n",
    "\n",
    "                        #this does the acutal alignment for end-to-end\n",
    "do_align = 'off'\n",
    "single_or_paired_reads = \"paired\"     #can be 'single' or 'paired' for the read alignment\n",
    "\n",
    "\n",
    "                    #### WHAT SAMPLES TO RUN ####\n",
    "\n",
    "lines_to_run = range(27)\n",
    "\n",
    "                    #for exporting summary of data\n",
    "results_df_all = pd.DataFrame()\n",
    "results_file = os.path.join(directory, 'results','local_alignment_filtered_summary.xlsx')\n",
    "\n",
    "\n",
    "for i in lines_to_run:\n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    print(' ')\n",
    "    print('running row', i)\n",
    "    print('the analysis is on AAV-seq sample:', amplicon_info['name'])\n",
    "    reference_assembly = amplicon_info['genome']\n",
    "    sample_name = amplicon_info['name']\n",
    "    \n",
    "    if do_align == 'on':\n",
    "        #####this aligns everything after the break to end-to-end keep_sam=1 means keep the sam file\n",
    "        align_guideseq_genome_local(directory, amplicon_info, reference_assembly, single_or_paired_reads, cpu, keep_sam=0)\n",
    "\n",
    "\n",
    "    #### Process the alignment files to make a bam file of only high quality reads\n",
    "\n",
    "    #this filters for good alignments based on AS score and MapQ and exports a final indexed bam file.\n",
    "\n",
    "    df_sample_results = guideseq_filtered_mapq_AS_primary(directory, 'local', amplicon_info, single_or_paired_reads)\n",
    "    \n",
    "    #this is a running dataframe of the clenaed up summary of filtering information\n",
    "    results_df_all = results_df_all.append(df_sample_results, ignore_index=True)\n",
    "\n",
    "\n",
    "    ### local alignment bed generation \n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "    print('-----  now making local bed file for', sample_name, ' ',N7, ' ', N5 )\n",
    "\n",
    "\n",
    "    if single_or_paired_reads == \"single\":\n",
    "        genome_local_bed_file = create_filename(directory, N7, N5, 'guideseq_local_bed_single')\n",
    "        filtered_and_sorted_bam_file = create_filename(directory, N7, N5, 'filtered_and_sorted_genome_local_single')\n",
    "    elif single_or_paired_reads == \"paired\":\n",
    "        genome_local_bed_file = create_filename(directory, N7, N5, 'guideseq_local_bed_paired')\n",
    "        filtered_and_sorted_bam_file = create_filename(directory, N7, N5, 'filtered_and_sorted_genome_local_paired')\n",
    "   \n",
    "    !bam2bed < {filtered_and_sorted_bam_file} > {genome_local_bed_file}\n",
    "    bed_folder = os.path.join(directory, 'bed_files')\n",
    "    \n",
    "    \n",
    "    if not os.path.exists(bed_folder):\n",
    "        os.mkdir(bed_folder)\n",
    "\n",
    "    file_location = create_filename(directory, N7, N5, 'bam_local')\n",
    "\n",
    "    new_bed_name = os.path.join(file_location, sample_name + '_' + 'local_'+ single_or_paired_reads + '.bed')\n",
    "    \n",
    "    os.rename(genome_local_bed_file, new_bed_name)\n",
    "\n",
    "    shutil.copy(new_bed_name, bed_folder)\n",
    "\n",
    "    #remove the original bed file in the bam folder\n",
    "    os.remove(new_bed_name)\n",
    "\n",
    "\n",
    "    print('made local bed file:', N7, '_', N5 )\n",
    "    print('It is called', sample_name )\n",
    "    print('doing the quantification of local alignments')\n",
    "\n",
    " \n",
    "    \n",
    " \n",
    " ## export all the read summaries\n",
    "results_df_all.to_excel(results_file)    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running row 0\n",
      "the analysis is on AAV-seq sample: 1_AAV_ELA_1\n",
      "single por paired paired\n",
      "   sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  1_AAV_ELA_1  N701  S502                 paired                 16484   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      15972  \n",
      "-----  now making bed file for 1_AAV_ELA_1   N701   S502\n",
      "made global bed file: N701 _ S502\n",
      "It is called 1_AAV_ELA_1\n",
      "doing the quantification of global alignments\n",
      "running row 1\n",
      "the analysis is on AAV-seq sample: 1_AAV_ELA_2\n",
      "single por paired paired\n",
      "   sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  1_AAV_ELA_2  N701  S503                 paired                 38874   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      38485  \n",
      "-----  now making bed file for 1_AAV_ELA_2   N701   S503\n",
      "made global bed file: N701 _ S503\n",
      "It is called 1_AAV_ELA_2\n",
      "doing the quantification of global alignments\n",
      "running row 2\n",
      "the analysis is on AAV-seq sample: 1_AAV_ELA_3\n",
      "single por paired paired\n",
      "   sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  1_AAV_ELA_3  N701  S505                 paired                    42   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                         28  \n",
      "-----  now making bed file for 1_AAV_ELA_3   N701   S505\n",
      "made global bed file: N701 _ S505\n",
      "It is called 1_AAV_ELA_3\n",
      "doing the quantification of global alignments\n",
      "running row 3\n",
      "the analysis is on AAV-seq sample: 2_Cas9sg5AAV_ELA_1\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  2_Cas9sg5AAV_ELA_1  N702  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                122563                      90040  \n",
      "-----  now making bed file for 2_Cas9sg5AAV_ELA_1   N702   S502\n",
      "made global bed file: N702 _ S502\n",
      "It is called 2_Cas9sg5AAV_ELA_1\n",
      "doing the quantification of global alignments\n",
      "running row 4\n",
      "the analysis is on AAV-seq sample: 2_Cas9sg5AAV_ELA_2\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  2_Cas9sg5AAV_ELA_2  N702  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                113859                      46493  \n",
      "-----  now making bed file for 2_Cas9sg5AAV_ELA_2   N702   S503\n",
      "made global bed file: N702 _ S503\n",
      "It is called 2_Cas9sg5AAV_ELA_2\n",
      "doing the quantification of global alignments\n",
      "running row 5\n",
      "the analysis is on AAV-seq sample: 2_Cas9sg5AAV_ELA_3\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  2_Cas9sg5AAV_ELA_3  N702  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 92348                      66384  \n",
      "-----  now making bed file for 2_Cas9sg5AAV_ELA_3   N702   S505\n",
      "made global bed file: N702 _ S505\n",
      "It is called 2_Cas9sg5AAV_ELA_3\n",
      "doing the quantification of global alignments\n",
      "running row 6\n",
      "the analysis is on AAV-seq sample: 3_D10Asg5_303_AAV_ELA_1\n",
      "single por paired paired\n",
      "               sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3_D10Asg5_303_AAV_ELA_1  N703  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 44237                      38112  \n",
      "-----  now making bed file for 3_D10Asg5_303_AAV_ELA_1   N703   S502\n",
      "made global bed file: N703 _ S502\n",
      "It is called 3_D10Asg5_303_AAV_ELA_1\n",
      "doing the quantification of global alignments\n",
      "running row 7\n",
      "the analysis is on AAV-seq sample: 3_D10Asg5_303_AAV_ELA_2\n",
      "single por paired paired\n",
      "               sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3_D10Asg5_303_AAV_ELA_2  N703  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                211261                     193303  \n",
      "-----  now making bed file for 3_D10Asg5_303_AAV_ELA_2   N703   S503\n",
      "made global bed file: N703 _ S503\n",
      "It is called 3_D10Asg5_303_AAV_ELA_2\n",
      "doing the quantification of global alignments\n",
      "running row 8\n",
      "the analysis is on AAV-seq sample: 3_D10Asg5_303_AAV_ELA_3\n",
      "single por paired paired\n",
      "               sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3_D10Asg5_303_AAV_ELA_3  N703  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                143445                     136546  \n",
      "-----  now making bed file for 3_D10Asg5_303_AAV_ELA_3   N703   S505\n",
      "made global bed file: N703 _ S505\n",
      "It is called 3_D10Asg5_303_AAV_ELA_3\n",
      "doing the quantification of global alignments\n",
      "running row 9\n",
      "the analysis is on AAV-seq sample: 4_AAV_PRF1_1\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  4_AAV_PRF1_1  N704  S502                 paired                103568   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      37787  \n",
      "-----  now making bed file for 4_AAV_PRF1_1   N704   S502\n",
      "made global bed file: N704 _ S502\n",
      "It is called 4_AAV_PRF1_1\n",
      "doing the quantification of global alignments\n",
      "running row 10\n",
      "the analysis is on AAV-seq sample: 4_AAV_PRF1_2\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  4_AAV_PRF1_2  N704  S503                 paired                 19894   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      13285  \n",
      "-----  now making bed file for 4_AAV_PRF1_2   N704   S503\n",
      "made global bed file: N704 _ S503\n",
      "It is called 4_AAV_PRF1_2\n",
      "doing the quantification of global alignments\n",
      "running row 11\n",
      "the analysis is on AAV-seq sample: 4_AAV_PRF1_3\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  4_AAV_PRF1_3  N704  S505                 paired                116794   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                     115326  \n",
      "-----  now making bed file for 4_AAV_PRF1_3   N704   S505\n",
      "made global bed file: N704 _ S505\n",
      "It is called 4_AAV_PRF1_3\n",
      "doing the quantification of global alignments\n",
      "running row 12\n",
      "the analysis is on AAV-seq sample: 5_Cas9RNP_AAV_PRF1_1\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5_Cas9RNP_AAV_PRF1_1  N705  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 86635                      51570  \n",
      "-----  now making bed file for 5_Cas9RNP_AAV_PRF1_1   N705   S502\n",
      "made global bed file: N705 _ S502\n",
      "It is called 5_Cas9RNP_AAV_PRF1_1\n",
      "doing the quantification of global alignments\n",
      "running row 13\n",
      "the analysis is on AAV-seq sample: 5_Cas9RNP_AAV_PRF1_2\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5_Cas9RNP_AAV_PRF1_2  N705  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                180007                     143739  \n",
      "-----  now making bed file for 5_Cas9RNP_AAV_PRF1_2   N705   S503\n",
      "made global bed file: N705 _ S503\n",
      "It is called 5_Cas9RNP_AAV_PRF1_2\n",
      "doing the quantification of global alignments\n",
      "running row 14\n",
      "the analysis is on AAV-seq sample: 5_Cas9RNP_AAV_PRF1_3\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5_Cas9RNP_AAV_PRF1_3  N705  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                288284                     106795  \n",
      "-----  now making bed file for 5_Cas9RNP_AAV_PRF1_3   N705   S505\n",
      "made global bed file: N705 _ S505\n",
      "It is called 5_Cas9RNP_AAV_PRF1_3\n",
      "doing the quantification of global alignments\n",
      "running row 15\n",
      "the analysis is on AAV-seq sample: 6_D10ARNP_AAV_PRF1_1\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  6_D10ARNP_AAV_PRF1_1  N706  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 54731                      39268  \n",
      "-----  now making bed file for 6_D10ARNP_AAV_PRF1_1   N706   S502\n",
      "made global bed file: N706 _ S502\n",
      "It is called 6_D10ARNP_AAV_PRF1_1\n",
      "doing the quantification of global alignments\n",
      "running row 16\n",
      "the analysis is on AAV-seq sample: 6_D10ARNP_AAV_PRF1_2\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  6_D10ARNP_AAV_PRF1_2  N706  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 99435                      78877  \n",
      "-----  now making bed file for 6_D10ARNP_AAV_PRF1_2   N706   S503\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "made global bed file: N706 _ S503\n",
      "It is called 6_D10ARNP_AAV_PRF1_2\n",
      "doing the quantification of global alignments\n",
      "running row 17\n",
      "the analysis is on AAV-seq sample: 6_D10ARNP_AAV_PRF1_3\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  6_D10ARNP_AAV_PRF1_3  N706  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                124300                      93945  \n",
      "-----  now making bed file for 6_D10ARNP_AAV_PRF1_3   N706   S505\n",
      "made global bed file: N706 _ S505\n",
      "It is called 6_D10ARNP_AAV_PRF1_3\n",
      "doing the quantification of global alignments\n",
      "running row 18\n",
      "the analysis is on AAV-seq sample: 7_AAV_IL7R_1\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  7_AAV_IL7R_1  N707  S502                 paired                 24022   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      19912  \n",
      "-----  now making bed file for 7_AAV_IL7R_1   N707   S502\n",
      "made global bed file: N707 _ S502\n",
      "It is called 7_AAV_IL7R_1\n",
      "doing the quantification of global alignments\n",
      "running row 19\n",
      "the analysis is on AAV-seq sample: 7_AAV_IL7R_2\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  7_AAV_IL7R_2  N707  S503                 paired                 53717   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      44169  \n",
      "-----  now making bed file for 7_AAV_IL7R_2   N707   S503\n",
      "made global bed file: N707 _ S503\n",
      "It is called 7_AAV_IL7R_2\n",
      "doing the quantification of global alignments\n",
      "running row 20\n",
      "the analysis is on AAV-seq sample: 7_AAV_IL7R_3\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  7_AAV_IL7R_3  N707  S505                 paired                 31050   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      28089  \n",
      "-----  now making bed file for 7_AAV_IL7R_3   N707   S505\n",
      "made global bed file: N707 _ S505\n",
      "It is called 7_AAV_IL7R_3\n",
      "doing the quantification of global alignments\n",
      "running row 21\n",
      "the analysis is on AAV-seq sample: 8_Cas9RNP_AAV_IL7R_1\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  8_Cas9RNP_AAV_IL7R_1  N710  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                183888                     119223  \n",
      "-----  now making bed file for 8_Cas9RNP_AAV_IL7R_1   N710   S502\n",
      "made global bed file: N710 _ S502\n",
      "It is called 8_Cas9RNP_AAV_IL7R_1\n",
      "doing the quantification of global alignments\n",
      "running row 22\n",
      "the analysis is on AAV-seq sample: 8_Cas9RNP_AAV_IL7R_2\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  8_Cas9RNP_AAV_IL7R_2  N710  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                105539                      97666  \n",
      "-----  now making bed file for 8_Cas9RNP_AAV_IL7R_2   N710   S503\n",
      "made global bed file: N710 _ S503\n",
      "It is called 8_Cas9RNP_AAV_IL7R_2\n",
      "doing the quantification of global alignments\n",
      "running row 23\n",
      "the analysis is on AAV-seq sample: 8_Cas9RNP_AAV_IL7R_3\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  8_Cas9RNP_AAV_IL7R_3  N710  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                130023                     119126  \n",
      "-----  now making bed file for 8_Cas9RNP_AAV_IL7R_3   N710   S505\n",
      "made global bed file: N710 _ S505\n",
      "It is called 8_Cas9RNP_AAV_IL7R_3\n",
      "doing the quantification of global alignments\n",
      "running row 24\n",
      "the analysis is on AAV-seq sample: 9_D10ARNP_AAV_IL7R_1\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  9_D10ARNP_AAV_IL7R_1  N711  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 72027                      49682  \n",
      "-----  now making bed file for 9_D10ARNP_AAV_IL7R_1   N711   S502\n",
      "made global bed file: N711 _ S502\n",
      "It is called 9_D10ARNP_AAV_IL7R_1\n",
      "doing the quantification of global alignments\n",
      "running row 25\n",
      "the analysis is on AAV-seq sample: 9_D10ARNP_AAV_IL7R_2\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  9_D10ARNP_AAV_IL7R_2  N711  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 23565                      20943  \n",
      "-----  now making bed file for 9_D10ARNP_AAV_IL7R_2   N711   S503\n",
      "made global bed file: N711 _ S503\n",
      "It is called 9_D10ARNP_AAV_IL7R_2\n",
      "doing the quantification of global alignments\n",
      "running row 26\n",
      "the analysis is on AAV-seq sample: 9_D10ARNP_AAV_IL7R_3\n",
      "single por paired paired\n",
      "            sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  9_D10ARNP_AAV_IL7R_3  N711  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                131926                      87454  \n",
      "-----  now making bed file for 9_D10ARNP_AAV_IL7R_3   N711   S505\n",
      "made global bed file: N711 _ S505\n",
      "It is called 9_D10ARNP_AAV_IL7R_3\n",
      "doing the quantification of global alignments\n",
      "/home/eric/Data/Spaced_Nicking/AAVSeq2_ELANE_PRF1_IL7R_2_2021/results/global_alignment_filtered_summary.xlsx\n"
     ]
    }
   ],
   "source": [
    "###  ALIGN END-TO-END GLOBAL\n",
    "#s. It makes a bam file that is filtered for AS and MAPQ \n",
    "#   It quantifies these bam files and sorts them. Then it converts to a bed file.\n",
    "#\n",
    "\n",
    "                #### Required Arguments\n",
    "cpu = 12\n",
    "min_MAPQ = 50\n",
    "min_AS = -180\n",
    "\n",
    "             ### WHAT TYPE OF ANALYSIS ARE WE RUNNING #####\n",
    "\n",
    "\n",
    "            #### this does the acutal alignment for end-to-end\n",
    "do_align = 'off'\n",
    "single_or_paired_reads = \"paired\"   #can be 'single' or 'paired' for the read alignment\n",
    "\n",
    "            #### WHAT SAMPLES TO RUN ####\n",
    "lines_to_run = range(27)\n",
    "\n",
    "    #for exporting summary of data\n",
    "results_df_all = pd.DataFrame()\n",
    "results_file = os.path.join(directory, 'results','global_alignment_filtered_summary.xlsx')\n",
    "\n",
    "\n",
    "for i in lines_to_run:\n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    print('running row', i)\n",
    "    print('the analysis is on AAV-seq sample:', amplicon_info['name'])\n",
    "    reference_assembly = amplicon_info['genome']\n",
    "    sample_name = amplicon_info['name']\n",
    "    \n",
    "    if do_align == 'on':\n",
    "        #####this aligns everything after the break to end-to-end keep_sam=1 means keep the sam file\n",
    "        align_guideseq_end_to_end_genome_global(directory, amplicon_info, reference_assembly, single_or_paired_reads, cpu, keep_sam=1)\n",
    "\n",
    "\n",
    "    #### Process the alignment files to make a bam file of only high quality reads\n",
    "\n",
    "    #this filters for good alignments based on AS score and MapQ and exports a final indexed bam file.\n",
    "\n",
    "    df_sample_results = guideseq_filtered_mapq_AS_primary(directory, 'global', amplicon_info, single_or_paired_reads)\n",
    "    #this is a running dataframe of the clenaed up read summary\n",
    "    results_df_all = results_df_all.append(df_sample_results, ignore_index=True)\n",
    "\n",
    "\n",
    "    ### global alignment bed generation \n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "    print('-----  now making bed file for', sample_name, ' ',N7, ' ', N5 )\n",
    "\n",
    "\n",
    "    if single_or_paired_reads == \"single\":\n",
    "        genome_global_bed_file = create_filename(directory, N7, N5, 'guideseq_global_bed_single')\n",
    "        filtered_and_sorted_bam_file = create_filename(directory, N7, N5, 'filtered_and_sorted_genome_global_single')\n",
    "    elif single_or_paired_reads == \"paired\":\n",
    "        genome_global_bed_file = create_filename(directory, N7, N5, 'guideseq_global_bed_paired')\n",
    "        filtered_and_sorted_bam_file = create_filename(directory, N7, N5, 'filtered_and_sorted_genome_global_paired')\n",
    "\n",
    "    !bam2bed < {filtered_and_sorted_bam_file} > {genome_global_bed_file}\n",
    "    bed_folder = os.path.join(directory, 'bed_files')\n",
    "\n",
    "    if not os.path.exists(bed_folder):\n",
    "        os.mkdir(bed_folder)\n",
    "\n",
    "    file_location = create_filename(directory, N7, N5, 'bam_global')\n",
    "\n",
    "    new_bed_name = os.path.join(file_location, sample_name + '_' + 'global_'+ single_or_paired_reads + '.bed')\n",
    "\n",
    "    os.rename(genome_global_bed_file, new_bed_name)\n",
    "\n",
    "    shutil.copy(new_bed_name, bed_folder)\n",
    "\n",
    "    #remove the original bed file in the bam folder\n",
    "    os.remove(new_bed_name)\n",
    "\n",
    "\n",
    "    print('made global bed file:', N7, '_', N5 )\n",
    "    print('It is called',sample_name )\n",
    "    print('doing the quantification of global alignments')\n",
    "\n",
    "\n",
    " ## export all the read summaries    \n",
    "results_df_all.to_excel(results_file)    \n",
    "print(results_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "listnumber2                 object\n",
      "Sequence                    object\n",
      "Name                        object\n",
      "Top off-target sites        object\n",
      "Strand                      object\n",
      "MutDist                      int64\n",
      "Risk                        object\n",
      "Annotation (if relevant)    object\n",
      "chr                         object\n",
      "position_start               int64\n",
      "position_end                object\n",
      "begin_site_range             int64\n",
      "end_site_range               int64\n",
      "dtype: object\n",
      "running row 1\n",
      "the analysis is on Guideseq sample: 2-1-3'GSP-TD\n",
      "index\n",
      "index N702 S502\n",
      "/home/eric/Data/Spaced_Nicking/GuideSeq_MiniSeq1/results/N702_S502_2-1-3'GSP-TD_guide_count.xlsx\n",
      "/home/eric/Data/Spaced_Nicking/GuideSeq_MiniSeq1/N702_S502/bam_genome_global_files\n",
      "/home/eric/Data/Spaced_Nicking/GuideSeq_MiniSeq1/N702_S502/bam_genome_global_files/2-1-3'GSP-TD_paired.bed\n",
      "chr                     object\n",
      "chromStart               int64\n",
      "chromEnd                 int64\n",
      "name                    object\n",
      "MapQ                     int64\n",
      "strand                  object\n",
      "flag                     int64\n",
      "CIGAR                   object\n",
      "paired_name             object\n",
      "read1start               int64\n",
      "paired_amplicon_size     int64\n",
      "sequence                object\n",
      "quality_score           object\n",
      "a                       object\n",
      "b                       object\n",
      "c                       object\n",
      "d                       object\n",
      "e                       object\n",
      "f                       object\n",
      "g                       object\n",
      "h                       object\n",
      "i                       object\n",
      "dtype: object\n",
      "list of read hits: [1, 2, 3, 4, 5, 6]\n"
     ]
    }
   ],
   "source": [
    "############ Use CIRPSR Gold and look at overlaps\n",
    "\n",
    "single_or_paired_reads = \"paired\"  #can be 'single' or 'paired' for the read alignment\n",
    "windowsize = 500\n",
    "\n",
    "#### WHAT SAMPLES TO RUN ####\n",
    "lines_to_run = [1]\n",
    "#lines_to_run = range(30)\n",
    "\n",
    "\n",
    "#off_target_sites \n",
    "crispr_gold_file = '/home/eric/Data/Spaced_Nicking/GuideSeq_MiniSeq1/CrispRgold_predictions.xlsx'\n",
    "df_targets = pd.read_excel(crispr_gold_file, index_col=0)\n",
    "\n",
    "# new data frame with split guide location. CRISPRGOLD puts it all in at once\n",
    "new = df_targets[\"Position\"].str.split(\":\", expand = True) \n",
    "df_targets[\"chr\"]= new[0] \n",
    "df_targets[\"position_start\"]= new[1] \n",
    "df_targets[\"position_end\"]= new[2] \n",
    "df_targets.drop(columns =[\"Position\"], inplace = True) \n",
    "df_targets['position_start'] = df_targets['position_start'].astype(int)\n",
    "df_targets['begin_site_range'] = df_targets['position_start'] - windowsize\n",
    "df_targets['end_site_range'] = df_targets['position_start'] + windowsize\n",
    "\n",
    "print(df_targets.dtypes)\n",
    "\n",
    "#print(df_targets)\n",
    "\n",
    "\n",
    "for i in lines_to_run:\n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    name = amplicon_info['name']\n",
    "    print('running row', i)\n",
    "    print('the analysis is on Guideseq sample:', name)\n",
    "    print('index',)\n",
    "    \n",
    "    #import bed file\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "    print('index', N7, N5)\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    print(results_file)\n",
    "\n",
    "    file_location = create_filename(directory, N7, N5, 'bam_global')   \n",
    "    print(file_location)\n",
    "    sample_name = amplicon_info['name']\n",
    "    bedfile = os.path.join(file_location, sample_name + '_' + single_or_paired_reads + '.bed')\n",
    "    print(bedfile)\n",
    "    \n",
    "    #needs to add header names to specify the number of columns so that \n",
    "    \n",
    "    header_names = ['chr', 'chromStart', 'chromEnd', 'name', 'MapQ', 'strand', 'flag', 'CIGAR', 'paired_name', 'read1start', 'paired_amplicon_size', 'sequence', 'quality_score','a','b','c','d','e','f','g','h','i']\n",
    "\n",
    "#    df_bed = pd.read_csv(bedfile, sep='\\t', comment='t', header=None)\n",
    "    df_bed = pd.read_csv(bedfile, sep='\\t', header=None, names = header_names)\n",
    "\n",
    "    \n",
    "    \n",
    "    #header = ['chr', 'chromStart', 'chromEnd', 'name', 'MapQ', 'strand', 'flag', 'CIGAR', 'paired_name', 'read1start', 'paired_amplicon_size', 'sequence', 'quality_score','a','b','c','d','e','f','g','h','i']\n",
    "    #df_bed.columns = header[:len(df_bed.columns)]\n",
    "    \n",
    "    print(df_bed.dtypes) \n",
    "    \n",
    "    read_hits = []\n",
    "    count = 0\n",
    "    for i,g in itertools.izip(df_bed.chromStart, df_bed.chr):\n",
    "        bed_alignment = i\n",
    "        chromosome = g\n",
    "        for a,b,c,d in itertools.izip(df_targets.listnumber2, df_targets.begin_site_range, df_targets.end_site_range, df_targets.chr): \n",
    "            counter_A = 0\n",
    "            if bed_alignment in range(b, c) and chromosome == d:\n",
    "                if counter = 0:\n",
    "                    print(b,c,'chromosome', d, ' guide:', a)\n",
    "                    read_hits.append(a)\n",
    "                    counter_A += 1\n",
    "                    print(b,c,'chromosome', d, ' guide:', a, 'for read starting at :', i)\n",
    "        count += 1 \n",
    "        if count >30:\n",
    "            break\n",
    "    results_df = pd.DataFrame({'sample_name': [name],\n",
    "                               'i7': [N7],\n",
    "                               'i5': [N5],\n",
    "                               'list_of_read_guide_matches' : [read_hits],\n",
    "                               },\n",
    "                                  columns=['sample_name',\n",
    "                                           'i7',\n",
    "                                           'i5',\n",
    "                                           'list_of_read_guide_matches'])            \n",
    "    \n",
    "    \n",
    "    \n",
    "    results_file = os.path.join(directory, 'results',N7 + '_' + N5 + '_'+ name +'_guide_count.xlsx')\n",
    "    \n",
    "    results_df.to_excel(results_file)    \n",
    "\n",
    "    print('list of read hits:',read_hits)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def create_filename(dir_sample, N7, N5, filetype):\n",
    "    main_folder = os.path.join(dir_sample, N7 + '_' + N5)\n",
    "    if filetype == 'mainfolder':\n",
    "        return main_folder\n",
    "    elif filetype == 'amplicons':\n",
    "        return os.path.join(main_folder, 'amplicons')\n",
    "    elif filetype == 'bam_global':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files')\n",
    "    elif filetype == 'bam_local':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files')\n",
    "    \n",
    "    elif filetype == 'R1fastq':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R1.fastq')\n",
    "    elif filetype == 'R1fastqgz':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R1.fastq.gz')\n",
    "    elif filetype == 'R2fastq':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R2.fastq')\n",
    "    elif filetype == 'R2fastqgz':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R2.fastq.gz')\n",
    "      \n",
    "    #### I added these as the sequences that were correctly primed\n",
    "    elif filetype == 'R1fastq_CorrPrime':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R1.CorrPrime.fastq')\n",
    "    elif filetype == 'R1fastqgz_CorrPrime':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R1.CorrPrime.fastq.gz')\n",
    "    elif filetype == 'R2fastq_CorrPrime':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R2.CorrPrime.fastq')\n",
    "    elif filetype == 'R2fastqgz_CorrPrime':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + '_R2.CorrPrime.fastq.gz')\n",
    "    elif filetype == 'localpriming_data':\n",
    "        return os.path.join(main_folder, 'fastq_files', N7 + '_' + N5 + 'priming.xlsx')\n",
    "    elif filetype == 'priming_summary':\n",
    "        return os.path.join(main_folder, 'results', 'priming_summary.xlsx')\n",
    "    #####\n",
    "   \n",
    "    \n",
    "    elif filetype == 'R1trimmed':\n",
    "        return os.path.join(main_folder, 'cutadapt_files', N7 + '_' + N5 + '_R1.trimmed.fastq.gz')\n",
    "    elif filetype == 'R2trimmed':\n",
    "        return os.path.join(main_folder, 'cutadapt_files', N7 + '_' + N5 + '_R2.trimmed.fastq.gz')\n",
    "    elif filetype == 'trimmed_report':\n",
    "        return os.path.join(main_folder, 'cutadapt_files', N7 + '_' + N5 + '.trimmed.report.txt')\n",
    "    elif filetype == 'sam_genome_local':\n",
    "        return os.path.join(main_folder, 'sam_genome_local_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_genome_local':\n",
    "        return os.path.join(main_folder, 'sam_genome_local_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_genome_local':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_genome_local':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_genome_local':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'sam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'sam_plasmid_local_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_plasmid_local':\n",
    "        return os.path.join(main_folder, 'sam_plasmid_local_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'unmapped_bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '_unmapped.bam')\n",
    "    elif filetype == 'qsorted_unmapped_bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '_qsorted_unmapped.bam')\n",
    "    elif filetype == 'unmapped_plasmid_R1fastq':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R1.fastq')\n",
    "    elif filetype == 'unmapped_plasmid_R2fastq':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R2.fastq')\n",
    "    elif filetype == 'unmapped_plasmid_R1fastqgz':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R1.fastq.gz')\n",
    "    elif filetype == 'unmapped_plasmid_R2fastqgz':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R2.fastq.gz')\n",
    "    elif filetype == 'sam_amplicons':\n",
    "        return os.path.join(main_folder, 'sam_amplicon_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_amplicons':\n",
    "        return os.path.join(main_folder, 'sam_amplicon_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'unmapped_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '_amplicons_unmapped.bam')\n",
    "    elif filetype == 'qsorted_unmapped_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '_qsorted_amplicons_unmapped.bam')\n",
    "    elif filetype == 'unmapped_amplicons_R1fastq':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files', N7 + '_' + N5 + '_amplicons_unmapped_R1.fastq')\n",
    "    elif filetype == 'unmapped_amplicons_R2fastq':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files', N7 + '_' + N5 + '_amplicons_unmapped_R2.fastq')\n",
    "    elif filetype == 'unmapped_amplicons_R1fastqgz':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files',\n",
    "                            N7 + '_' + N5 + '_amplicons_unmapped_R1.fastq.gz')\n",
    "    elif filetype == 'unmapped_amplicons_R2fastqgz':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files',\n",
    "                            N7 + '_' + N5 + '_amplicons_unmapped_R2.fastq.gz')\n",
    "    elif filetype == 'unmapped_amplicons_report':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files', N7 + '_' + N5 + '.unmapped.report.txt')\n",
    "    elif filetype == 'sam_genome_global':\n",
    "        return os.path.join(main_folder, 'sam_genome_global_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_genome_global':\n",
    "        return os.path.join(main_folder, 'sam_genome_global_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_genome_global':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_genome_global':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_genome_global':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'results_amplicons':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5)  # We will append the window size later\n",
    "    elif filetype == 'results_plasmid':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_plasmid.xlsx')\n",
    "    elif filetype == 'results_all_amplicons':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_all_amplicons.xlsx')\n",
    "    elif filetype == 'results_genomewide':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_genomewide.xlsx')\n",
    "    elif filetype == 'summary_all_alignments':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_summary_all_alignments.xlsx')\n",
    "    elif filetype == 'read_counts':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_read_counts.xlsx')\n",
    "    ## more added\n",
    "    elif filetype == 'results_pipeline2_global':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_pipeline2_global.xlsx')\n",
    "    elif filetype == 'results_pipeline2_local':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_pipeline2_local.xlsx')\n",
    "    elif filetype == 'R1trimmed':\n",
    "        return os.path.join(main_folder, 'cutadapt_files', N7 + '_' + N5 + '_R1.trimmed.fastq.gz')\n",
    "    elif filetype == 'R2trimmed':\n",
    "        return os.path.join(main_folder, 'cutadapt_files', N7 + '_' + N5 + '_R2.trimmed.fastq.gz')\n",
    "    elif filetype == 'trimmed_report':\n",
    "        return os.path.join(main_folder, 'cutadapt_files', N7 + '_' + N5 + '.trimmed.report.txt')\n",
    "    elif filetype == 'sam_genome_local':\n",
    "        return os.path.join(main_folder, 'sam_genome_local_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_genome_local':\n",
    "        return os.path.join(main_folder, 'sam_genome_local_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_genome_local':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_genome_local':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_genome_local':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'sam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'sam_plasmid_local_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_plasmid_local':\n",
    "        return os.path.join(main_folder, 'sam_plasmid_local_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'unmapped_bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '_unmapped.bam')\n",
    "    elif filetype == 'qsorted_unmapped_bam_plasmid_local':\n",
    "        return os.path.join(main_folder, 'bam_plasmid_local_files', N7 + '_' + N5 + '_qsorted_unmapped.bam')\n",
    "    elif filetype == 'unmapped_plasmid_R1fastq':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R1.fastq')\n",
    "    elif filetype == 'unmapped_plasmid_R2fastq':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R2.fastq')\n",
    "    elif filetype == 'unmapped_plasmid_R1fastqgz':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R1.fastq.gz')\n",
    "    elif filetype == 'unmapped_plasmid_R2fastqgz':\n",
    "        return os.path.join(main_folder, 'plasmid_unmapped_fastq_files', N7 + '_' + N5 + '_plasmid_unmapped_R2.fastq.gz')\n",
    "    elif filetype == 'sam_amplicons':\n",
    "        return os.path.join(main_folder, 'sam_amplicon_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_amplicons':\n",
    "        return os.path.join(main_folder, 'sam_amplicon_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'unmapped_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '_amplicons_unmapped.bam')\n",
    "    elif filetype == 'qsorted_unmapped_bam_amplicons':\n",
    "        return os.path.join(main_folder, 'bam_amplicon_files', N7 + '_' + N5 + '_qsorted_amplicons_unmapped.bam')\n",
    "    elif filetype == 'unmapped_amplicons_R1fastq':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files', N7 + '_' + N5 + '_amplicons_unmapped_R1.fastq')\n",
    "    elif filetype == 'unmapped_amplicons_R2fastq':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files', N7 + '_' + N5 + '_amplicons_unmapped_R2.fastq')\n",
    "    elif filetype == 'unmapped_amplicons_R1fastqgz':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files',\n",
    "                            N7 + '_' + N5 + '_amplicons_unmapped_R1.fastq.gz')\n",
    "    elif filetype == 'unmapped_amplicons_R2fastqgz':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files',\n",
    "                            N7 + '_' + N5 + '_amplicons_unmapped_R2.fastq.gz')\n",
    "    elif filetype == 'unmapped_amplicons_report':\n",
    "        return os.path.join(main_folder, 'amplicons_unmapped_fastq_files', N7 + '_' + N5 + '.unmapped.report.txt')\n",
    "    elif filetype == 'sam_genome_global':\n",
    "        return os.path.join(main_folder, 'sam_genome_global_files', N7 + '_' + N5 + '.sam')\n",
    "    elif filetype == 'sam_report_genome_global':\n",
    "        return os.path.join(main_folder, 'sam_genome_global_files', N7 + '_' + N5 + '.sam.report.txt')\n",
    "    elif filetype == 'bam_genome_global':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '.bam')\n",
    "    elif filetype == 'sorted_bam_genome_global':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '.sorted.bam')\n",
    "    elif filetype == 'sorted_bai_genome_global':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '.sorted.bam.bai')\n",
    "    elif filetype == 'results_amplicons':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5)  # We will append the window size later\n",
    "    elif filetype == 'results_plasmid':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_plasmid.xlsx')\n",
    "    elif filetype == 'results_all_amplicons':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_all_amplicons.xlsx')\n",
    "    elif filetype == 'results_genomewide':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_genomewide.xlsx')\n",
    "    elif filetype == 'summary_all_alignments':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_summary_all_alignments.xlsx')\n",
    "    elif filetype == 'read_counts':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_read_counts.xlsx')\n",
    "    ## more added\n",
    "    elif filetype == 'results_pipeline2_global':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_pipeline2_global.xlsx')\n",
    "    elif filetype == 'results_pipeline2_local':\n",
    "        return os.path.join(main_folder, 'results', N7 + '_' + N5 + '_results_pipeline2_local.xlsx')\n",
    "    elif filetype == 'sam_genome_global_single':\n",
    "        return os.path.join(main_folder, 'sam_genome_global_files', N7 + '_' + N5 + '_single.sam')\n",
    "    elif filetype == 'sam_genome_local_single':\n",
    "        return os.path.join(main_folder, 'sam_genome_local_files', N7 + '_' + N5 + '_single.sam')\n",
    "    elif filetype == 'sam_report_genome_global_single':\n",
    "        return os.path.join(main_folder, 'sam_genome_global_files', N7 + '_' + N5 + '_single.sam.report.txt')\n",
    "    elif filetype == 'sam_report_genome_local_single':\n",
    "        return os.path.join(main_folder, 'sam_genome_local_files', N7 + '_' + N5 + '_single.sam.report.txt')\n",
    "    elif filetype == 'bam_genome_global_single':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_single.bam')\n",
    "    elif filetype == 'bam_genome_local_single':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '_single.bam')\n",
    "    elif filetype == 'sorted_bam_genome_global_single':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_single.sorted.bam')\n",
    "    elif filetype == 'sorted_bam_genome_local_single':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '_single.sorted.bam')\n",
    "    elif filetype == 'sam_genome_global_paired':\n",
    "        return os.path.join(main_folder, 'sam_genome_global_files', N7 + '_' + N5 + '_paired.sam')\n",
    "    elif filetype == 'sam_genome_local_paired':\n",
    "        return os.path.join(main_folder, 'sam_genome_local_files', N7 + '_' + N5 + '_paired.sam')\n",
    "    elif filetype == 'sam_report_genome_global_paired':\n",
    "        return os.path.join(main_folder, 'sam_genome_global_files', N7 + '_' + N5 + '_paired.sam.report.txt')\n",
    "    elif filetype == 'sam_report_genome_local_paired':\n",
    "        return os.path.join(main_folder, 'sam_genome_local_files', N7 + '_' + N5 + '_paired.sam.report.txt')\n",
    "    elif filetype == 'bam_genome_global_paired':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_paired.bam')\n",
    "    elif filetype == 'bam_genome_local_paired':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '_paired.bam')\n",
    "    elif filetype == 'sorted_bam_genome_global_paired':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_paired.sorted.bam')\n",
    "    elif filetype == 'sorted_bam_genome_local_paired':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '_paired.sorted.bam')\n",
    "    elif filetype == 'filtered_and_sorted_genome_global_single':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_single_primary_as_mapq.sorted.bam')    \n",
    "    elif filetype == 'filtered_and_sorted_genome_local_single':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '_single_primary_as_mapq.sorted.bam')    \n",
    "    elif filetype == 'genome_global_bed_single':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_single_global.sorted.bed')\n",
    "    elif filetype == 'filtered_and_sorted_genome_global_paired':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_paired_primary_as_mapq.sorted.bam')  \n",
    "    elif filetype == 'filtered_and_sorted_genome_local_paired':\n",
    "        return os.path.join(main_folder, 'bam_genome_local_files', N7 + '_' + N5 + '_paired_primary_as_mapq.sorted.bam')  \n",
    "    elif filetype == 'genome_global_bed_paired':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_paired_global.sorted.bed')\n",
    "    elif filetype == 'guideseq_global_bed_single':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_single.global.sorted.bed')\n",
    "    elif filetype == 'guideseq_global_bed_paired':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_paired.global.sorted.bed')\n",
    "    elif filetype == 'guideseq_local_bed_single':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_single.local.sorted.bed')\n",
    "    elif filetype == 'guideseq_local_bed_paired':\n",
    "        return os.path.join(main_folder, 'bam_genome_global_files', N7 + '_' + N5 + '_paired.local.sorted.bed')\n",
    "###########################\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "apple += 1\n",
    "print(apple)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "apple += 1\n",
    "print(apple)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def guideseq_filtered_mapq_AS_primary(dir_sample, global_or_local, amplicon_info, single_or_paired = 'single', min_MAPQ = 25, min_AS=-180):\n",
    "    #function beginning\n",
    "    # define samples\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "    name = amplicon_info['name']\n",
    "\n",
    "\n",
    "    \n",
    "    if global_or_local == 'global':\n",
    "        if single_or_paired == 'single':\n",
    "            bam_file = create_filename(dir_sample, N7, N5, 'sorted_bam_genome_global_single')\n",
    "            final_trimmed_bam_filtered_file = create_filename(dir_sample, N7, N5, 'filtered_and_sorted_genome_global_single')\n",
    "        elif single_or_paired == 'paired':\n",
    "            bam_file = create_filename(dir_sample, N7, N5, 'sorted_bam_genome_global_paired')\n",
    "            final_trimmed_bam_filtered_file = create_filename(dir_sample, N7, N5, 'filtered_and_sorted_genome_global_paired')\n",
    " \n",
    "    if global_or_local == 'local':\n",
    "        if single_or_paired == 'single':\n",
    "            bam_file = create_filename(dir_sample, N7, N5, 'sorted_bam_genome_local_single')\n",
    "            final_trimmed_bam_filtered_file = create_filename(dir_sample, N7, N5, 'filtered_and_sorted_genome_local_single')\n",
    "        elif single_or_paired == 'paired':\n",
    "            print('wooo')\n",
    "            bam_file = create_filename(dir_sample, N7, N5, 'sorted_bam_genome_local_paired')\n",
    "            final_trimmed_bam_filtered_file = create_filename(dir_sample, N7, N5, 'filtered_and_sorted_genome_local_paired')    \n",
    "    \n",
    "    all_reads_unfiltered = []\n",
    "    all_reads_pass_filter = []\n",
    "    print(bam_file)\n",
    "    bam_alignment_file = pysam.AlignmentFile(bam_file, 'rb')\n",
    "    bam_in = bam_alignment_file.fetch()\n",
    "    \n",
    "    #make output file\n",
    "    filtered_reads = pysam.AlignmentFile(final_trimmed_bam_filtered_file, \"wb\", template=bam_alignment_file)\n",
    "    \n",
    "    #test each read\n",
    "    for read in bam_in:\n",
    "        all_reads_unfiltered.append(read.query_name)\n",
    "        if read.has_tag('AS'):\n",
    "            read_AS = read.get_tag('AS')\n",
    "        #check read quality is good\n",
    "        if read.mapping_quality >= min_MAPQ and read_AS >= min_AS and not read.is_secondary:\n",
    "            all_reads_pass_filter.append(read.query_name)\n",
    "            filtered_reads.write(read)\n",
    "    filtered_reads.close()\n",
    "    bam_alignment_file.close()\n",
    "\n",
    "\n",
    "    count_unfiltered = len(all_reads_unfiltered)\n",
    "\n",
    "    count_filtered = len(all_reads_pass_filter)\n",
    "    \n",
    "    \n",
    "    \n",
    "    results_df = pd.DataFrame({'sample_name': [name],\n",
    "                               'i7': [N7],\n",
    "                               'i5': [N5],\n",
    "                               'single_or_paired_reads' : [single_or_paired],\n",
    "                               'all_alignments_count': [count_unfiltered],\n",
    "                               'filtered_alignments_count': [count_filtered]\n",
    "                               },\n",
    "                                  columns=['sample_name',\n",
    "                                           'i7',\n",
    "                                           'i5',\n",
    "                                           'single_or_paired_reads',\n",
    "                                           'all_alignments_count',\n",
    "                                           'filtered_alignments_count'])\n",
    "    print('single por paired', single_or_paired)\n",
    "    \n",
    "    print(results_df)   \n",
    "    return results_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1, 2, 3, 4]\n"
     ]
    }
   ],
   "source": [
    "print(range(1,5))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
