{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This is for PRF1 ELANE and ILR7 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import matplotlib\n",
    "matplotlib.use('Agg')\n",
    "import matplotlib.pyplot as plt\n",
    "import pylab\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import gzip\n",
    "import itertools\n",
    "import operator\n",
    "import subprocess\n",
    "import twobitreader\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import pysam\n",
    "import shutil\n",
    "\n",
    "#from uditas.uditas_replace_helpers import *\n",
    "#from uditas.uditas_guideseq import *\n",
    "\n",
    "from guideseq_functions.guideseq_helpers_ElaneIL7Prf import *\n",
    "from guideseq_functions.guideseq_ElaneIL7Prf import *\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem log\n",
    "\n",
    "** makes sure you keep your I5 primers binding to the ODN sequence as this becomes \"Read1 by the machine\"\n",
    "1. do 200 long read for Read1 \n",
    "\n",
    "2.naming files and folder\n",
    "2. longer site in front of primer binding\n",
    "3. cut out amplicons that are 400-700 so I don't loose reads\n"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# Danner_GuideseqLogic\n",
    "\n",
    "#  This is a guideseq pipeline. Different than the oirignal but built in a simliar form. \n",
    "## It checks for the on target priming, the uses bowtie2 trimming off the adapter seq. Then exports a bed file for comparing to CRISPRgold\n",
    "\n",
    "\n",
    "### 1. Check reads were correctly primed.\n",
    "### 2. Trim the short reads that run into the illumina adapters or ODN primer sequences   \n",
    "### 3. Do global end-to-end alignment\n",
    "### 4. process bam file and pull out the high mapQ reads\n",
    "### 5. convert to bed\n",
    "### 6. import bed file as dataframe\n",
    "### 7. check for bed mapping to within 500-1000bp of the predicted cut sites\n",
    "### 8. check if 5' and 3' read both have mappings to the same off target site.\n",
    "\n",
    "We use read1 for checking priming. Also we could use that for mapping (or paired end). On both the 3 and 5' direction the first 34 nucleotides of the read contain the integrated ODN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/eric/Data/Spaced_Nicking/GuideSeq_MiniSeq3_ELANE_IL7R_PRF1\n"
     ]
    }
   ],
   "source": [
    "#Directory\n",
    "\n",
    "directory = '/home/eric/Data/Spaced_Nicking/GuideSeq_MiniSeq3_ELANE_IL7R_PRF1'\n",
    "\n",
    "print(directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Discard Mispriming Reads\n",
    "When you put a universal primer on the ends of everything, every mispriming event will amplify. An effect we normally don't deal with. I did nested PCR to reduce this. However 85% of the alignments in the UDITAS data I looked at seemed to be mispriming. They did all their blasting and analysis before removing mispriming. But to save computational power and remove error early on I will discard mispriming events. \n",
    "They discard these only for plasmid alignments analyze_alignments_plasmid for some reason which comes from the bam file.\n",
    "\n",
    "### for guideseq read1 is where the ODN primer binds and so that is how you tell targeting. For the reverse it was binidng to the transposon which is random but then the seq primer binds right on it so the the read for read2 begins in the genomic sequence. the Rev primer could misprime but it shouldnt matter as it would then just bind and make smaller amplicons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with sample 0\n",
      "done with sample 1\n",
      "done with sample 2\n",
      "done with sample 3\n",
      "done with sample 4\n",
      "done with sample 5\n",
      "done with sample 6\n",
      "done with sample 7\n",
      "done with sample 8\n",
      "done with sample 9\n",
      "done with sample 10\n",
      "done with sample 11\n",
      "done with sample 12\n",
      "done with sample 13\n",
      "done with sample 14\n",
      "done with sample 15\n",
      "done with sample 16\n",
      "done with sample 17\n",
      "done with sample 18\n",
      "done with sample 19\n",
      "done with sample 20\n",
      "done with sample 21\n",
      "done with sample 22\n",
      "done with sample 23\n",
      "done with sample 24\n",
      "done with sample 25\n",
      "done with sample 26\n",
      "done with sample 27\n",
      "done with sample 28\n",
      "done with sample 29\n",
      "done with sample 30\n",
      "done with sample 31\n",
      "done with sample 32\n",
      "done with sample 33\n",
      "done with sample 34\n",
      "done with sample 35\n",
      "done with sample 36\n",
      "done with sample 37\n",
      "done with sample 38\n",
      "done with sample 39\n",
      "done with sample 40\n",
      "done with sample 41\n",
      "done with sample 42\n",
      "done with sample 43\n",
      "done with sample 44\n",
      "done with sample 45\n",
      "done with sample 46\n",
      "done with sample 47\n",
      "            sample_name    i7    i5  total_reads  reads_with_good_priming  \\\n",
      "0          3'GSP-Ctrl-1  N701  S502        64404                     6877   \n",
      "1          3'GSP-Ctrl-2  N702  S502        72205                    22672   \n",
      "2          3'GSP-Ctrl-3  N703  S502        53019                     7033   \n",
      "3         3'GSP-dsODN-1  N704  S502       129743                    32847   \n",
      "4         3'GSP-dsODN-2  N705  S502       159999                    51166   \n",
      "5         3'GSP-dsODN-3  N706  S502       145447                    67960   \n",
      "6    3'GSP-ELANE-Cas9-1  N707  S502       141688                    80396   \n",
      "7    3'GSP-ELANE-Cas9-2  N710  S502       151868                    75775   \n",
      "8    3'GSP-ELANE-Cas9-3  N711  S502       146639                    68451   \n",
      "9   3'GSP-ELANE-Cas9n-1  N712  S502        62033                     9792   \n",
      "10  3'GSP-ELANE-Cas9n-2  N714  S502       125428                    22025   \n",
      "11  3'GSP-ELANE-Cas9n-3  N715  S502        40784                    12717   \n",
      "12    3'GSP-IL7R-Cas9-1  N701  S503       114093                    39544   \n",
      "13    3'GSP-IL7R-Cas9-2  N702  S503       135756                    53030   \n",
      "14    3'GSP-IL7R-Cas9-3  N703  S503       106142                    44404   \n",
      "15   3'GSP-IL7R-Cas9n-1  N704  S503        55162                     7666   \n",
      "16   3'GSP-IL7R-Cas9n-2  N705  S503        62155                    15806   \n",
      "17   3'GSP-IL7R-Cas9n-3  N706  S503       187113                    13583   \n",
      "18    3'GSP-PRF1-Cas9-1  N707  S503       170923                    35549   \n",
      "19    3'GSP-PRF1-Cas9-2  N710  S503        63268                    20780   \n",
      "20    3'GSP-PRF1-Cas9-3  N711  S503        85966                    27687   \n",
      "21   3'GSP-PRF1-Cas9n-1  N712  S503       115551                    19946   \n",
      "22   3'GSP-PRF1-Cas9n-2  N714  S503       120417                    25130   \n",
      "23   3'GSP-PRF1-Cas9n-3  N715  S503        62569                    17309   \n",
      "24         5'GSP-Ctrl-1  N701  S505       144253                     6435   \n",
      "25         5'GSP-Ctrl-2  N702  S505       146497                     7496   \n",
      "26         5'GSP-Ctrl-3  N703  S505       222701                    15284   \n",
      "27        5'GSP-dsODN-1  N704  S505       113855                    25184   \n",
      "28        5'GSP-dsODN-2  N705  S505       197573                    32385   \n",
      "29        5'GSP-dsODN-3  N706  S505       108911                    31832   \n",
      "30   5'GSP-ELANE-Cas9-1  N707  S505       230925                   178455   \n",
      "31   5'GSP-ELANE-Cas9-2  N710  S505       203235                   165586   \n",
      "32   5'GSP-ELANE-Cas9-3  N711  S505       131786                    96939   \n",
      "33  5'GSP-ELANE-Cas9n-1  N712  S505       178495                    12887   \n",
      "34  5'GSP-ELANE-Cas9n-2  N714  S505       135687                    21041   \n",
      "35  5'GSP-ELANE-Cas9n-3  N715  S505       118437                    24828   \n",
      "36    5'GSP-IL7R-Cas9-1  N701  S506       219425                   102695   \n",
      "37    5'GSP-IL7R-Cas9-2  N702  S506       143868                    62151   \n",
      "38    5'GSP-IL7R-Cas9-3  N703  S506       184401                   115200   \n",
      "39   5'GSP-IL7R-Cas9n-1  N704  S506       169630                    14500   \n",
      "40   5'GSP-IL7R-Cas9n-2  N705  S506       124616                     3815   \n",
      "41   5'GSP-IL7R-Cas9n-3  N706  S506       172061                     3679   \n",
      "42    5'GSP-PRF1-Cas9-1  N707  S506       213167                    16733   \n",
      "43    5'GSP-PRF1-Cas9-2  N710  S506       130000                    53322   \n",
      "44    5'GSP-PRF1-Cas9-3  N711  S506       192663                    86203   \n",
      "45   5'GSP-PRF1-Cas9n-1  N712  S506       201967                    30578   \n",
      "46   5'GSP-PRF1-Cas9n-2  N714  S506       217078                    18854   \n",
      "47   5'GSP-PRF1-Cas9n-3  N715  S506       205038                    49921   \n",
      "\n",
      "    reads_with_guideseq_primer_misprimed  \\\n",
      "0                                  22341   \n",
      "1                                  22361   \n",
      "2                                  15063   \n",
      "3                                  45771   \n",
      "4                                  58394   \n",
      "5                                  42952   \n",
      "6                                  29082   \n",
      "7                                  41590   \n",
      "8                                  50555   \n",
      "9                                  23232   \n",
      "10                                 37329   \n",
      "11                                 12928   \n",
      "12                                 37269   \n",
      "13                                 45851   \n",
      "14                                 28608   \n",
      "15                                 19947   \n",
      "16                                 19939   \n",
      "17                                 44878   \n",
      "18                                 35335   \n",
      "19                                 26946   \n",
      "20                                 22186   \n",
      "21                                 37312   \n",
      "22                                 23791   \n",
      "23                                 17268   \n",
      "24                                 16148   \n",
      "25                                 10230   \n",
      "26                                  7975   \n",
      "27                                 27969   \n",
      "28                                 44581   \n",
      "29                                 15286   \n",
      "30                                  6122   \n",
      "31                                  3554   \n",
      "32                                  6061   \n",
      "33                                  9250   \n",
      "34                                 16138   \n",
      "35                                 31394   \n",
      "36                                 21758   \n",
      "37                                 31657   \n",
      "38                                 13893   \n",
      "39                                 21378   \n",
      "40                                 21227   \n",
      "41                                 13553   \n",
      "42                                 19571   \n",
      "43                                 23761   \n",
      "44                                 11553   \n",
      "45                                 24159   \n",
      "46                                 16306   \n",
      "47                                 29880   \n",
      "\n",
      "    reads_with_indexing_primer_mispriming  \n",
      "0                                   35186  \n",
      "1                                   27172  \n",
      "2                                   30923  \n",
      "3                                   51125  \n",
      "4                                   50439  \n",
      "5                                   34535  \n",
      "6                                   32210  \n",
      "7                                   34503  \n",
      "8                                   27633  \n",
      "9                                   29009  \n",
      "10                                  66074  \n",
      "11                                  15139  \n",
      "12                                  37280  \n",
      "13                                  36875  \n",
      "14                                  33130  \n",
      "15                                  27549  \n",
      "16                                  26410  \n",
      "17                                 128652  \n",
      "18                                 100039  \n",
      "19                                  15542  \n",
      "20                                  36093  \n",
      "21                                  58293  \n",
      "22                                  71496  \n",
      "23                                  27992  \n",
      "24                                 121670  \n",
      "25                                 128771  \n",
      "26                                 199442  \n",
      "27                                  60702  \n",
      "28                                 120607  \n",
      "29                                  61793  \n",
      "30                                  46348  \n",
      "31                                  34095  \n",
      "32                                  28786  \n",
      "33                                 156358  \n",
      "34                                  98508  \n",
      "35                                  62215  \n",
      "36                                  94972  \n",
      "37                                  50060  \n",
      "38                                  55308  \n",
      "39                                 133752  \n",
      "40                                  99574  \n",
      "41                                 154829  \n",
      "42                                 176863  \n",
      "43                                  52917  \n",
      "44                                  94907  \n",
      "45                                 147230  \n",
      "46                                 181918  \n",
      "47                                 125237  \n"
     ]
    }
   ],
   "source": [
    "#the amplicon info is related to the line on the csv file. It is indexed from 0.\n",
    "#it is on my extera space so need to make sure that is mounted. (first thing to check if it throws and error)\n",
    "\n",
    "\n",
    "# THIS IS WHAT RUNS THE PRIMER CHECK THROUGH ALL SAMPLES\n",
    "#input the sample range or list below\n",
    "\n",
    "\n",
    "results_df_all = pd.DataFrame()\n",
    "\n",
    "results_folder = os.path.join(directory, 'results')\n",
    "if not os.path.exists(results_folder):\n",
    "    os.mkdir(results_folder)\n",
    "results_file = os.path.join(directory, 'results','all_priming.xlsx')\n",
    "    \n",
    "\n",
    "#make results tree\n",
    "\n",
    "for i in range(48):\n",
    "        \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "\n",
    "    #5primer is everything but AT, AT is for checking mispriming. The full sequence is the olgio from guideseq.\n",
    "    #EVERYTHING IS CAPITAL\n",
    "    guideseq3_seq = 'GTTTAATTGAGTTGTCATATGTTAATAACGGTAT'\n",
    "    guideseq3_primeronly = 'GTTTAATTGAGTTGTCATATGTTAATAACGGT'\n",
    "    #3primer is everything but the final AC. The AC is for checking mispriming\n",
    "    guideseq5_seq = 'ATACCGTTATTAACATATGACAACTCAATTAAAC'\n",
    "    guideseq5_primeronly = 'ATACCGTTATTAACATATGACAACTCAATTAA'\n",
    "\n",
    "    direction = amplicon_info['Direction']\n",
    "\n",
    "    if direction == 3:\n",
    "        primer_seq_plus_downstream = guideseq3_seq\n",
    "        primer_seq = guideseq3_primeronly\n",
    "    elif direction == 5:\n",
    "        primer_seq_plus_downstream = guideseq5_seq\n",
    "        primer_seq = guideseq5_primeronly\n",
    "    \n",
    "    df_sample_results = correct_priming_guideseq(directory, amplicon_info, primer_seq, primer_seq_plus_downstream)\n",
    "    results_df_all = results_df_all.append(df_sample_results, ignore_index=True)\n",
    "    print('done with sample', i)\n",
    "\n",
    "print(results_df_all)\n",
    "results_df_all.to_excel(results_file)    \n",
    "#print reults to excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with sample 0\n",
      "done with sample 1\n",
      "done with sample 2\n",
      "done with sample 3\n",
      "done with sample 4\n",
      "done with sample 5\n",
      "done with sample 6\n",
      "done with sample 7\n",
      "done with sample 8\n",
      "done with sample 9\n",
      "done with sample 10\n",
      "done with sample 11\n",
      "done with sample 12\n",
      "done with sample 13\n",
      "done with sample 14\n",
      "done with sample 15\n",
      "done with sample 16\n",
      "done with sample 17\n",
      "done with sample 18\n",
      "done with sample 19\n",
      "done with sample 20\n",
      "done with sample 21\n",
      "done with sample 22\n",
      "done with sample 23\n",
      "done with sample 24\n",
      "done with sample 25\n",
      "done with sample 26\n",
      "done with sample 27\n",
      "done with sample 28\n",
      "done with sample 29\n",
      "done with sample 30\n",
      "done with sample 31\n",
      "done with sample 32\n",
      "done with sample 33\n",
      "done with sample 34\n",
      "done with sample 35\n",
      "done with sample 36\n",
      "done with sample 37\n",
      "done with sample 38\n",
      "done with sample 39\n",
      "done with sample 40\n",
      "done with sample 41\n",
      "done with sample 42\n",
      "done with sample 43\n",
      "done with sample 44\n",
      "done with sample 45\n",
      "done with sample 46\n",
      "done with sample 47\n"
     ]
    }
   ],
   "source": [
    "### TRIMMING ####\n",
    "#need to trim off the end of the short reads. This is for amplicons that were too short and have the other side on them.\n",
    "\n",
    "for i in range(48):\n",
    "        \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    \n",
    "    trim_guideseq(directory, amplicon_info)\n",
    "    \n",
    "    print('done with sample', i)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "env: BOWTIE2_INDEXES=/home/eric/Data/Ref_Genomes\n"
     ]
    }
   ],
   "source": [
    "# BOWTIE2_INDEXES are needed for global alignments\n",
    "#check in bash: > ECHO $GENOMES_2BIT\n",
    "%env BOWTIE2_INDEXES=/home/eric/Data/Ref_Genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Need to make an new bowtie2 index file that includes targeting for alignment agains the whole genome so it is all in one sheet together. \n",
    "\n",
    "### Other option would be to align it to the amplicon, extract unaligned files and then align to the genome but seems cleaner this way. \n",
    "#### Ideally every sequence is unique between the targeting vector and genome.\n",
    "\n",
    "\n",
    "1. Build your fastas of interest and label .fa files.\n",
    "    1. You need fasta of hg38 or reference genome. You can pull this from downloaded bowtie indexed sampels and then use the following command to turn the index into a fasta file: bowtie2-inspect hg38 > hg38.fa   \n",
    "    2. Put all the fasta files in the same folder. Should also use the transfected plasmid\n",
    "2. index the files with bowtie\n",
    "    1. use the command bowtie2-build -f pE049,pe038_mc.fa,hg38.fa -p hg38_plus_targetvectorandplasmid\n",
    "    2. this has the -p to make it take less ram in my case.\n",
    "    3. In this case it adds the hg38 and the minicircle targeting file together\n",
    "    4. I have a Intel® Core™ i7-5500U CPU @ 2.40GHz × 4 with 15.1 GiB ram and it required about 13.8 gigs of ram and 2 hours to do a hg38+small fasta index\n",
    "    5. be sure to pay attention to the name of the new indexed file. \"hg38_plus_targetvector\" in example above. Add it to the sample_info.csv sheet. Under the tab 'genome_plus_targeting'.\n",
    "    6. you can check it indexed correcctly: bowtie2-inspect -s hg38_plus_targetvector\n",
    "    \n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Now we blast end-to-end the correctly primed sequences.\n",
    "We don't need to trim off anything. But we should remove the length of the ODN primer sequence before blasting for Read1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running row 0\n",
      "the analysis is on Guideseq sample: 3'GSP-Ctrl-1\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-Ctrl-1  N701  S502                 paired                 28484   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      11147  \n",
      "-----  now making bed file for 3'GSP-Ctrl-1   N701   S502\n",
      "made global bed file: N701 _ S502\n",
      "It is called 3'GSP-Ctrl-1\n",
      "doing the quantification of global alignments\n",
      "running row 1\n",
      "the analysis is on Guideseq sample: 3'GSP-Ctrl-2\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-Ctrl-2  N702  S502                 paired                 91679   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      37364  \n",
      "-----  now making bed file for 3'GSP-Ctrl-2   N702   S502\n",
      "made global bed file: N702 _ S502\n",
      "It is called 3'GSP-Ctrl-2\n",
      "doing the quantification of global alignments\n",
      "running row 2\n",
      "the analysis is on Guideseq sample: 3'GSP-Ctrl-3\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-Ctrl-3  N703  S502                 paired                 33714   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      12014  \n",
      "-----  now making bed file for 3'GSP-Ctrl-3   N703   S502\n",
      "made global bed file: N703 _ S502\n",
      "It is called 3'GSP-Ctrl-3\n",
      "doing the quantification of global alignments\n",
      "running row 3\n",
      "the analysis is on Guideseq sample: 3'GSP-dsODN-1\n",
      "single por paired paired\n",
      "     sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-dsODN-1  N704  S502                 paired                 85333   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      42213  \n",
      "-----  now making bed file for 3'GSP-dsODN-1   N704   S502\n",
      "made global bed file: N704 _ S502\n",
      "It is called 3'GSP-dsODN-1\n",
      "doing the quantification of global alignments\n",
      "running row 4\n",
      "the analysis is on Guideseq sample: 3'GSP-dsODN-2\n",
      "single por paired paired\n",
      "     sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-dsODN-2  N705  S502                 paired                192114   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      75288  \n",
      "-----  now making bed file for 3'GSP-dsODN-2   N705   S502\n",
      "made global bed file: N705 _ S502\n",
      "It is called 3'GSP-dsODN-2\n",
      "doing the quantification of global alignments\n",
      "running row 5\n",
      "the analysis is on Guideseq sample: 3'GSP-dsODN-3\n",
      "single por paired paired\n",
      "     sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-dsODN-3  N706  S502                 paired                317432   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                     117807  \n",
      "-----  now making bed file for 3'GSP-dsODN-3   N706   S502\n",
      "made global bed file: N706 _ S502\n",
      "It is called 3'GSP-dsODN-3\n",
      "doing the quantification of global alignments\n",
      "running row 6\n",
      "the analysis is on Guideseq sample: 3'GSP-ELANE-Cas9-1\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-ELANE-Cas9-1  N707  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                150549                     128897  \n",
      "-----  now making bed file for 3'GSP-ELANE-Cas9-1   N707   S502\n",
      "made global bed file: N707 _ S502\n",
      "It is called 3'GSP-ELANE-Cas9-1\n",
      "doing the quantification of global alignments\n",
      "running row 7\n",
      "the analysis is on Guideseq sample: 3'GSP-ELANE-Cas9-2\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-ELANE-Cas9-2  N710  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                112073                     103757  \n",
      "-----  now making bed file for 3'GSP-ELANE-Cas9-2   N710   S502\n",
      "made global bed file: N710 _ S502\n",
      "It is called 3'GSP-ELANE-Cas9-2\n",
      "doing the quantification of global alignments\n",
      "running row 8\n",
      "the analysis is on Guideseq sample: 3'GSP-ELANE-Cas9-3\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-ELANE-Cas9-3  N711  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                130093                     101057  \n",
      "-----  now making bed file for 3'GSP-ELANE-Cas9-3   N711   S502\n",
      "made global bed file: N711 _ S502\n",
      "It is called 3'GSP-ELANE-Cas9-3\n",
      "doing the quantification of global alignments\n",
      "running row 9\n",
      "the analysis is on Guideseq sample: 3'GSP-ELANE-Cas9n-1\n",
      "single por paired paired\n",
      "           sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-ELANE-Cas9n-1  N712  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 19006                      12810  \n",
      "-----  now making bed file for 3'GSP-ELANE-Cas9n-1   N712   S502\n",
      "made global bed file: N712 _ S502\n",
      "It is called 3'GSP-ELANE-Cas9n-1\n",
      "doing the quantification of global alignments\n",
      "running row 10\n",
      "the analysis is on Guideseq sample: 3'GSP-ELANE-Cas9n-2\n",
      "single por paired paired\n",
      "           sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-ELANE-Cas9n-2  N714  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 72174                      28546  \n",
      "-----  now making bed file for 3'GSP-ELANE-Cas9n-2   N714   S502\n",
      "made global bed file: N714 _ S502\n",
      "It is called 3'GSP-ELANE-Cas9n-2\n",
      "doing the quantification of global alignments\n",
      "running row 11\n",
      "the analysis is on Guideseq sample: 3'GSP-ELANE-Cas9n-3\n",
      "single por paired paired\n",
      "           sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-ELANE-Cas9n-3  N715  S502                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 53590                      21348  \n",
      "-----  now making bed file for 3'GSP-ELANE-Cas9n-3   N715   S502\n",
      "made global bed file: N715 _ S502\n",
      "It is called 3'GSP-ELANE-Cas9n-3\n",
      "doing the quantification of global alignments\n",
      "running row 12\n",
      "the analysis is on Guideseq sample: 3'GSP-IL7R-Cas9-1\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-IL7R-Cas9-1  N701  S503                 paired                 97943   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      67409  \n",
      "-----  now making bed file for 3'GSP-IL7R-Cas9-1   N701   S503\n",
      "made global bed file: N701 _ S503\n",
      "It is called 3'GSP-IL7R-Cas9-1\n",
      "doing the quantification of global alignments\n",
      "running row 13\n",
      "the analysis is on Guideseq sample: 3'GSP-IL7R-Cas9-2\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-IL7R-Cas9-2  N702  S503                 paired                129644   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      63225  \n",
      "-----  now making bed file for 3'GSP-IL7R-Cas9-2   N702   S503\n",
      "made global bed file: N702 _ S503\n",
      "It is called 3'GSP-IL7R-Cas9-2\n",
      "doing the quantification of global alignments\n",
      "running row 14\n",
      "the analysis is on Guideseq sample: 3'GSP-IL7R-Cas9-3\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-IL7R-Cas9-3  N703  S503                 paired                102204   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      78397  \n",
      "-----  now making bed file for 3'GSP-IL7R-Cas9-3   N703   S503\n",
      "made global bed file: N703 _ S503\n",
      "It is called 3'GSP-IL7R-Cas9-3\n",
      "doing the quantification of global alignments\n",
      "running row 15\n",
      "the analysis is on Guideseq sample: 3'GSP-IL7R-Cas9n-1\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-IL7R-Cas9n-1  N704  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 16339                       8950  \n",
      "-----  now making bed file for 3'GSP-IL7R-Cas9n-1   N704   S503\n",
      "made global bed file: N704 _ S503\n",
      "It is called 3'GSP-IL7R-Cas9n-1\n",
      "doing the quantification of global alignments\n",
      "running row 16\n",
      "the analysis is on Guideseq sample: 3'GSP-IL7R-Cas9n-2\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-IL7R-Cas9n-2  N705  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 55181                      26614  \n",
      "-----  now making bed file for 3'GSP-IL7R-Cas9n-2   N705   S503\n",
      "made global bed file: N705 _ S503\n",
      "It is called 3'GSP-IL7R-Cas9n-2\n",
      "doing the quantification of global alignments\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "running row 17\n",
      "the analysis is on Guideseq sample: 3'GSP-IL7R-Cas9n-3\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-IL7R-Cas9n-3  N706  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 34217                      23174  \n",
      "-----  now making bed file for 3'GSP-IL7R-Cas9n-3   N706   S503\n",
      "made global bed file: N706 _ S503\n",
      "It is called 3'GSP-IL7R-Cas9n-3\n",
      "doing the quantification of global alignments\n",
      "running row 18\n",
      "the analysis is on Guideseq sample: 3'GSP-PRF1-Cas9-1\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-PRF1-Cas9-1  N707  S503                 paired                130455   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      58151  \n",
      "-----  now making bed file for 3'GSP-PRF1-Cas9-1   N707   S503\n",
      "made global bed file: N707 _ S503\n",
      "It is called 3'GSP-PRF1-Cas9-1\n",
      "doing the quantification of global alignments\n",
      "running row 19\n",
      "the analysis is on Guideseq sample: 3'GSP-PRF1-Cas9-2\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-PRF1-Cas9-2  N710  S503                 paired                 58989   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      29615  \n",
      "-----  now making bed file for 3'GSP-PRF1-Cas9-2   N710   S503\n",
      "made global bed file: N710 _ S503\n",
      "It is called 3'GSP-PRF1-Cas9-2\n",
      "doing the quantification of global alignments\n",
      "running row 20\n",
      "the analysis is on Guideseq sample: 3'GSP-PRF1-Cas9-3\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  3'GSP-PRF1-Cas9-3  N711  S503                 paired                 55707   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      40455  \n",
      "-----  now making bed file for 3'GSP-PRF1-Cas9-3   N711   S503\n",
      "made global bed file: N711 _ S503\n",
      "It is called 3'GSP-PRF1-Cas9-3\n",
      "doing the quantification of global alignments\n",
      "running row 21\n",
      "the analysis is on Guideseq sample: 3'GSP-PRF1-Cas9n-1\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-PRF1-Cas9n-1  N712  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 59073                      26602  \n",
      "-----  now making bed file for 3'GSP-PRF1-Cas9n-1   N712   S503\n",
      "made global bed file: N712 _ S503\n",
      "It is called 3'GSP-PRF1-Cas9n-1\n",
      "doing the quantification of global alignments\n",
      "running row 22\n",
      "the analysis is on Guideseq sample: 3'GSP-PRF1-Cas9n-2\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-PRF1-Cas9n-2  N714  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                109455                      44166  \n",
      "-----  now making bed file for 3'GSP-PRF1-Cas9n-2   N714   S503\n",
      "made global bed file: N714 _ S503\n",
      "It is called 3'GSP-PRF1-Cas9n-2\n",
      "doing the quantification of global alignments\n",
      "running row 23\n",
      "the analysis is on Guideseq sample: 3'GSP-PRF1-Cas9n-3\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  3'GSP-PRF1-Cas9n-3  N715  S503                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 63079                      28982  \n",
      "-----  now making bed file for 3'GSP-PRF1-Cas9n-3   N715   S503\n",
      "made global bed file: N715 _ S503\n",
      "It is called 3'GSP-PRF1-Cas9n-3\n",
      "doing the quantification of global alignments\n",
      "running row 24\n",
      "the analysis is on Guideseq sample: 5'GSP-Ctrl-1\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-Ctrl-1  N701  S505                 paired                 17083   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                        627  \n",
      "-----  now making bed file for 5'GSP-Ctrl-1   N701   S505\n",
      "made global bed file: N701 _ S505\n",
      "It is called 5'GSP-Ctrl-1\n",
      "doing the quantification of global alignments\n",
      "running row 25\n",
      "the analysis is on Guideseq sample: 5'GSP-Ctrl-2\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-Ctrl-2  N702  S505                 paired                 29016   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                       9215  \n",
      "-----  now making bed file for 5'GSP-Ctrl-2   N702   S505\n",
      "made global bed file: N702 _ S505\n",
      "It is called 5'GSP-Ctrl-2\n",
      "doing the quantification of global alignments\n",
      "running row 26\n",
      "the analysis is on Guideseq sample: 5'GSP-Ctrl-3\n",
      "single por paired paired\n",
      "    sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-Ctrl-3  N703  S505                 paired                 13115   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                       5683  \n",
      "-----  now making bed file for 5'GSP-Ctrl-3   N703   S505\n",
      "made global bed file: N703 _ S505\n",
      "It is called 5'GSP-Ctrl-3\n",
      "doing the quantification of global alignments\n",
      "running row 27\n",
      "the analysis is on Guideseq sample: 5'GSP-dsODN-1\n",
      "single por paired paired\n",
      "     sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-dsODN-1  N704  S505                 paired                 60370   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      26369  \n",
      "-----  now making bed file for 5'GSP-dsODN-1   N704   S505\n",
      "made global bed file: N704 _ S505\n",
      "It is called 5'GSP-dsODN-1\n",
      "doing the quantification of global alignments\n",
      "running row 28\n",
      "the analysis is on Guideseq sample: 5'GSP-dsODN-2\n",
      "single por paired paired\n",
      "     sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-dsODN-2  N705  S505                 paired                 71289   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      35817  \n",
      "-----  now making bed file for 5'GSP-dsODN-2   N705   S505\n",
      "made global bed file: N705 _ S505\n",
      "It is called 5'GSP-dsODN-2\n",
      "doing the quantification of global alignments\n",
      "running row 29\n",
      "the analysis is on Guideseq sample: 5'GSP-dsODN-3\n",
      "single por paired paired\n",
      "     sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-dsODN-3  N706  S505                 paired                 44450   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      37470  \n",
      "-----  now making bed file for 5'GSP-dsODN-3   N706   S505\n",
      "made global bed file: N706 _ S505\n",
      "It is called 5'GSP-dsODN-3\n",
      "doing the quantification of global alignments\n",
      "running row 30\n",
      "the analysis is on Guideseq sample: 5'GSP-ELANE-Cas9-1\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-ELANE-Cas9-1  N707  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                323608                     317578  \n",
      "-----  now making bed file for 5'GSP-ELANE-Cas9-1   N707   S505\n",
      "made global bed file: N707 _ S505\n",
      "It is called 5'GSP-ELANE-Cas9-1\n",
      "doing the quantification of global alignments\n",
      "running row 31\n",
      "the analysis is on Guideseq sample: 5'GSP-ELANE-Cas9-2\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-ELANE-Cas9-2  N710  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                310880                     299882  \n",
      "-----  now making bed file for 5'GSP-ELANE-Cas9-2   N710   S505\n",
      "made global bed file: N710 _ S505\n",
      "It is called 5'GSP-ELANE-Cas9-2\n",
      "doing the quantification of global alignments\n",
      "running row 32\n",
      "the analysis is on Guideseq sample: 5'GSP-ELANE-Cas9-3\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-ELANE-Cas9-3  N711  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                159584                     155559  \n",
      "-----  now making bed file for 5'GSP-ELANE-Cas9-3   N711   S505\n",
      "made global bed file: N711 _ S505\n",
      "It is called 5'GSP-ELANE-Cas9-3\n",
      "doing the quantification of global alignments\n",
      "running row 33\n",
      "the analysis is on Guideseq sample: 5'GSP-ELANE-Cas9n-1\n",
      "single por paired paired\n",
      "           sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-ELANE-Cas9n-1  N712  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 28614                      21361  \n",
      "-----  now making bed file for 5'GSP-ELANE-Cas9n-1   N712   S505\n",
      "made global bed file: N712 _ S505\n",
      "It is called 5'GSP-ELANE-Cas9n-1\n",
      "doing the quantification of global alignments\n",
      "running row 34\n",
      "the analysis is on Guideseq sample: 5'GSP-ELANE-Cas9n-2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "single por paired paired\n",
      "           sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-ELANE-Cas9n-2  N714  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 40221                      20411  \n",
      "-----  now making bed file for 5'GSP-ELANE-Cas9n-2   N714   S505\n",
      "made global bed file: N714 _ S505\n",
      "It is called 5'GSP-ELANE-Cas9n-2\n",
      "doing the quantification of global alignments\n",
      "running row 35\n",
      "the analysis is on Guideseq sample: 5'GSP-ELANE-Cas9n-3\n",
      "single por paired paired\n",
      "           sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-ELANE-Cas9n-3  N715  S505                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 41774                      32051  \n",
      "-----  now making bed file for 5'GSP-ELANE-Cas9n-3   N715   S505\n",
      "made global bed file: N715 _ S505\n",
      "It is called 5'GSP-ELANE-Cas9n-3\n",
      "doing the quantification of global alignments\n",
      "running row 36\n",
      "the analysis is on Guideseq sample: 5'GSP-IL7R-Cas9-1\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-IL7R-Cas9-1  N701  S506                 paired                181069   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                     161389  \n",
      "-----  now making bed file for 5'GSP-IL7R-Cas9-1   N701   S506\n",
      "made global bed file: N701 _ S506\n",
      "It is called 5'GSP-IL7R-Cas9-1\n",
      "doing the quantification of global alignments\n",
      "running row 37\n",
      "the analysis is on Guideseq sample: 5'GSP-IL7R-Cas9-2\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-IL7R-Cas9-2  N702  S506                 paired                 93187   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      66289  \n",
      "-----  now making bed file for 5'GSP-IL7R-Cas9-2   N702   S506\n",
      "made global bed file: N702 _ S506\n",
      "It is called 5'GSP-IL7R-Cas9-2\n",
      "doing the quantification of global alignments\n",
      "running row 38\n",
      "the analysis is on Guideseq sample: 5'GSP-IL7R-Cas9-3\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-IL7R-Cas9-3  N703  S506                 paired                229430   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                     201394  \n",
      "-----  now making bed file for 5'GSP-IL7R-Cas9-3   N703   S506\n",
      "made global bed file: N703 _ S506\n",
      "It is called 5'GSP-IL7R-Cas9-3\n",
      "doing the quantification of global alignments\n",
      "running row 39\n",
      "the analysis is on Guideseq sample: 5'GSP-IL7R-Cas9n-1\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-IL7R-Cas9n-1  N704  S506                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 36483                      17412  \n",
      "-----  now making bed file for 5'GSP-IL7R-Cas9n-1   N704   S506\n",
      "made global bed file: N704 _ S506\n",
      "It is called 5'GSP-IL7R-Cas9n-1\n",
      "doing the quantification of global alignments\n",
      "running row 40\n",
      "the analysis is on Guideseq sample: 5'GSP-IL7R-Cas9n-2\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-IL7R-Cas9n-2  N705  S506                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 13784                       3953  \n",
      "-----  now making bed file for 5'GSP-IL7R-Cas9n-2   N705   S506\n",
      "made global bed file: N705 _ S506\n",
      "It is called 5'GSP-IL7R-Cas9n-2\n",
      "doing the quantification of global alignments\n",
      "running row 41\n",
      "the analysis is on Guideseq sample: 5'GSP-IL7R-Cas9n-3\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-IL7R-Cas9n-3  N706  S506                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                  4085                       2583  \n",
      "-----  now making bed file for 5'GSP-IL7R-Cas9n-3   N706   S506\n",
      "made global bed file: N706 _ S506\n",
      "It is called 5'GSP-IL7R-Cas9n-3\n",
      "doing the quantification of global alignments\n",
      "running row 42\n",
      "the analysis is on Guideseq sample: 5'GSP-PRF1-Cas9-1\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-PRF1-Cas9-1  N707  S506                 paired                 24035   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      20069  \n",
      "-----  now making bed file for 5'GSP-PRF1-Cas9-1   N707   S506\n",
      "made global bed file: N707 _ S506\n",
      "It is called 5'GSP-PRF1-Cas9-1\n",
      "doing the quantification of global alignments\n",
      "running row 43\n",
      "the analysis is on Guideseq sample: 5'GSP-PRF1-Cas9-2\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-PRF1-Cas9-2  N710  S506                 paired                 95654   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                      81221  \n",
      "-----  now making bed file for 5'GSP-PRF1-Cas9-2   N710   S506\n",
      "made global bed file: N710 _ S506\n",
      "It is called 5'GSP-PRF1-Cas9-2\n",
      "doing the quantification of global alignments\n",
      "running row 44\n",
      "the analysis is on Guideseq sample: 5'GSP-PRF1-Cas9-3\n",
      "single por paired paired\n",
      "         sample_name    i7    i5 single_or_paired_reads  all_alignments_count  \\\n",
      "0  5'GSP-PRF1-Cas9-3  N711  S506                 paired                135509   \n",
      "\n",
      "   filtered_alignments_count  \n",
      "0                     128193  \n",
      "-----  now making bed file for 5'GSP-PRF1-Cas9-3   N711   S506\n",
      "made global bed file: N711 _ S506\n",
      "It is called 5'GSP-PRF1-Cas9-3\n",
      "doing the quantification of global alignments\n",
      "running row 45\n",
      "the analysis is on Guideseq sample: 5'GSP-PRF1-Cas9n-1\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-PRF1-Cas9n-1  N712  S506                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 54217                      42232  \n",
      "-----  now making bed file for 5'GSP-PRF1-Cas9n-1   N712   S506\n",
      "made global bed file: N712 _ S506\n",
      "It is called 5'GSP-PRF1-Cas9n-1\n",
      "doing the quantification of global alignments\n",
      "running row 46\n",
      "the analysis is on Guideseq sample: 5'GSP-PRF1-Cas9n-2\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-PRF1-Cas9n-2  N714  S506                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                 21027                      18242  \n",
      "-----  now making bed file for 5'GSP-PRF1-Cas9n-2   N714   S506\n",
      "made global bed file: N714 _ S506\n",
      "It is called 5'GSP-PRF1-Cas9n-2\n",
      "doing the quantification of global alignments\n",
      "running row 47\n",
      "the analysis is on Guideseq sample: 5'GSP-PRF1-Cas9n-3\n",
      "single por paired paired\n",
      "          sample_name    i7    i5 single_or_paired_reads  \\\n",
      "0  5'GSP-PRF1-Cas9n-3  N715  S506                 paired   \n",
      "\n",
      "   all_alignments_count  filtered_alignments_count  \n",
      "0                119696                      75023  \n",
      "-----  now making bed file for 5'GSP-PRF1-Cas9n-3   N715   S506\n",
      "made global bed file: N715 _ S506\n",
      "It is called 5'GSP-PRF1-Cas9n-3\n",
      "doing the quantification of global alignments\n"
     ]
    }
   ],
   "source": [
    "### This section does two things. It makes a bam file that is filtered for AS and MAPQ \n",
    "#   It quantifies these bam files and sorts them. Then it converts to a bed file.\n",
    "#\n",
    "\n",
    "#### Required Arguments\n",
    "cpu = 12\n",
    "min_MAPQ = 50\n",
    "min_AS = -180\n",
    "\n",
    "### WHAT TYPE OF ANALYSIS ARE WE RUNNING\n",
    "#####type in 'on' or 'off'\n",
    "local = 'off' #local is softclipped NEED TO MAKE THIS TO USE IT\n",
    "global_align_branch = 'on' #global is end-to-end and this does the analysis\n",
    "#this does the acutal alignment for end-to-end\n",
    "do_align_global = 'on'\n",
    "single_or_paired_reads = \"paired\"  #can be 'single' or 'paired' for the read alignment\n",
    "\n",
    "#### WHAT SAMPLES TO RUN ####\n",
    "#lines_to_run = [20]\n",
    "lines_to_run = range(48)\n",
    "\n",
    "#for exporting summary of data\n",
    "results_df_all = pd.DataFrame()\n",
    "results_file = os.path.join(directory, 'results','alignment_filtered_summary.xlsx')\n",
    "\n",
    "\n",
    "for i in lines_to_run:\n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    print('running row', i)\n",
    "    print('the analysis is on Guideseq sample:', amplicon_info['name'])\n",
    "    reference_assembly = amplicon_info['genome']\n",
    "    sample_name = amplicon_info['name']\n",
    "    \n",
    "    if global_align_branch == 'on':\n",
    "        if do_align_global == 'on':\n",
    "            #####this aligns everything after the break to end-to-end keep_sam=1 means keep the sam file\n",
    "            align_guideseq_end_to_end_genome_global(directory, amplicon_info, reference_assembly, single_or_paired_reads, cpu, keep_sam=1)\n",
    "\n",
    "            \n",
    "        #### Process the alignment files to make a bam file of only high quality reads\n",
    "        \n",
    "        #this filters for good alignments based on AS score and MapQ and exports a final indexed bam file.\n",
    "\n",
    "        df_sample_results = guideseq_filtered_mapq_AS_primary(directory, 'global', amplicon_info, single_or_paired_reads)\n",
    "        #this is a running dataframe of the clenaed up read summary\n",
    "        results_df_all = results_df_all.append(df_sample_results, ignore_index=True)\n",
    "        \n",
    "        \n",
    "        ### global alignment bed generation \n",
    "        N7 = amplicon_info['index_I1']\n",
    "        N5 = amplicon_info['index_I2']\n",
    "        print('-----  now making bed file for', sample_name, ' ',N7, ' ', N5 )\n",
    "        \n",
    "        \n",
    "        ##### Need to fix this!!!!!!!!!!!! the names of the input are not right and mabye i need to sort things\n",
    "        if single_or_paired_reads == \"single\":\n",
    "            genome_global_bed_file = create_filename(directory, N7, N5, 'guideseq_global_bed_single')\n",
    "            filtered_and_sorted_bam_file = create_filename(directory, N7, N5, 'filtered_and_sorted_genome_global_single')\n",
    "        elif single_or_paired_reads == \"paired\":\n",
    "            genome_global_bed_file = create_filename(directory, N7, N5, 'guideseq_global_bed_paired')\n",
    "            filtered_and_sorted_bam_file = create_filename(directory, N7, N5, 'filtered_and_sorted_genome_global_paired')\n",
    "    \n",
    "        !bam2bed < {filtered_and_sorted_bam_file} > {genome_global_bed_file}\n",
    "        bed_folder = os.path.join(directory, 'bed_files')\n",
    "\n",
    "        if not os.path.exists(bed_folder):\n",
    "            os.mkdir(bed_folder)\n",
    "        \n",
    "        file_location = create_filename(directory, N7, N5, 'bam_global')\n",
    "        \n",
    "        new_bed_name = os.path.join(file_location, sample_name + '_' + single_or_paired_reads + '.bed')\n",
    "\n",
    "        os.rename(genome_global_bed_file, new_bed_name)\n",
    "\n",
    "        shutil.copy(new_bed_name, bed_folder)\n",
    "        \n",
    "        \n",
    "        \n",
    "\n",
    "        print('made global bed file:', N7, '_', N5 )\n",
    "        print('It is called',sample_name )\n",
    "        print('doing the quantification of global alignments')\n",
    "\n",
    " \n",
    "    \n",
    "    ########### I don't use local so ignore this #############\n",
    "    if local == 'on':\n",
    "        #### this aligns everything in the local format wehre it can soft clip the ends keep_sam=1 means keep the sam file\n",
    "        # need to make this section if I want to use it\n",
    "        align_afterbreaks_genome_local(directory, 1, amplicon_info, assembly_plus_targetvector, lam_or_tn5='lam', keep_sam=0)\n",
    "        \n",
    "        #make bed file\n",
    "        print('-----  now making bed file   ------')\n",
    "        N7 = amplicon_info['index_I1']\n",
    "        N5 = amplicon_info['index_I2']\n",
    "        file_trimmed_genome_local_bed = create_filename(directory, N7, N5, 'break_trimmed_genome_local_bed')\n",
    "        file_sorted_bam_genome_local = create_filename(directory, N7, N5, 'break_trimmed_filtered_and_sorted_bam_genome_local')\n",
    "\n",
    "        !bam2bed < {file_sorted_bam_genome_local} > {file_trimmed_genome_local_bed}\n",
    "        bed_folder = os.path.join(directory, 'bed_files')\n",
    "        \n",
    "        if not os.path.exists(bed_folder):\n",
    "            os.mkdir(bed_folder)\n",
    "        shutil.copy(file_trimmed_genome_local_bed, bed_folder)\n",
    "    \n",
    "        \n",
    "        print('made local bed file:', N7, '_', N5 )\n",
    "\n",
    "        # running overview of alignment quantification\n",
    "        print('doing the quantification of local alignments')\n",
    "        quantify_local = quantify_pipeline2_alignments(directory, 'local', amplicon_info, 'lam')\n",
    "        print(quantify_local)\n",
    "\n",
    "#export all the read summaries\n",
    "results_df_all.to_excel(results_file)    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "listnumber2                 object\n",
      "Sequence                    object\n",
      "Name                        object\n",
      "Top off-target sites        object\n",
      "Strand                      object\n",
      "MutDist                      int64\n",
      "Risk                        object\n",
      "Annotation (if relevant)    object\n",
      "chr                         object\n",
      "position_start               int64\n",
      "position_end                object\n",
      "begin_site_range             int64\n",
      "end_site_range               int64\n",
      "dtype: object\n",
      "running row 1\n",
      "the analysis is on Guideseq sample: 2-1-3'GSP-TD\n",
      "index\n",
      "index N702 S502\n",
      "/home/eric/Data/Spaced_Nicking/GuideSeq_MiniSeq1/results/N702_S502_2-1-3'GSP-TD_guide_count.xlsx\n",
      "/home/eric/Data/Spaced_Nicking/GuideSeq_MiniSeq1/N702_S502/bam_genome_global_files\n",
      "/home/eric/Data/Spaced_Nicking/GuideSeq_MiniSeq1/N702_S502/bam_genome_global_files/2-1-3'GSP-TD_paired.bed\n",
      "chr                     object\n",
      "chromStart               int64\n",
      "chromEnd                 int64\n",
      "name                    object\n",
      "MapQ                     int64\n",
      "strand                  object\n",
      "flag                     int64\n",
      "CIGAR                   object\n",
      "paired_name             object\n",
      "read1start               int64\n",
      "paired_amplicon_size     int64\n",
      "sequence                object\n",
      "quality_score           object\n",
      "a                       object\n",
      "b                       object\n",
      "c                       object\n",
      "d                       object\n",
      "e                       object\n",
      "f                       object\n",
      "g                       object\n",
      "h                       object\n",
      "i                       object\n",
      "dtype: object\n",
      "list of read hits: [1, 2, 3, 4, 5, 6]\n"
     ]
    }
   ],
   "source": [
    "############ Use CIRPSR Gold and look at overlaps\n",
    "\n",
    "single_or_paired_reads = \"paired\"  #can be 'single' or 'paired' for the read alignment\n",
    "windowsize = 500\n",
    "\n",
    "#### WHAT SAMPLES TO RUN ####\n",
    "lines_to_run = [1]\n",
    "#lines_to_run = range(30)\n",
    "\n",
    "\n",
    "#off_target_sites \n",
    "crispr_gold_file = '/home/eric/Data/Spaced_Nicking/GuideSeq_MiniSeq1/CrispRgold_predictions.xlsx'\n",
    "df_targets = pd.read_excel(crispr_gold_file, index_col=0)\n",
    "\n",
    "# new data frame with split guide location. CRISPRGOLD puts it all in at once\n",
    "new = df_targets[\"Position\"].str.split(\":\", expand = True) \n",
    "df_targets[\"chr\"]= new[0] \n",
    "df_targets[\"position_start\"]= new[1] \n",
    "df_targets[\"position_end\"]= new[2] \n",
    "df_targets.drop(columns =[\"Position\"], inplace = True) \n",
    "df_targets['position_start'] = df_targets['position_start'].astype(int)\n",
    "df_targets['begin_site_range'] = df_targets['position_start'] - windowsize\n",
    "df_targets['end_site_range'] = df_targets['position_start'] + windowsize\n",
    "\n",
    "print(df_targets.dtypes)\n",
    "\n",
    "#print(df_targets)\n",
    "\n",
    "\n",
    "for i in lines_to_run:\n",
    "    \n",
    "    amplicon_info = get_csv_data(directory, i)\n",
    "    name = amplicon_info['name']\n",
    "    print('running row', i)\n",
    "    print('the analysis is on Guideseq sample:', name)\n",
    "    print('index',)\n",
    "    \n",
    "    #import bed file\n",
    "    N7 = amplicon_info['index_I1']\n",
    "    N5 = amplicon_info['index_I2']\n",
    "    print('index', N7, N5)\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    print(results_file)\n",
    "\n",
    "    file_location = create_filename(directory, N7, N5, 'bam_global')   \n",
    "    print(file_location)\n",
    "    sample_name = amplicon_info['name']\n",
    "    bedfile = os.path.join(file_location, sample_name + '_' + single_or_paired_reads + '.bed')\n",
    "    print(bedfile)\n",
    "    \n",
    "    #needs to add header names to specify the number of columns so that \n",
    "    \n",
    "    header_names = ['chr', 'chromStart', 'chromEnd', 'name', 'MapQ', 'strand', 'flag', 'CIGAR', 'paired_name', 'read1start', 'paired_amplicon_size', 'sequence', 'quality_score','a','b','c','d','e','f','g','h','i']\n",
    "\n",
    "#    df_bed = pd.read_csv(bedfile, sep='\\t', comment='t', header=None)\n",
    "    df_bed = pd.read_csv(bedfile, sep='\\t', header=None, names = header_names)\n",
    "\n",
    "    \n",
    "    \n",
    "    #header = ['chr', 'chromStart', 'chromEnd', 'name', 'MapQ', 'strand', 'flag', 'CIGAR', 'paired_name', 'read1start', 'paired_amplicon_size', 'sequence', 'quality_score','a','b','c','d','e','f','g','h','i']\n",
    "    #df_bed.columns = header[:len(df_bed.columns)]\n",
    "    \n",
    "    print(df_bed.dtypes) \n",
    "    \n",
    "    read_hits = []\n",
    "    count = 0\n",
    "    for i,g in itertools.izip(df_bed.chromStart, df_bed.chr):\n",
    "        bed_alignment = i\n",
    "        chromosome = g\n",
    "        for a,b,c,d in itertools.izip(df_targets.listnumber2, df_targets.begin_site_range, df_targets.end_site_range, df_targets.chr): \n",
    "            counter_A = 0\n",
    "            if bed_alignment in range(b, c) and chromosome == d:\n",
    "                if counter = 0:\n",
    "                    print(b,c,'chromosome', d, ' guide:', a)\n",
    "                    read_hits.append(a)\n",
    "                    counter_A += 1\n",
    "                    print(b,c,'chromosome', d, ' guide:', a, 'for read starting at :', i)\n",
    "        count += 1 \n",
    "        if count >30:\n",
    "            break\n",
    "    results_df = pd.DataFrame({'sample_name': [name],\n",
    "                               'i7': [N7],\n",
    "                               'i5': [N5],\n",
    "                               'list_of_read_guide_matches' : [read_hits],\n",
    "                               },\n",
    "                                  columns=['sample_name',\n",
    "                                           'i7',\n",
    "                                           'i5',\n",
    "                                           'list_of_read_guide_matches'])            \n",
    "    \n",
    "    \n",
    "    \n",
    "    results_file = os.path.join(directory, 'results',N7 + '_' + N5 + '_'+ name +'_guide_count.xlsx')\n",
    "    \n",
    "    results_df.to_excel(results_file)    \n",
    "\n",
    "    print('list of read hits:',read_hits)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
